---
title: "STRATAA Microbiome - metadata basics"
author: "Philip Ashton"
date: "`r Sys.Date()`"
output: 
  html_document:
      toc: true
      toc_float: true
      theme: united
      number_sections: true
---

```{r setup, error=FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)
```

# STRATAA microbiome metadata basics

## Sources

The file handles are set in config.R as they're used by multiple scripts.

The imports are done in the core_functions file.

```{r source, error=FALSE}
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/00.core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```

## Read in metadata

```{r metadata_read_in, error=FALSE}
metadata <- read_metadata(metadata_handle)
# putting this here so that the output files of maaslin get named accroding to the variable names in the metadata file.
metadata <- metadata %>% mutate(Group = if_else(Group == 'Control_HealthySerosurvey', 'Household contact', Group)) %>% mutate(Group = if_else(Group == 'Acute_Typhi', 'Acute typhoid', Group)) %>% mutate(Group = if_else(Group == 'Carrier', 'High Vi-titre', Group))
```

```{r metadata_basics, error=FALSE}
number_per_country <- metadata %>% group_by(Country) %>% summarise(count = n())

# number per country
number_per_country %>% kbl() %>% kable_styling()

print(sum(number_per_country$count))

# number per group, and median age
metadata %>% group_by(Group) %>% summarise(count = n(), median_age = median(Age)) %>% kbl() %>% kable_styling()

# number per country per group
metadata %>% group_by(Country, Group) %>% summarise(count = n()) %>% kbl() %>% kable_styling()


baseline_chars <- get_baseline_characteristics(metadata)
comparisons_groups <- list(c("Acute typhoid", "High Vi-titre"), c("Acute typhoid", "Household contact"), c("High Vi-titre", "Household contact"))

ggboxplot(metadata, y = "Age", x = "Group", color = "Group") + 
  stat_compare_means() + 
  stat_compare_means(comparisons = comparisons_groups) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rremove("x.ticks") +
  scale_color_manual(values = participant_group_colours)
   # +rotate_x_text(angle = 45)


baseline_chars_table <- baseline_chars %>% 
  rename(`Median age` = median_age, `Women (%)` = pct_fem, `Antibiotics in last 2 weeks (%)` = pct_anti) %>% 
  pivot_longer(!c(Country, Group)) %>% 
  rename(variable_name = name) %>% 
  pivot_wider(names_from = Country, values_from = value)

baseline_chars_table %>% kbl(digits = c(NA, NA, 1, 1, 1)) %>% kable_styling()

# the kruskal wallis test is positive, showing a difference between the groups, so we move onto pairwise tests.
# ggplot(metadata, aes(x = Country, y = Age, fill = Group)) + geom_boxplot() + stat_compare_means(method = 'kruskal.test', label = "p")
comparisons_countries <- list(c('Bangladesh', 'Malawi'), c('Bangladesh', 'Nepal'), c('Malawi', 'Nepal'))
# default stat method when doing pairwise tests in wilcoxon.
metadata_for_plotting <- metadata
metadata_for_plotting$Group <- factor(metadata_for_plotting$Group, levels = c("Household contact", "Acute typhoid", "High Vi-titre"))
ggboxplot(metadata_for_plotting, facet.by = "Country", y = "Age", x = "Group", color = "Group") + stat_compare_means(comparisons = comparisons_groups) + rremove("x.text") + rremove("xlab") + rremove("x.ticks") + scale_color_manual(values = participant_group_colours) # +rotate_x_text(angle = 45)
# ggboxplot(metadata_for_plotting, facet.by = "Group", y = "Age", x = "Country", color = "Country") + stat_compare_means(comparisons = comparisons_countries) + rotate_x_text(angle = 45)

```

```{r get_age_range}


```

```{r metadata_basic_plots, error=FALSE, fig.width = 4, fig.height= 8}
country_group_sex <- metadata %>% group_by(Country, Group, Sex) %>% summarise(count = n()) 
comparisons_groups <- list(c("Acute typhoid", "High Vi-titre"), c("Acute typhoid", "Household contact"), c("High Vi-titre", "Household contact"))
# country_group_sex %>% group_by(Country, Sex) %>% summarise(n = sum(count))
plot_sex <- function(eg1, c){
  d <- eg1 %>% filter(Country == c)
  p <- ggplot(d, aes(x = Group, y = count, fill = Sex)) + 
    geom_bar(stat ='identity', position = 'fill') + 
    ylab('Proportion') + 
    ggtitle(c)
    # +
    #theme(axis.text=element_text(size=34), axis.title=element_text(size=36,face="bold"), plot.title = element_text(size = 40, face = "bold"), legend.key.size = unit(4, 'cm'), legend.title = element_text(size = 34), legend.text = element_text(size = 28))
  
  return(p)
}

country_group_sex$Group <- factor(country_group_sex$Group, levels = c("Household contact", "Acute typhoid", "High Vi-titre"))

m_sex <- plot_sex(country_group_sex, 'Malawi')
b_sex <- plot_sex(country_group_sex, 'Bangladesh')
n_sex <- plot_sex(country_group_sex, 'Nepal')
m_sex / b_sex / n_sex 
```

## sequencing read stats

```{r sequencing_read_stats, error=FALSE}

# todo - replace with the read stats from the metadata file.
# read_stats_path <- '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/data_munging/2023.08.28/2023.08.28 strataa microbiome number of reads.csv'
# read_stats <- read_csv(read_stats_path) %>% filter(Country != '#N/A')

ggboxplot(metadata, facet.by = "Country", x = "Group", y = "number_of_reads", color = "Group") + rremove("x.text") + rremove("xlab") + rremove("x.ticks") + rotate_x_text(angle = 45) + stat_compare_means()

```