
---
title: "STRATAA Microbiome data processing"
author: "Philip Ashton"
date: "`r Sys.Date()`"
output: 
    html_document:
        toc: true
        toc_float: true
        theme: united
        number_sections: true

---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```

## Sources
```{r imports}

source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/00.core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```


## read in data

Metadata
```{r read_in_metadata}
metadata <- read_metadata(metadata_handle)

rownames(metadata) <- metadata$SampleID
strataa_metaphlan_metadata <- metadata %>% dplyr::select(SampleID, sequencing_lane, Group, Sex, Country, Age, Antibiotics_taken_before_sampling_assumptions, sequencing_lane)

# strataa_metaphlan_metadata %>% filter(Group == 'Acute_Typhi', Country == 'Malawi') %>% group_by(Antibiotics_taken_before_sampling_assumptions) %>% summarise(n = n())
# strataa_metaphlan_metadata
```

Metaphlan data
```{r read_in_metaphlan, include=FALSE}
strataa_metaphlan_data = read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.all_strataa_metaphlan.csv'), header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
```

bigmap data

```{r read_in_bigmap_results}

# strataa_bigmap_data = read.csv(strataa_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_disease_health_bigmap_data = read.csv(file = strataa_disease_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_carrier_health_bigmap_data = read.csv(file = strataa_carrier_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_nepal_bigmap_data <- read.csv(file = strataa_nepal_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)

rownames(strataa_disease_health_bigmap_data) <- rownames(strataa_disease_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')
rownames(strataa_carrier_health_bigmap_data) <- rownames(strataa_carrier_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')
rownames(strataa_nepal_bigmap_data) <- rownames(strataa_nepal_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')

# strataa_bigmap_metadata is the same as strataa_metaphlan_metadata
#strataa_bigmap_metadata = read.csv(file = strataa_bigmap_metadata_path, header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

#strataa_bigmap_metadata <- strataa_bigmap_metadata %>% filter(Gender != '#N/A')
#strataa_bigmap_metadata$Age <- as.numeric(as.character(strataa_bigmap_metadata$Age))

strataa_bigmap_metadata <- metadata %>% mutate(isolate_bigmap = str_replace_all(isolate, '_', '.'))
rownames(strataa_bigmap_metadata) <- strataa_bigmap_metadata$isolate_bigmap

```

## Run MAASLIN2 taxonomic

Acute vs household contact.

```{r run_maaslin_metaphlan_healthy_typhoid include=FALSE}
#variables_for_analysis <- c("Group")
groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")

run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"), 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
# variables_for_analysis <- c("Group", "Sex")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG',  NA, 'metaphlan')
```


For healthy vs carrier, not including antibiotics becayse they're all "No"

```{r run_maaslin_metaphlan_healthy_carrier}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')

#variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")
variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG',  c("sequencing_lane,32580_7"), 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'metaphlan')
```

health vs acute vs carrier

```{r}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey', 'Acute_Typhi')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey", "sequencing_lane,32712_2"), 'metaphlan')
# checking the effect of excluding sequencing lane
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey"), 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey"), 'metaphlan')

```



## run maaslin2 functional

### typhi vs health

- todo = haven't changed the function for including the prevalence, probably just need to filter/select the meta-data or something?

```{r }

groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_disease_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", 'sequencing_lane')
run_maaslin(strataa_disease_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"), 'bigmap')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_nepal_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')

```

#### grouping reads by mgc class and doing maaslin


```{r}
# manually add the mgc suffix to the output dir in the run_maaslin function.
# todo = add a parameter to the function to allow this to be done automatically
strataa_disease_health_bigmap_data_l <- strataa_disease_health_bigmap_data %>% rownames_to_column(var = 'sample_id') %>%  
  pivot_longer(!sample_id, names_to = "mgc_class", values_to = "reads") %>%
  mutate(extracted_column = str_extract(mgc_class, "(?<=Entryname=)[^\\-]+"))

strataa_disease_health_bigmap_mgc <- strataa_disease_health_bigmap_data_l %>% group_by(sample_id, extracted_column) %>% summarise(reads = sum(reads)) %>% ungroup() %>% pivot_wider(names_from = extracted_column, values_from = reads, values_fill = 0) %>% column_to_rownames(var = 'sample_id')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_disease_health_bigmap_mgc, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", 'sequencing_lane')
run_maaslin(strataa_disease_health_bigmap_mgc, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"), 'bigmap')


strataa_nepal_disease_health_bigmap_data_l <- strataa_nepal_bigmap_data %>% rownames_to_column(var = 'sample_id') %>%  
  pivot_longer(!sample_id, names_to = "mgc_class", values_to = "reads") %>%
  mutate(extracted_column = str_extract(mgc_class, "(?<=Entryname=)[^\\-]+"))

strataa_nepal_disease_health_bigmap_mgc <- strataa_nepal_disease_health_bigmap_data_l %>% group_by(sample_id, extracted_column) %>% summarise(reads = sum(reads)) %>% ungroup() %>% pivot_wider(names_from = extracted_column, values_from = reads, values_fill = 0) %>% column_to_rownames(var = 'sample_id')


variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
run_maaslin(strataa_nepal_disease_health_bigmap_mgc, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')


```
### carrier vs health

```{r}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')

# variables_for_analysis <- c("Group", "Sex", "Age")
# variables_for_analysis <- c("Group")

variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
run_maaslin(strataa_carrier_health_bigmap_data, strataa_metaphlan_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32712_2"), 'bigmap')

variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_nepal_bigmap_data, strataa_metaphlan_metadata, maaslin_functional_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')

```

