## Imports
```{r}
# library(rbiom)
library(kableExtra)
library(ggplot2)
library(ggpubr)
# library(ggforce)
library(patchwork)
# library(vegan)
library(dplyr)
library(purrr)

# library(forcats)
# library(RColorBrewer)

```
## Sources

The file handles are set in config.R as they're used by both this script and data_cleaning.

```{r source}
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```


## Metadata basics

First do the plots with kruskall wallis comparison between groups to see if there's an overall difference, then do the pairwise tests because KW says there is a difference.


```{r metadata_read_in}
metadata <- read_metadata(metadata_handle)

metadata <- metadata %>% mutate(Group = if_else(Group == 'Control', 'Control', Group))
metadata <- metadata %>% mutate(Group = if_else(Group == 'Acute typhoid', 'Acute typhoid', Group))

```


##Â amr analysis

Code here is not great, maybe come back and move some into functions/utils.

```{r amr_paths}
amr_data_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/2023.08.30.strataa_microbiome.subsamp.rgi.txt"
ncbi_amr_info_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/ncbi_amr/2023.08.31/refgenes.modified.csv"

```

```{r amr_read_in}

ncbi_amr_info <- read_csv(ncbi_amr_info_path)

amr_data <- read_delim(amr_data_path, delim = "\t") %>% separate_wider_delim(`ARO Term`, names = c("sample", "ARO Term"), delim = ".gene_mapping_data.txt:") %>% mutate(reads_per_kb = `Completely Mapped Reads` / (`Reference Length` / 1000)) 

bgd_carriers <- metadata %>% filter(Country == 'Bangladesh') %>% filter(Group == 'Carrier') %>% pull(Lane)
# filter out the bgd carriers
amr_data <- amr_data %>% filter(!(sample %in% bgd_carriers))

amr_data_with_ncbi <- left_join(amr_data, ncbi_amr_info, by = join_by(`ARO Term` == combined_name))

# take the "found" genes
# we filter for 100% cov here, and these are the "present" genes
amr_data_with_ncbi_cln <- amr_data_with_ncbi %>% filter(`Average Percent Coverage` == 100) %>% filter(Class != 'NA') %>% filter(Scope == 'core')

# summarise the data (by class)
amr_data_with_ncbi_cln_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))
# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)
# join the data to the template (amr_summarised_meta)
amr_summarised_meta_with_data <- left_join(amr_summarised_meta, amr_data_with_ncbi_cln_summarised, by = join_by(sample == sample, Class == Class)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% mutate(Class = ifelse(is.na(Class), "None", Class))

# amr_data_with_ncbi_cln_meta <- left_join(amr_data_with_ncbi_cln, metadata, by = join_by(sample == Lane))
# View(amr_summarised_meta_with_data %>% filter(Class == 'AMINOGLYCOSIDE') %>% select(sample, Class, n, Group, Country) %>% arrange(Country, Group))

```

Need to setup a data frame with every sample/drug combination so that capture the zeroes. 

```{r setup_df, include=FALSE}

## need to remove the Bangladesh carriers from this.

write_csv(amr_data_with_ncbi, file.path('~/Desktop', "amr_data_with_ncbi.csv"))

# 1. set the samples and classes to be analysed
samples <- metadata %>% pull(Lane)
classes <- distinct(amr_data_with_ncbi, Class) %>% filter(!is.na(Class)) %>% pull(Class) 

# 2. make an empty data frame, with just the samples and classes
amr_summarised <- expand.grid(sample = samples, Class = classes)
# 3. add the metadata
amr_summarised_meta <- left_join(amr_summarised, metadata, by = join_by(sample == Lane))

```

## Graphs of overall comparison. Do not include carriers here, because of different sample handling in Bangladesh.

```{r overall_graphs_per_country}
number_of_amr_genes_per_sample <- amr_summarised_meta_with_data %>% filter(!(Group == 'Carrier')) %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
number_of_amr_genes_per_sample  %>% group_by(Country) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
# number of amr genes per sample, by country
p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  ylab('Number of AMR genes identified per sample') +
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_boxplot.png"), plot = p, width = 5, height = 4)

p
# rpk amr genes per sample by country
p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpkm_boxplot.png"), plot = p, width = 5, height = 4)

p
```


```{r overall_graphs_per_group}
number_of_amr_genes_per_sample_group <- amr_summarised_meta_with_data %>% filter(!(Group == 'Carrier' & Country == 'Bangladesh')) %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
number_of_amr_genes_per_sample_group$Group <- factor(number_of_amr_genes_per_sample_group$Group, levels = c("Control", "Acute typhoid", "Carrier"))

number_of_amr_genes_per_sample_group  %>% group_by(Group) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
# number of amr genes per sample, by country
p <- ggplot(number_of_amr_genes_per_sample_group, aes(x = Group, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  ylab('Number of AMR genes identified per sample') +
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_boxplot_group.png"), plot = p, width = 5, height = 4)

p
# rpk amr genes per sample by country
p <- ggplot(number_of_amr_genes_per_sample_group, aes(x = Group, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpkm_boxplot_group.png"), plot = p, width = 5, height = 4)

p
```


```{r amr_with_antibiotic_treatment}

number_of_amr_genes_per_sample_abu <- amr_summarised_meta_with_data %>% filter(Group == 'Acute typhoid') %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb))
number_of_amr_genes_per_sample_abu_meta <- left_join(number_of_amr_genes_per_sample_abu, metadata, by = join_by(sample == Lane))

number_of_amr_genes_per_sample_abu_meta  %>% group_by(Antibiotics_taken_before_sampling_yes_no_assumptions) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
number_of_amr_genes_per_sample_abu_meta$Antibiotics_taken_before_sampling_yes_no_assumptions <- factor(number_of_amr_genes_per_sample_abu_meta$Antibiotics_taken_before_sampling_yes_no_assumptions, levels = c("Yes", "No", "Unknown"))

p <- ggplot(number_of_amr_genes_per_sample_abu_meta, aes(x = Antibiotics_taken_before_sampling_yes_no_assumptions, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_amu.png"), plot = p, width = 5, height = 4)

p

p <- ggplot(number_of_amr_genes_per_sample_abu_meta, aes(x = Antibiotics_taken_before_sampling_yes_no_assumptions, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpk_amu.png"), plot = p, width = 5, height = 4)

p

```

```{r ncbi_amr_graphs}

# amr_data_with_ncbi_cln_meta_summarised <- amr_data_with_ncbi_cln_meta %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
amr_summarised_meta_with_data$Group <- factor(amr_summarised_meta_with_data$Group, levels = c("Control", "Acute typhoid", "Carrier"))

# classes <- unique(amr_summarised_meta_with_data$Class)
# cut down classes with some manual removal of low numbers
classes <- c("AMINOGLYCOSIDE", "BETA-LACTAM", "LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN", "TETRACYCLINE", "TRIMETHOPRIM", "EFFLUX", "SULFONAMIDE")

# by country by group, number of genes
for (drug_class in classes) {
  # print(drug_class)
  drug_class_data <- filter(amr_summarised_meta_with_data, Class == drug_class)
  # View(drug_class_data)
  p <- ggviolin(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group", draw_quantiles = 0.5) + 
    stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
    stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
    rremove("x.text") + 
    rremove("xlab") + 
    rotate_x_text(angle = 45) +
    ggtitle(drug_class) +
    geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
  ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/number/by_country_by_group',  paste(gsub("/", "_", drug_class), "number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, number of genes
# remove carrier because not valid in bangladesh
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class, Group != 'Carrier')
    p <- ggboxplot(drug_class_data, y = "n", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/number/by_country',  paste(gsub("/", "_", drug_class), "number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}
# by country, by group, rpkm
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class)
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    # print(drug_class)
    # View(drug_class_data)
    
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) + scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/rpkb/by country by group', paste(gsub("/", "_", drug_class), "rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, rpkm
for (drug_class in classes) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class, Group != 'Carrier')
    # add 1 to the reads so that the log plot works ok
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = log10(max(drug_class_data$sum_reads_per_kb) * 0.9)) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0) +
      scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/rpkb/by country', paste(gsub("/", "_", drug_class), "rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```

```{r esbl_data}
esbls <- amr_data_with_ncbi_cln %>% filter(grepl('extended-spectrum', `Product name`))
esbls_template <- expand.grid(sample = samples, Subclass = 'ESBL')

esbls_meta <- left_join(esbls, metadata, by = join_by(sample == Lane))
esbls_summarised <- esbls_meta %>% group_by(sample) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))
esbls_all_samples_summarised <- left_join(esbls_template, esbls_summarised, by = join_by(sample == sample)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb))
esbls_summarised_meta <- left_join(esbls_all_samples_summarised, metadata, by = join_by(sample == Lane))
esbls_summarised_meta$Group <- factor(esbls_summarised_meta$Group, levels = c("Control", "Acute typhoid", "Carrier"))

```

```{r esbl_graphs}

p <- ggboxplot(esbls_summarised_meta, facet.by = "Country", y = "n", x = "Group", color = "Group") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
p
p <- ggboxplot(esbls_summarised_meta, y = "sum_reads_per_kb", x = "Country", color = "Country") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  scale_y_log10()
p

# ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)

write_csv(esbls_summarised_meta, file.path('~/Desktop', 'esbls_summarised_meta.csv'))

p <- ggboxplot(esbls_summarised_meta, y = "n", x = "Country", color = "Country") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  geom_jitter(aes(color = Country), width = 0.2, height = 0, alpha = 0.2)
p
```


```{r betalactam_graphs}
# 1. set the samples and classes to be analysed (samples set above)
# 2. make an empty data frame, with just the samples and classes
amr_summarised_betalac <- expand.grid(sample = samples, Subclass = subclasses)
# 3. add the metadata
amr_summarised_betalac_meta <- left_join(amr_summarised_betalac, metadata, by = join_by(sample == Lane))
# 4. take the found genes (can just use amr_data_with_ncbi_cln from above)
# 5. summarise the data (by Subclass)
amr_data_with_ncbi_cln_betalac_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Subclass) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))

# 6. join the data to the template (amr_summarised_betalac_meta)
amr_summarised_meta_with_data_betalac <- left_join(amr_summarised_betalac_meta, amr_data_with_ncbi_cln_betalac_summarised, by = join_by(sample == sample, Subclass == Subclass)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb))
View(amr_summarised_meta_with_data_betalac %>% filter(Subclass == 'CARBAPENEM', n > 0))

# only four carbapenem samples, so taking out.
amr_summarised_meta_with_data_betalac <- amr_summarised_meta_with_data_betalac %>% filter(!(Subclass == 'CARBAPENEM'))
# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)

amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control", "Acute typhoid", "Carrier"))

# amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control", "Acute typhoid", "Carrier"))
subclasses <- c('BETA-LACTAM', 'CEPHALOSPORIN')

# by country by group, boxplot of number of genes
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group") + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of number of genes
for (drug_class in subclasses) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class, Group != 'Carrier')
    p <- ggboxplot(drug_class_data, y = "n", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +       rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country by group, boxplot of rpkm
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of rpkm
for (drug_class in subclasses) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class, Group != 'Carrier')
    # View(drug_class_data)
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/rpk', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```

## Typhi gene analysis

Do an analysis of only genes found in Typhi. allele_names and gene_family_names come from the pathogen watch paper supp table 2.

```{r typhi_only}
allele_names <- c('blaCMY-2', 'blaOXA-1', 'blaOXA-11', 'blaOXA-15', 'blaOXA-7', 'blaTEM-1', 'blaTEM-10', 'blaTEM-124', 'cmlA1')
gene_family_names <- c('blaAIM', 'blaBIC', 'blaCTX-M', 'blaGES', 'blaIMI', 'blaIMP', 'blaKPC', 'blaLMB', 'blaNDM', 'blaSHV', 'blaSIM', 'blaSPM', 'blaVIM', 'dfrA1', 'dfrA15', 'dfrA17', 'dfrA18', 'dfrA5', 'dfrA7', 'drfA14', 'erm(A)', 'erm(B)', 'erm(C)', 'mcr-1', 'mcr-2', 'mcr-3', 'mcr-4', 'mcr-5', 'mcr-6', 'mcr-7', 'mcr-8', 'mph(A)', 'mph(B)', 'qnrA', 'qnrB', 'qnrD', 'qnrS', 'sul1', 'sul2', 'tet(A)', 'tet(B)', 'tet(C)', 'tet(D)')

amr_data_typhi <- amr_data_with_ncbi_cln %>% filter(`#Allele` %in% allele_names | `Gene family` %in% gene_family_names)
typhi_classes <- unique(amr_data_typhi$Class)
samples <- metadata %>% filter(!(Country == 'Bangladesh' & Group == 'Carrier')) %>% pull(Lane)

# 1. make a template with all the samples and classes
amr_summarised_typhi <- expand.grid(sample = samples, `Class` = typhi_classes)

# 2. count the number of genes per sample, per class.
tmp <- amr_data_typhi %>% group_by(sample, Class) %>% summarise(n = n()) #%>% kbl() %>% kable_styling()

# 3. join the data to the template (amr_summarised_typhi)
amr_summarised_typhi_d <- left_join(amr_summarised_typhi, tmp, by = join_by(sample == sample, Class == Class)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(at_least_one = ifelse(n > 0, 1, 0)) 

# 4. join to metadata
amr_summarised_typhi_d_meta <- left_join(amr_summarised_typhi_d, select(metadata, c(Lane, Country, Group)), by = join_by(sample == Lane))


# 5. Group by class, country, and summarise
typhi_amr_class_country <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Country) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 
typhi_amr_class_country %>% filter(Country == 'Bangladesh') %>% kbl() %>% kable_styling()

# function to run chi sq test on a given antibiotic class, assuming country and class.
run_chisq_on_amr <- function(df, antibiotic_class){
  test <- df %>% 
    ungroup() %>% 
    filter(Class == antibiotic_class) %>% 
    mutate(n_samples_without_gene = n_samples - n_samples_with_gene) %>% 
    select(n_samples_with_gene, n_samples_without_gene)
  test <- as.data.frame(test)

  # row.names(test) <- test$Country
  # test <- test[, -which(names(test) == "Country")]
  chisq.test(test)
}

# 6. run chi-sq tests on each antibiotic class
run_chisq_on_amr(typhi_amr_class_country, 'SULFONAMIDE')
run_chisq_on_amr(typhi_amr_class_country, 'TETRACYCLINE')
run_chisq_on_amr(typhi_amr_class_country, 'MACROLIDE')
run_chisq_on_amr(typhi_amr_class_country, 'QUINOLONE')
run_chisq_on_amr(typhi_amr_class_country, 'TRIMETHOPRIM')
run_chisq_on_amr(typhi_amr_class_country, 'BETA-LACTAM')
run_chisq_on_amr(typhi_amr_class_country, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN')

# 7. Group by class and group and summarise

typhi_amr_class_group <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Group) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 

typhi_amr_class_group %>% kbl() %>% kable_styling()

# 8. run chi sq 

run_chisq_on_amr(typhi_amr_class_group, 'SULFONAMIDE')
run_chisq_on_amr(typhi_amr_class_group, 'TETRACYCLINE')
run_chisq_on_amr(typhi_amr_class_group, 'MACROLIDE')
run_chisq_on_amr(typhi_amr_class_group, 'QUINOLONE')
run_chisq_on_amr(typhi_amr_class_group, 'TRIMETHOPRIM')
run_chisq_on_amr(typhi_amr_class_group, 'BETA-LACTAM')
run_chisq_on_amr(typhi_amr_class_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN')

# 7. Group by class, country, group, and summarise
typhi_amr_class_country_group <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Country, Group) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 
  # kbl() %>% kable_styling()

# function to run chi sq test on a given antibiotic class, assuming country, class and group.
run_chisq_on_amr_w_group <- function(df, antibiotic_class, country){
  test <- df %>% 
    ungroup() %>% 
    filter(Class == antibiotic_class, Country == country) %>% 
    select(Country, Group, n_samples_with_gene, n_samples) %>% 
    mutate(n_samples_without_gene = n_samples - n_samples_with_gene) %>% 
    select(-n_samples, Country)
  test <- as.data.frame(test)
  row.names(test) <- test$Group
  test <- test[, -which(names(test) == "Group")]
  test <- test[, -which(names(test) == "Country")]
  View(test)
  fisher.test(test)
}

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Nepal')

run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Bangladesh')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Malawi')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Nepal')

run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Bangladesh')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Malawi')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Nepal')

typhi_amr_class_country_group %>% kbl() %>% kable_styling()

amr_data_typhi_meta <- left_join(amr_data_typhi, metadata, by = join_by(sample == Lane))

# amr_data_typhi_meta %>% group_by(`Gene family`, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% arrange(`Gene family`, Country, Group) %>% kbl() %>% kable_styling()
# get all the unique class names
typhi_classes <- unique(amr_data_typhi_meta$Class)
# make an empty data frame, with just the classes, countries, and groups.
typhi_amr_classes <- expand.grid(Class = typhi_classes, Country = c('Bangladesh', 'Malawi', 'Nepal'), Group = c('Control', 'Acute typhoid', 'Carrier'))
# group by class, country, group, and summarise
tmp <- amr_data_typhi_meta %>% group_by(Class, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% kbl() %>% kable_styling()
# join the data to the template (typhi_amr_classes), set any NAs as 0.
typhi_amr_classes <- left_join(typhi_amr_classes, tmp, by = join_by(Class == Class, Country == Country, Group == Group)) %>% 
  mutate(n = ifelse(is.na(n), 0, n)) %>% 
  mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% 
  arrange(Class, Country, Group)
# print the table
typhi_amr_classes %>% kbl() %>% kable_styling()

# same as above, but for gene families
typhi_gene_family_names <- unique(amr_data_typhi_meta$`Gene family`)
typhi_amr_gene_families <- expand.grid(`Gene family` = typhi_gene_family_names, Country = c('Bangladesh', 'Malawi', 'Nepal'), Group = c('Control', 'Acute typhoid', 'Carrier'))

tmp <- amr_data_typhi_meta %>% group_by(`Gene family`, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) # %>% kbl() %>% kable_styling()
typhi_amr_families <- left_join(typhi_amr_gene_families, tmp, by = join_by(`Gene family` == `Gene family`, Country == Country, Group == Group)) %>% 
  mutate(n = ifelse(is.na(n), 0, n)) %>% 
  mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% 
  arrange(`Gene family`, Country, Group)

typhi_amr_families %>% kbl() %>% kable_styling()
```


## correlation 

```{r}
data <- read_tsv('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2024.01.15/2024.01.15.country_drug_percent.tsv')

# Data inspection and cleaning (if needed)
# data <- na.omit(data)  # Example of removing rows with NA

# data <- data %>% filter(country != 'Nepal')

microbiome_data <- data %>% filter(type == "Microbiome") %>% 
                  select(-type) %>% 
                  rename(percent_microbiome = percent)
isolate_data <- data %>% filter(type == "Isolate") %>% 
                select(-type) %>% 
                rename(percent_isolate = percent)

combined_data <- left_join(microbiome_data, isolate_data, by = c("country", "drug"))

# Plotting
combined_data$percent_microbiome <- combined_data$percent_microbiome / 100
combined_data$percent_isolate <- combined_data$percent_isolate / 100

```

```{r}

calculate_correlation <- function(data, drug_name) {
  cor_test <- cor.test(data$percent_isolate, data$percent_microbiome, use = "complete.obs")
#   mean_x <- mean(data$percent_isolate, na.rm = TRUE)
#   mean_y <- mean(data$percent_microbiome, na.rm = TRUE)
  return(data.frame(drug = drug_name, correlation = cor_test$estimate, p_value = cor_test$p.value,
                    label_pos_x = NA, label_pos_y = NA))
}

# Calculate Pearson's correlation coefficient, p-value, and label positions for each drug
correlation_data <- combined_data %>%
  group_by(drug) %>%
  do(calculate_correlation(., unique(.$drug)))
correlation_data$label_pos_x <- c(0.3, 0.3, 0.3)
correlation_data$label_pos_y <- c(0.9, 0.8, 0.7)
# Add label text
correlation_data$label <- with(correlation_data, paste(drug, ": r =", round(correlation, 2), "p =", signif(p_value, 2)))

# Plotting
plot <- ggplot(combined_data, aes(x = percent_isolate, y = percent_microbiome, color = drug)) +
  geom_point(aes(shape = country)) +
  geom_smooth(method = "lm", se = TRUE, aes(group = drug), linetype = "dotted", alpha = 0.8) +
  geom_text(data = correlation_data, aes(x = label_pos_x, y = label_pos_y, label = label), hjust = 1, vjust = 1, inherit.aes = FALSE) +
  labs(x = "Percent Isolate", y = "Percent Microbiome", color = "Drug", shape = "Country") +
  theme_minimal() +
  ylim(0, 1) +
  ggtitle("Correlation between Microbiome and Isolate Percentages with Linear Fit and Correlation Coefficients")


show(plot)

```


```{r}
calculate_correlation_and_lm <- function(data, drug_name) {
  cor_test <- cor.test(data$percent_isolate, data$percent_microbiome, use = "complete.obs")
  lm_model <- lm(percent_microbiome ~ percent_isolate, data = data)
  eq_slope <- coef(lm_model)[2]
  eq_intercept <- coef(lm_model)[1]
  equation <- paste("y = ", round(eq_slope, 2), "x + ", round(eq_intercept, 2))
  mean_x <- mean(data$percent_isolate, na.rm = TRUE)
  mean_y <- mean(data$percent_microbiome, na.rm = TRUE)
  return(data.frame(drug = drug_name, correlation = cor_test$estimate, p_value = cor_test$p.value,
                    equation = equation, label_pos_x = NA, label_pos_y = NA))
}

# Calculate Pearson's correlation coefficient, p-value, equation, and label positions for each drug
correlation_data <- combined_data %>%
  group_by(drug) %>%
  do(calculate_correlation_and_lm(., unique(.$drug)))

correlation_data$label_pos_x <- c(0.3, 0.3, 0.3)
correlation_data$label_pos_y <- c(0.95, 0.8, 0.65)

# Add label text
correlation_data$label <- with(correlation_data, paste(drug, ": r =", round(correlation, 2), "p =", signif(p_value, 2), "\n", equation))

correlation_data %>% kbl() %>% kable_styling()

# Plotting
plot <- ggplot(combined_data, aes(x = percent_isolate, y = percent_microbiome, color = drug)) +
  geom_point(aes(shape = country)) +
  geom_smooth(method = "lm", aes(group = drug), linetype = "dotted", alpha = 0.8) +
#   geom_text(data = correlation_data, aes(x = label_pos_x, y = label_pos_y, label = label), hjust = 1, vjust = 1, inherit.aes = FALSE) +
  labs(x = "Proportion Isolates\nwith Resistance Gene", y = "Proportion Microbiome\nwith Resistance Gene", color = "Drug", shape = "Country") +
  theme(text = element_text(size = 15)) +
  ylim(0, 1) +
  guides(fill = guide_legend(override.aes = list(fill = "transparent"))) +
  coord_fixed()

# Save and display the plot
# ggsave("microbiome_isolate_linear_fit_plot.png", plot)
print(plot)

```
