
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```

## Sources
```{r}
library(kableExtra)
library(dplyr)
library(stringr)
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```

## paths

```{r}
strataa_disease_health_bigmap_data_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.01/analyse_output/tsv-results/all_RPKMs.transpose.csv"
strataa_carrier_health_bigmap_data_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.12/analyse_output_healthy_carrier/tsv-results/all_RPKMs.transpose.csv"

#strataa_bigmap_metadata_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.01/2023.05.01.big_map_input.maaslin.csv"
```

## Setting general params

```{r}
filename_regex = "\\d+_\\d+_\\d+" #for patch and strata
```

## read in metadata

```{r}
metadata <- read_metadata(metadata_handle)

```



## Run beta diversity analysis

* Beta diversity analysis is done on relative abundances (see call to sweep function on )
```{r}

taxonomic_level <- "species"
all_three_countries <- c('Malawi', 'Bangladesh', 'Nepal')
just_malawi <- c('Malawi')
just_bangladesh <- c('Bangladesh')
just_nepal <- c('Nepal')
malawi_and_bangladesh <- c('Malawi', 'Bangladesh')

run_calculate_beta <- function(output_folder, tax_level, countries, meta, fn_regex, braken_folder){
  filtered_otu.unnormalised <- run_make_filtered_otu_table(output_folder, tax_level, countries, fn_regex, braken_folder, metadata)
  meta_filtered <- meta %>% filter(Country %in% countries)
  beta_dir <- file.path(output_folder, '4_beta')
  if (!dir.exists(beta_dir)){ dir.create(beta_dir) }
  calculate_beta(filtered_otu.unnormalised, meta_filtered, beta_dir)
}

run_calculate_beta(output_folder_all_three, taxonomic_level, all_three_countries, metadata, filename_regex, braken_folder)
run_calculate_beta(output_folder_blantyre, taxonomic_level, just_malawi, metadata, filename_regex, braken_folder)
run_calculate_beta(output_folder_dhaka, taxonomic_level, just_bangladesh, metadata, filename_regex, braken_folder)
run_calculate_beta(output_folder_kathmandu, taxonomic_level, just_nepal, metadata, filename_regex, braken_folder)
run_calculate_beta(output_folder_blantyre_dhaka, taxonomic_level, malawi_and_bangladesh, metadata, filename_regex, braken_folder)

```


## Run PERMANOVA

We do this separately from the calculate_beta This takes some time (~2-10 minutes per dataset). R2 is essentially effect size, p-value is p-value (calc'd by permutation).

https://groups.google.com/g/qiime-forum/c/qyDJAdECBnc/m/F44lWRsyKWIJ
https://forum.qiime2.org/t/group-significance-changes-and-interpretation/2396


```{r cache=TRUE}
run_permanova <- function(output_folder, inc_country){
  d.bray <- readRDS(file.path(output_folder, '4_beta', 'd.bray.RDS'))
  meta_permanova <- readRDS(file.path(output_folder, '4_beta', 'meta_for_permanova.RDS'))
  if (isTRUE(inc_country)) {
    pn <- adonis(d.bray~Country*Sex*Group*age_bracket*Antibiotics_taken_before_sampling_yes_no_assumptions, data = meta_permanova, permutations = 1000)
  } else {
    pn <- adonis(d.bray~Sex*Group*age_bracket*Antibiotics_taken_before_sampling_yes_no_assumptions, data = meta_permanova, permutations = 1000)
  }
  
  pn_with_var_names <- cbind(rownames(pn$aov.tab), data.frame(pn$aov.tab, row.names = NULL))
  pn_res <- pn_with_var_names %>% rename(variable = `rownames(pn$aov.tab)`) %>% mutate(is_it_significant = ifelse(`Pr..F.` < 0.01, 'significant', 'not_significant')) %>% arrange(desc(is_it_significant), desc(R2)) %>% select(!c(Df, SumsOfSqs, MeanSqs, F.Model))
  write_delim(pn_res, file.path(output_folder, '4_beta', 'permanova.results.tsv'), delim = '\t')
}

run_permanova(output_folder_all_three, inc_country = TRUE)
run_permanova(output_folder_blantyre, inc_country = FALSE)
run_permanova(output_folder_dhaka, inc_country = FALSE)
run_permanova(output_folder_kathmandu, inc_country = FALSE)
run_permanova(output_folder_blantyre_dhaka, inc_country = TRUE)

```

## Run alpha diversity

https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/alpha-diversities.html

* Not going to normalise before alpha diversity - https://www.nature.com/articles/s41579-018-0029-9
"Measures of species richness (for example, the number of observed species or the Chao1 abundance estimator, which estimates true species diversity) and phylogenetic measures (Faithâ€™s phylogenetic diversity) are sensitive to the number of sequences per sample, whereas measures that combine richness and evenness (Shannon index) are much less so. However, it should be noted that these methods have been evaluated exclusively for 16S rRNA data and may not apply to other microbiome data types."
* Slightly unsatisfying that run_calculate_alpha has a different structure to run_calc_beta/run_permanova, with the stats bit done separately for beta, but integrated in alpha, but whaddya gonna do?

```{r}

taxonomic_level <- "species"
all_three_countries <- c('Malawi', 'Bangladesh', 'Nepal')
just_malawi <- c('Malawi')
just_bangladesh <- c('Bangladesh')
just_nepal <- c('Nepal')
malawi_and_bangladesh <- c('Malawi', 'Bangladesh')

run_calculate_alpha <- function(output_folder, tax_level, countries, fn_regex, braken_folder, meta, inc_country){
  # if run_make_filtered_otu_table has already been run i.e. for the beta diversity analysi
  # then here, will just read in the otu table and return it, rather than gathering all the per sample files.
  filtered_otu.unnormalised <- run_make_filtered_otu_table(output_folder, tax_level, countries, fn_regex, braken_folder, meta)
  meta_subset <- meta %>% filter(Country %in% countries)
  anova_res <- calculate_alpha(filtered_otu.unnormalised, meta_subset, quote("Group"), output_folder, "Phenotype", inc_country)
  # anova_res[[1]] is the output dataframe
  anova_res_with_var_names <- cbind(rownames(anova_res[[1]]), data.frame(anova_res[[1]], row.names = NULL))
  anova_res_with_var_names <- anova_res_with_var_names %>% mutate(is_it_significant = ifelse(`Pr..F.` < 0.01, 'significant', 'not_significant')) %>% arrange(desc(is_it_significant), `Pr..F.`)
  write_delim(anova_res_with_var_names, file.path(output_folder, '3_alpha', 'shannon.anova.results.tsv'), delim = '\t')
  #View(anova_res_with_var_names)
  anova_res_with_var_names %>% kbl(digits = 3)  %>% kable_styling()
}

# all_three_countries won't be needed if the beta diversity has already been calculated, but should include just in case.
run_calculate_alpha(output_folder_all_three, taxonomic_level, all_three_countries, filename_regex, braken_folder, metadata, inc_country = TRUE)
run_calculate_alpha(output_folder_blantyre, taxonomic_level, just_malawi, filename_regex, braken_folder, metadata, inc_country = FALSE)
run_calculate_alpha(output_folder_dhaka, taxonomic_level, just_bangladesh, filename_regex, braken_folder, metadata, inc_country = FALSE)
run_calculate_alpha(output_folder_kathmandu, taxonomic_level, just_nepal, filename_regex, braken_folder, metadata, inc_country = FALSE)


```


## Run EdegR DGE

This takes some time 2-10 minutes per run.

```{r}

taxonomic_level <- "species"
all_three_countries <- c('Malawi', 'Bangladesh', 'Nepal')
just_malawi <- c('Malawi')
just_bangladesh <- c('Bangladesh')
just_nepal <- c('Nepal')
malawi_and_bangladesh <- c('Malawi', 'Bangladesh')

run_calculate_dge <- function(meta, output_folder, tax_level, countries_to_include, covariates, groups_for_comparison){
  meta_subset <- meta %>% filter(Country %in% countries_to_include)
  filtered_otu.unnormalised <- run_make_filtered_otu_table(output_folder, tax_level, countries_to_include, fn_regex, braken_folder, meta)
  calculate_dge(meta_subset, output_folder, tax_level, filtered_otu.unnormalised, covariates, groups_for_comparison)
}

groups_for_comparison <- c("Control_HealthySerosurvey", "Acute_Typhi")
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions', 'sequencing_lane')
run_calculate_dge(metadata, output_folder_all_three, taxonomic_level, all_three_countries, covars, groups_for_comparison)
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions', 'sequencing_lane')
run_calculate_dge(metadata, output_folder_blantyre, taxonomic_level, just_malawi, covars, groups_for_comparison)
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions')
run_calculate_dge(metadata, output_folder_dhaka, taxonomic_level, just_bangladesh, covars, groups_for_comparison)
run_calculate_dge(metadata, output_folder_kathmandu, taxonomic_level, just_nepal, covars, groups_for_comparison)
run_calculate_dge(metadata, output_folder_blantyre_dhaka, taxonomic_level, malawi_and_bangladesh, covars, groups_for_comparison)

groups_for_comparison <- c("Control_HealthySerosurvey", "Carrier")
#covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions', 'sequencing_lane')
#run_calculate_dge(metadata, output_folder_all_three, taxonomic_level, all_three_countries, covars, groups_for_comparison)
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions', 'sequencing_lane')
run_calculate_dge(metadata, output_folder_blantyre, taxonomic_level, just_malawi, covars, groups_for_comparison)
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions')
run_calculate_dge(metadata, output_folder_dhaka, taxonomic_level, just_bangladesh, covars, groups_for_comparison)
#run_calculate_dge(metadata, output_folder_kathmandu, taxonomic_level, just_nepal, covars, groups_for_comparison)
#run_calculate_dge(metadata, output_folder_blantyre_dhaka, taxonomic_level, malawi_and_bangladesh, covars, groups_for_comparison)
```



## Run MAASLIN2 DGE

```{r  include=FALSE}

strataa_metaphlan_data = read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.all_strataa_metaphlan.csv'), header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_metaphlan_metadata = read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.strataa_metadata.metaphlan.csv'), header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)


#variables_for_analysis <- c("Group")
groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")

run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"))

# not including antibiotics becayse they're all "No"

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')
#bgd_acute_healthy_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste())
variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA)

#variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG',  c("sequencing_lane,32580_7"))

```
## run maaslin2 functional

```{r echo = T, results = 'hide'}

strataa_bigmap_data = read.csv(strataa_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_disease_health_bigmap_data = read.csv(file = strataa_disease_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_carrier_health_bigmap_data = read.csv(file = strataa_carrier_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)

rownames(strataa_carrier_health_bigmap_data) <- rownames(strataa_carrier_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')
rownames(strataa_disease_health_bigmap_data) <- rownames(strataa_disease_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')


# strataa_bigmap_metadata is the same as strataa_metaphlan_metadata
#strataa_bigmap_metadata = read.csv(file = strataa_bigmap_metadata_path, header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

#strataa_bigmap_metadata <- strataa_bigmap_metadata %>% filter(Gender != '#N/A')
#strataa_bigmap_metadata$Age <- as.numeric(as.character(strataa_bigmap_metadata$Age))

strataa_bigmap_metadata <- metadata %>% mutate(isolate_bigmap = str_replace_all(isolate, '_', '.'))
rownames(strataa_bigmap_metadata) <- strataa_bigmap_metadata$isolate_bigmap


```

typhi vs health

```{r}

groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_disease_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA)
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", 'Lane')
run_maaslin(strataa_carrier_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Lane,32580_7"))

```

carrier vs health

```{r}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age")
# variables_for_analysis <- c("Group")
run_maaslin(strataa_carrier_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA)
variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_carrier_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Lane,32712_2"))

```

