---
title: "STRATAA Microbiome"
author: "Philip Ashton"
date: "`r Sys.Date()`"
output: 
  html_document:
      toc: true
      toc_float: true
      theme: united
      number_sections: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)

```

# STRATAA microbiome analysis

## Imports
```{r}
library(kableExtra)
library(ggplot2)
library(ggpubr)
library(patchwork)
library(vegan)
```


## Sources

The file handles are set in config.R as they're used by both this script and data_cleaning.

```{r}
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```


## Metadata basics

First do the plots with kruskall wallis comparison between groups to see if there's an overall difference, then do the pairwise tests because KW says there is a difference.


```{r}
metadata <- read_metadata(metadata_handle)

metadata <- metadata %>% mutate(Group = if_else(Group == 'Control_HealthySerosurvey', 'Healthy Control', Group))
metadata <- metadata %>% mutate(Group = if_else(Group == 'Acute_Typhi', 'Acute typhoid', Group))
number_per_country <- metadata %>% group_by(Country) %>% summarise(count = n())

number_per_country %>% kbl() %>% kable_styling()

print(sum(number_per_country$count))

metadata %>% group_by(Group) %>% summarise(count = n()) %>% kbl() %>% kable_styling()

baseline_chars <- get_baseline_characteristics(metadata)

baseline_chars$pct_anti
baseline_chars$pct_anti %>% na_if(0)

# BEWARE - na_if(0.0) will convert ALL 0s to NAs, this is ok as they're only in the antibiotics of carriers and controls
# at the moment, but if others get added in, will screw with it.
baseline_chars %>% rename(`Median age` = median_age, `Women (%)` = pct_fem, `Antibiotics in last 2 weeks (%)` = pct_anti) %>% pivot_longer(!c(Country, Group)) %>% rename(variable_name = name) %>% pivot_wider(names_from = Group, values_from = value) %>% filter(variable_name != 'number') %>% na_if(0.0) %>% kbl(digits = c(NA, NA, 1, 1, 1)) %>% kable_styling()


ggplot(metadata, aes(x = Country, y = Age, fill = Group)) + geom_boxplot() + stat_compare_means(method = 'kruskal.test', label = "p")
comparisons_groups <- list(c("Acute typhoid", "Carrier"), c("Acute typhoid", "Healthy Control"), c("Carrier", "Healthy Control"))
comparisons_countries <- list(c('Bangladesh', 'Malawi'), c('Bangladesh', 'Nepal'), c('Malawi', 'Nepal'))
# default stat method when doing pairwise tests in wilcoxon.
ggboxplot(metadata, facet.by = "Country", y = "Age", x = "Group", color = "Group") + stat_compare_means(comparisons = comparisons_groups) + rremove("x.text") + rremove("xlab") + rremove("x.ticks") # +rotate_x_text(angle = 45)
ggboxplot(metadata, facet.by = "Group", y = "Age", x = "Country", color = "Country") + stat_compare_means(comparisons = comparisons_countries) + rotate_x_text(angle = 45)



```

```{r fig.width = 4, fig.height= 8}
country_group_sex <- metadata %>% group_by(Country, Group, Sex) %>% summarise(count = n()) 

plot_sex <- function(eg1, c){
  d <- eg1 %>% filter(Country == c)
  p <- ggplot(d, aes(x = Group, y = count, fill = Sex)) + 
    geom_bar(stat ='identity', position = 'fill') + 
    ylab('Proportion') + 
    ggtitle(c)# +
    #theme(axis.text=element_text(size=34), axis.title=element_text(size=36,face="bold"), plot.title = element_text(size = 40, face = "bold"), legend.key.size = unit(4, 'cm'), legend.title = element_text(size = 34), legend.text = element_text(size = 28))
  
  return(p)
}

m_sex <- plot_sex(country_group_sex, 'Malawi')
b_sex <- plot_sex(country_group_sex, 'Bangladesh')
n_sex <- plot_sex(country_group_sex, 'Nepal')
m_sex / b_sex / n_sex 
```



## Alpha diversity

```{r}
meta_for_alpha <- metadata %>% select(isolate, Group, Sex, Country, Antibiotics_taken_before_sampling_yes_no_assumptions, age_bracket)

```


### all three

ANOVA results

```{r}
anova_all_three_csgaa_res <- read_delim(file.path(output_folder_all_three, '3_alpha', 'shannon.anova.results.tsv'), delim = '\t')
anova_all_three_csgaa_res %>% kbl %>% kable_styling()
```

Plots

```{r}
all_three_shannon_alpha <- read_delim(file.path(output_folder_all_three, '3_alpha', 'shannon.alpha_results.tsv'))
all_three_shannon_alpha <- left_join(all_three_shannon_alpha, meta_for_alpha, by = 'isolate', )
all_three_shannon_alpha <- all_three_shannon_alpha %>% mutate(Group = if_else(Group == 'Control_HealthySerosurvey', 'Healthy Control', Group))
all_three_shannon_alpha <- all_three_shannon_alpha %>% mutate(Group = if_else(Group == 'Acute_Typhi', 'Acute typhoid', Group)) %>% arrange(Country)


country_comps <- list(c('Bangladesh', 'Malawi'), c('Malawi', 'Nepal'), c('Bangladesh', 'Nepal'))
ggplot(all_three_shannon_alpha, aes(x=Country, y=alpha)) + 
  ggtitle("By country") + 
  labs(x="", y="Alpha diversity") +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12), axis.text.y = element_text(size = 12), axis.title = element_text(size = 14, face = 'bold')) +
  stat_compare_means(size = 4, label = "p.format", comparisons = country_comps)

group_comps <- list(c(1, 2), c(2, 3), c(1, 3))
ggplot(all_three_shannon_alpha, aes(x=Group, y=alpha)) + 
  ggtitle("By group") + 
  labs(x="", y="Alpha diversity") +
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 12), axis.text.y = element_text(size = 12), axis.title = element_text(size = 14, face = 'bold')) +
  stat_compare_means(size = 4, label = "p.format", comparisons = group_comps)
  #geom_point(alpha = 0.3, position = "jitter")

country_group_comps <- list(c('Acute typhoid', 'Carrier'), c('Acute typhoid', 'Healthy Control'), c('Healthy Control', 'Carrier'))

ggboxplot(all_three_shannon_alpha, facet.by = "Country", y = "alpha", x = "Group", color = "Group") + stat_compare_means(comparisons = country_group_comps, label = 'p.signif') + rremove("x.text") + rremove("xlab") + rremove("x.ticks") # +rotate_x_text(angle = 45)


ggboxplot(all_three_shannon_alpha, facet.by = "Group", y = "alpha", x = "Country", color = "Country") + stat_compare_means(comparisons = country_comps) + rremove("x.text") + rremove("xlab") + rremove("x.ticks") # +rotate_x_text(angle = 45)


antibiotic_comps <- list(c('Yes', 'No'))
ggboxplot(all_three_shannon_alpha, facet.by = "Country", y = "alpha", x = "Antibiotics_taken_before_sampling_yes_no_assumptions", color = "Antibiotics_taken_before_sampling_yes_no_assumptions", legend.title = 'Antibiotics') + stat_compare_means(comparisons = antibiotic_comps) + rremove("x.text") + rremove("xlab") + rremove("x.ticks")


```

### blantyre

ANOVA

```{r}

anova_blantyre_sgaa_res <- read_delim(file.path(output_folder_blantyre, '3_alpha', 'shannon.anova.results.tsv'), delim = '\t')
anova_blantyre_sgaa_res %>% kbl %>% kable_styling()

```

Plots

- The p-value on the graph is a kruskal wallis test (which is just wilcoxon i think?)

```{r}
blantyre_shannon_alpha <- read_delim(file.path(output_folder_blantyre, '3_alpha', 'shannon.alpha_results.tsv'))
blantyre_shannon_alpha <- left_join(blantyre_shannon_alpha, meta_for_alpha, by = 'isolate')

my_comparisons <- list(c('No', 'Yes'), c('No', 'Yes'), c('No', 'Yes'))

ggplot(blantyre_shannon_alpha, aes(x=age_bracket, y=alpha, fill = Antibiotics_taken_before_sampling_yes_no_assumptions)) + 
  #ggtitle("All three countries - country by group") + 
  labs(x="Age bracket", y="Alpha diversity", fill='Antibiotics', color = 'Antibiotics') +
  geom_boxplot(outlier.shape = NA, alpha = 0.2) +
  #geom_point(aes(color=as.factor(Antibiotics_taken_before_sampling_yes_no_assumptions)), alpha = 0.3) +
  geom_point(alpha = 0.5, position=position_dodge(width=0.75),aes(group=Antibiotics_taken_before_sampling_yes_no_assumptions, color=as.factor(Antibiotics_taken_before_sampling_yes_no_assumptions))) +
  stat_compare_means(label.y = c(5.2, 5.5, 5.7), size = 4, label = "p.format")
  

```


### dhaka

```{r}

anova_dhaka_sgaa_res <- read_delim(file.path(output_folder_dhaka, '3_alpha', 'shannon.anova.results.tsv'), delim = '\t')
anova_dhaka_sgaa_res %>% kbl %>% kable_styling()

```

### kathmandu

```{r}

anova_kathmandu_sgaa_res <- read_delim(file.path(output_folder_kathmandu, '3_alpha', 'shannon.anova.results.tsv'), delim = '\t')
anova_kathmandu_sgaa_res %>% kbl %>% kable_styling()

```


## Beta diversity

Information on interpretation of PERMANOVA R2 from here - https://forum.qiime2.org/t/group-significance-changes-and-interpretation/2396 , some info on the p-values here https://forum.qiime2.org/t/understanding-beta-group-significance-permanova-results/12648

```{r}
run_plot_beta <- function(output_folder, meta, vars_to_plot){
  pcoa.data.handle <- file.path(output_folder, '4_beta', "first_2d_coords.txt")
  pcoa.var.handle <- file.path(output_folder, '4_beta', "pcoa_var.txt")
  pcoa.data <- read_delim(pcoa.data.handle, delim = " ", escape_double = FALSE, trim_ws = TRUE)
  #pcoa.data <- pcoa.data %>% separate(col = Y, sep = '\t', into = c('x', 'y')) %>% rename(ID = X)
  pcoa.var <- read_delim(pcoa.var.handle, delim = " ", escape_double = FALSE, trim_ws = TRUE)
  #pcoa.data <- read_tsv(pcoa.data.handle)
  
  meta_for_beta <- meta %>% select(isolate, Country, Group, Sex, age_bracket, Antibiotics_taken_before_sampling_yes_no_assumptions, group_country, group_antibiotic, sequencing_lane)
  
  pcoa.data <- left_join(pcoa.data, meta_for_beta, by = c('Sample' = 'isolate'))

  # add some different plots to the below
  # or, change this function to take the variable to colour by
  plots <- plot_beta(pcoa.data, pcoa.var, vars_to_plot)
  return(plots)
}
```

### All three countries

#### Plots

```{r}
all_three_vars_to_plot <- c('Country', 'Group', 'Sex', 'age_bracket', 'group_country', 'group_antibiotic')
all_three_plots <- run_plot_beta(output_folder_all_three, metadata, all_three_vars_to_plot)

a3_c <- all_three_plots$country_plot
a3_c
a3_g <- all_three_plots$group_plot
a3_g
all_three_plots$sex_plot
all_three_plots$age_plot
all_three_plots$group_country_plot
all_three_plots$group_antibiotic_plot


```

```{r fig.height = 3, fig.width = 3}
a3_c / a3_g
```


#### PERMANOVA

```{r}
pn_all3_csgaa_res <- read_delim(file.path(output_folder_all_three, '4_beta', 'permanova.results.tsv'))
pn_all3_csgaa_res %>% kbl %>% kable_styling()
```


### Just Blantyre

#### Plots

```{r}
blantyre_vars_to_plot <- c('Group', 'Sex', 'age_bracket', 'group_antibiotic', 'sequencing_lane')
blantyre_plots <- run_plot_beta(output_folder_blantyre, metadata, blantyre_vars_to_plot)
b_g <- blantyre_plots$group_plot
b_g
blantyre_plots$sex_plot
blantyre_plots$age_plot
blantyre_plots$group_antibiotic_plot
blantyre_plots$sequencing_lane
```

#### PERMANOVA

```{r}
pn_blantyre_sgaa_res <- read_delim(file.path(output_folder_blantyre, '4_beta', 'permanova.results.tsv'))
pn_blantyre_sgaa_res %>% kbl %>% kable_styling()

```

### Just Dhaka

#### Plots

```{r}
dhaka_vars_to_plot <- c('Group', 'Sex', 'age_bracket', 'group_antibiotic', 'sequencing_lane')
dhaka_plots <- run_plot_beta(output_folder_dhaka, metadata, dhaka_vars_to_plot)
d_g <- dhaka_plots$group_plot
d_g
dhaka_plots$sex_plot
dhaka_plots$age_plot
dhaka_plots$group_antibiotic_plot
dhaka_plots$sequencing_lane
```

#### PERMANOVA

```{r}
pn_dhaka_sgaa_res <- read_delim(file.path(output_folder_dhaka, '4_beta', 'permanova.results.tsv'))
pn_dhaka_sgaa_res %>% kbl %>% kable_styling()
```


### Just Kathmandu

#### Plots

```{r}
kathmandu_vars_to_plot <- c('Group', 'Sex', 'age_bracket', 'group_antibiotic', 'sequencing_lane')
kathmandu_plots <- run_plot_beta(output_folder_kathmandu, metadata, all_three_vars_to_plot)
k_g <- kathmandu_plots$group_plot
k_g
kathmandu_plots$sex_plot
kathmandu_plots$age_plot
kathmandu_plots$group_antibiotic_plot
```

#### PERMANOVA

```{r}
pn_kathmandu_sgaa_res <- read_delim(file.path(output_folder_kathmandu, '4_beta', 'permanova.results.tsv'))
pn_kathmandu_sgaa_res %>% kbl %>% kable_styling()
```

### Blantyre and Dhaka


#### Plots

```{r}
blantyre_dhaka_vars_to_plot <- c('Group', 'Sex', 'age_bracket', 'group_antibiotic')
blantyre_dhaka_plots <- run_plot_beta(output_folder_blantyre_dhaka, metadata, all_three_vars_to_plot)
blantyre_dhaka_plots$group_plot
blantyre_dhaka_plots$sex_plot
blantyre_dhaka_plots$age_plot
blantyre_dhaka_plots$group_antibiotic_plot
```

#### PERMANOVA

```{r}
pn_blantyre_dhaka_csgaa_res <- read_delim(file.path(output_folder_blantyre_dhaka, '4_beta', 'permanova.results.tsv'))
pn_blantyre_dhaka_csgaa_res %>% kbl %>% kable_styling()
```

### Combined plots

It looks like some might be being trimmed by the coordinate limits in the plot_beta function, but they're not (i've checked both blantyre and dhaka).

```{r fig.height= 3, fig.width = 3}
#b_g / d_g | k_g / plot_spacer() + plot_layout(guides = 'collect')
b_g + d_g + k_g + guide_area() + plot_layout(ncol = 2, guides = 'collect')
```


## Taxa associated with participant groups

### EdgeR

```{r}
combined_output_root <- '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/phil/2022.11.08/combined_analysis'
if (!dir.exists(combined_output_root)){ dir.create(combined_output_root) }

groups_for_comparison <- c("Control_HealthySerosurvey", "Acute_Typhi")
comp <- paste(groups_for_comparison[1], 'vs', groups_for_comparison[2], sep = '_')
covars <- c('Country', 'Sex', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions')
covar_initials <- paste(str_sub(covars, 1, 6), sep = '', collapse = '')

#to_combine <- c(output_folder_all_three, output_folder_blantyre, output_folder_dhaka, output_folder_kathmandu, output_folder_blantyre_dhaka)

blantyre_output_path <- file.path(output_folder_blantyre, '5_glm', paste('results_all', comp, covar_initials, 'edgeR.tsv', sep = '.'))
dhaka_output_path <- file.path(output_folder_dhaka, '5_glm', paste('results_all', comp, covar_initials, 'edgeR.tsv', sep = '.'))
View(blantyre_output_path)
blantyre_output <- read_delim(blantyre_output_path, delim = "\t", escape_double = FALSE,  trim_ws = TRUE)
dhaka_output <- read_delim(dhaka_output_path, delim = "\t", escape_double = FALSE,  trim_ws = TRUE)


# this is the ridiculous shit you have to do in R to get a list of dataframes.
to_combine_paths <- c(blantyre_output_path, dhaka_output_path)
# this is a vec of strings that will be passed to dge_output to give sensible names to the joined output
# needs to be same order as to_combine_paths
to_combine_strings_for_join <- c("_mal", "_bang")
to_combine <- list()
for (i in seq_along(to_combine_paths)) {
  to_combine[[i]] <- read_delim(to_combine_paths[i], delim = "\t", escape_double = FALSE,  trim_ws = TRUE)
  to_combine[[i]] <- to_combine[[i]] %>% rename(species=...1)
}

# the_date doesn't have to be todays date, it's just the current working dir
the_date <- '2022.11.10'

dge_output <- combine_and_compare_edgeRs(to_combine, to_combine_strings_for_join)

#dge_output <- old_combine_and_compare_edgeRs(to_combine, covar_initials, output_folder_all_three, output_folder_blantyre, output_folder_dhaka, output_folder_kathmandu, combined_output_root, the_date, comp)
#options(scipen = 999)
#options(scipen = 0) # to switch scientific notation back on

View(dge_output$sig_up_cases_controls)

sig_up_cases_controls_for_writing <- dge_output$sig_up_cases_controls %>% select(species, malawi_bangladesh_logFC, malawi_bangladesh_FDR, bangladesh_logFC, bangladesh_FDR, malawi_logFC, malawi_FDR) %>% arrange(malawi_bangladesh_FDR, malawi_bangladesh_logFC)

sig_up_cases_controls_for_writing %>% kbl() %>% kable_styling()

write_csv(sig_up_cases_controls_for_writing, file.path(combined_output_root, 'dge', paste(the_date, comp, covar_initials, 'sig_up_dge.csv', sep = '.')))

sig_down_cases_controls_for_writing <- dge_output$sig_down_cases_controls %>% select(species, malawi_bangladesh_logFC, malawi_bangladesh_FDR, bangladesh_logFC, bangladesh_FDR, malawi_logFC, malawi_FDR) %>% arrange(malawi_bangladesh_FDR, malawi_bangladesh_logFC)

sig_down_cases_controls_for_writing %>% kbl() %>% kable_styling()

write_csv(sig_down_cases_controls_for_writing, file.path(combined_output_root, 'dge', paste(the_date, comp, covar_initials, 'sig_down_dge.csv', sep = '.')))
#covar_initials <- paste(str_sub(covars, 1, 6), sep = '', collapse = '')

#combined_dge_output_folder <- file.path(combined_output_root, paste(covar_initials, 'combined_edgeR'))

```

### maaslin

Combine the taxonomic maaslins, and print out the species that are sig in both and share directions.

Because sequencing run and participant type are totally confounded for Bangladesh, need to exclude sequencing run from the final model for Bangladesh (otherwise, wipes out the signals).

Things in common 

```{r}
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
#mwi_variables_for_analysis <- c("Group")
#bang_variables_for_analysis <- c("Group")
mwi_vars_for_dirname <- paste(mwi_variables_for_analysis, collapse = '.')
bang_vars_for_dirname <- paste(bang_variables_for_analysis, collapse = '.')
```


between acute <-> health.

```{r}

mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
#mwi_variables_for_analysis <- c("Group")
#bang_variables_for_analysis <- c("Group")
mwi_vars_for_dirname <- paste(mwi_variables_for_analysis, collapse = '.')
bang_vars_for_dirname <- paste(bang_variables_for_analysis, collapse = '.')

groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')
groups_for_dirname <- paste(groups_to_analyse, collapse = '.')

bang_maaslin_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste('Bangladesh', paste(groups_to_analyse, collapse = '_vs_'), bang_vars_for_dirname, sep = '_'))

malawi_maaslin_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste('Malawi', paste(groups_to_analyse, collapse = '_vs_'), mwi_vars_for_dirname, sep = '_'))

bang_maaslin <- read_delim(file.path(bang_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
malawi_maaslin <- read_delim(file.path(malawi_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)

combined_maaslins <- combine_maaslins(bang_maaslin, malawi_maaslin)

combined_maaslins_positive_coef <- combined_maaslins$positive_coef
combined_maaslins_negative_coef <- combined_maaslins$negative_coef


vars_for_output_dirname <- paste(mwi_vars_for_dirname, 'mwi_only', sep = '_')
#combined_maaslins_positive_coef %>%  kbl() %>% kable_styling()
write_csv(combined_maaslins_positive_coef, file.path(maaslin_taxonomic_output_root_folder, paste(groups_for_dirname, vars_for_output_dirname, 'combined_maaslins_positive_coef.csv', sep = '.')))
#combined_maaslins_negative_coef %>%  kbl() %>% kable_styling()
write_csv(combined_maaslins_negative_coef, file.path(maaslin_taxonomic_output_root_folder, paste(groups_for_dirname, vars_for_output_dirname, 'combined_maaslins_negative_coef.csv', sep = '.')))
```

Carrier <-> health

```{r}


groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')
groups_for_dirname <- paste(groups_to_analyse, collapse = '.')

mwi_variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
bang_variables_for_analysis <- c("Group", "Sex", "Age")
#mwi_variables_for_analysis <- c("Group")
#bang_variables_for_analysis <- c("Group")
mwi_vars_for_dirname <- paste(mwi_variables_for_analysis, collapse = '.')
bang_vars_for_dirname <- paste(bang_variables_for_analysis, collapse = '.')


bang_maaslin_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste('Bangladesh', paste(groups_to_analyse, collapse = '_vs_'), bang_vars_for_dirname, sep = '_'))

malawi_maaslin_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste('Malawi', paste(groups_to_analyse, collapse = '_vs_'), mwi_vars_for_dirname, sep = '_'))

bang_maaslin <- read_delim(file.path(bang_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
malawi_maaslin <- read_delim(file.path(malawi_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)

combined_maaslins <- combine_maaslins(bang_maaslin, malawi_maaslin)

combined_maaslins_positive_coef <- combined_maaslins$positive_coef
combined_maaslins_negative_coef <- combined_maaslins$negative_coef


vars_for_output_dirname <- paste(mwi_vars_for_dirname, 'mwi_only', sep = '_')
#vars_for_output_dirname <- mwi_vars_for_dirname
#combined_maaslins_positive_coef %>%  kbl() %>% kable_styling()
write_csv(combined_maaslins_positive_coef, file.path(maaslin_taxonomic_output_root_folder, paste(groups_for_dirname, vars_for_output_dirname, 'combined_maaslins_positive_coef.csv', sep = '.')))
#combined_maaslins_negative_coef %>%  kbl() %>% kable_styling()
write_csv(combined_maaslins_negative_coef, file.path(maaslin_taxonomic_output_root_folder, paste(groups_for_dirname, vars_for_output_dirname, 'combined_maaslins_negative_coef.csv', sep = '.')))
```


Print the results for all the P. copri clades

```{r}

all_taxa_maaslin <- combined_maaslins$all_features
p_copri_maaslin <- all_taxa_maaslin %>% filter(grepl('_Prevotella_copri_', feature)) %>% filter(qval_bang < 0.05 | qval_mal < 0.05)
write_csv(p_copri_maaslin, file.path(maaslin_taxonomic_output_root_folder, 'combined_maaslins_p_copri.csv'))

```

## functional groups associated with health



```{r}
variables_for_analysis <- c("Group", "Gender", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
vars_for_dirname <- paste(variables_for_analysis, collapse = '.')
groups_to_analyse <- c('Acute_Typhi', 'Control_Healthy')

bang_maaslin_output_dir <- file.path(maaslin_functional_output_root_folder, paste('Bangladesh', paste(groups_to_analyse, collapse = '_vs_'), vars_for_dirname, sep = '_'))

malawi_maaslin_output_dir <- file.path(maaslin_functional_output_root_folder, paste('Malawi', paste(groups_to_analyse, collapse = '_vs_'), vars_for_dirname, sep = '_'))

bang_maaslin <- read_delim(file.path(bang_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)
malawi_maaslin <- read_delim(file.path(malawi_maaslin_output_dir, "all_results.tsv"), delim = "\t", escape_double = FALSE, trim_ws = TRUE)

combined_maaslins <- combine_maaslins(bang_maaslin, malawi_maaslin)

combined_maaslins_positive_coef <- combined_maaslins$positive_coef
combined_maaslins_negative_coef <- combined_maaslins$negative_coef

# hacky as hell thing to split the feature name to extract the organism and the enzyme class.
# because separate isn't working with .. or \\.\\. which is the actual separator used.
combined_maaslins_positive_coef <- combined_maaslins_positive_coef %>% separate_wider_delim(feature, delim = 'Entryname.', names_sep = '', cols_remove = FALSE) %>% separate_wider_delim(feature2, delim = '..OS.', names_sep = '') %>% separate_wider_delim(feature22, delim = '..SMASH', names_sep = '') %>% select(!c(feature1, feature222)) %>% rename(MGC_class = feature21, Species = feature221, feature = featurefeature)

combined_maaslins_negative_coef <- combined_maaslins_negative_coef %>% separate_wider_delim(feature, delim = 'Entryname.', names_sep = '', cols_remove = FALSE) %>% separate_wider_delim(feature2, delim = '..OS.', names_sep = '') %>% separate_wider_delim(feature22, delim = '..SMASH', names_sep = '') %>% select(!c(feature1, feature222)) %>% rename(MGC_class = feature21, Species = feature221, feature = featurefeature)



combined_maaslins_positive_coef %>%  kbl() %>% kable_styling()

combined_maaslins_negative_coef %>%  kbl() %>% kable_styling()

write_csv(combined_maaslins_positive_coef, file.path(maaslin_functional_output_root_folder, 'combined_maaslins_positive_coef.csv'))
combined_maaslins_negative_coef %>%  kbl() %>% kable_styling()
write_csv(combined_maaslins_negative_coef, file.path(maaslin_functional_output_root_folder, 'combined_maaslins_negative_coef.csv'))

```



