
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```

## Sources
```{r imports}
library(kableExtra)
library(dplyr)
library(stringr)
library(tibble)
library(broom)
library(rstatix)
library(ggplot2)
library(ggpubr)
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```

## paths

```{r paths}
strataa_disease_health_bigmap_data_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.01/analyse_output/tsv-results/all_RPKMs.transpose.csv"
strataa_carrier_health_bigmap_data_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.12/analyse_output_healthy_carrier/tsv-results/all_RPKMs.transpose.csv"

amr_data_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/2023.08.30.strataa_microbiome.subsamp.rgi.txt"
#strataa_bigmap_metadata_path <- "~/Dropbox/GordonGroup/STRATAA_Microbiome/big-map/2023.05.01/2023.05.01.big_map_input.maaslin.csv"
ncbi_amr_info_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/ncbi_amr/2023.08.31/refgenes.modified.csv"
```

## Setting general params

```{r filename regex}
filename_regex = "\\d+_\\d+_\\d+" #for patch and strata
```

## read in metadata

```{r read_in_metadata}
metadata <- read_metadata(metadata_handle)
# View(metadata)

```


## Run MAASLIN2 taxonomic

```{r read_in_metaphlan, include=FALSE}

strataa_metaphlan_data = read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.all_strataa_metaphlan.csv'), header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_metaphlan_metadata = read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.strataa_metadata.metaphlan.csv'), header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

```

```{r run_maaslin_metaphlan_healthy_typhoid include=FALSE}

#variables_for_analysis <- c("Group")
groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")

run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"), 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
# variables_for_analysis <- c("Group", "Sex")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG',  NA, 'metaphlan')
```


For healthy vs carrier, not including antibiotics becayse they're all "No"



```{r run_maaslin_metaphlan_healthy_carrier}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')
#bgd_acute_healthy_output_dir <- file.path(maaslin_taxonomic_output_root_folder, paste())
variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'metaphlan')

#variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG',  c("sequencing_lane,32580_7"), 'metaphlan')

variables_for_analysis <- c("Group", "Sex", "Age")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'metaphlan')
```

health vs acute vs carrier

```{r}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey', 'Acute_Typhi')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey", "Lane,32712_2"), 'metaphlan')
# checking the effect of excluding sequencing lane
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey"), 'metaphlan')


variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_metaphlan_data, strataa_metaphlan_metadata, maaslin_taxonomic_output_root_folder, 'Nepal', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Group,Control_HealthySerosurvey"), 'metaphlan')

```



## run maaslin2 functional

```{r read_in_bigmap_results echo = T, results = 'hide'}

# strataa_bigmap_data = read.csv(strataa_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_disease_health_bigmap_data = read.csv(file = strataa_disease_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_carrier_health_bigmap_data = read.csv(file = strataa_carrier_health_bigmap_data_path, header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)

rownames(strataa_disease_health_bigmap_data) <- rownames(strataa_disease_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')
rownames(strataa_carrier_health_bigmap_data) <- rownames(strataa_carrier_health_bigmap_data) %>% str_replace_all('-', '.') %>% str_replace_all('#', '.')

# strataa_bigmap_metadata is the same as strataa_metaphlan_metadata
#strataa_bigmap_metadata = read.csv(file = strataa_bigmap_metadata_path, header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

#strataa_bigmap_metadata <- strataa_bigmap_metadata %>% filter(Gender != '#N/A')
#strataa_bigmap_metadata$Age <- as.numeric(as.character(strataa_bigmap_metadata$Age))

strataa_bigmap_metadata <- metadata %>% mutate(isolate_bigmap = str_replace_all(isolate, '_', '.'))
rownames(strataa_bigmap_metadata) <- strataa_bigmap_metadata$isolate_bigmap

```

typhi vs health

- todo = haven't changed the function for including the prevalence, probably just need to filter/select the meta-data or something?

```{r }

groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
run_maaslin(strataa_disease_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", 'sequencing_lane')
run_maaslin(strataa_disease_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("sequencing_lane,32580_7"), 'bigmap')

```

carrier vs health

```{r}

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')

variables_for_analysis <- c("Group", "Sex", "Age")
# variables_for_analysis <- c("Group")
run_maaslin(strataa_carrier_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Bangladesh', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', NA, 'bigmap')
variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
run_maaslin(strataa_carrier_health_bigmap_data, strataa_bigmap_metadata, maaslin_functional_output_root_folder, 'Malawi', groups_to_analyse, variables_for_analysis, 'NONE', 'LOG', c("Lane,32712_2"), 'bigmap')

```

##Â amr analysis

Need to setup a data frame with every sample/drug combination so that capture the zeroes. 

```{r read_in_amr, include=FALSE}

ncbi_amr_info <- read_csv(ncbi_amr_info_path)

amr_data <- read_delim(amr_data_path, delim = "\t") %>% separate_wider_delim(`ARO Term`, names = c("sample", "ARO Term"), delim = ".gene_mapping_data.txt:") %>% mutate(reads_per_kb = `Completely Mapped Reads` / (`Reference Length` / 1000)) 

# write_csv(amr_data_with_ncbi, file.path('~/Desktop', "amr_data_with_ncbi.csv"))
amr_data_with_ncbi <- left_join(amr_data, ncbi_amr_info, by = join_by(`ARO Term` == combined_name))

# 1. set the samples and classes to be analysed
samples <- distinct(amr_data_with_ncbi, sample) %>% pull(sample)
classes <- distinct(amr_data_with_ncbi, Class) %>% filter(!is.na(Class)) %>% pull(Class) 

# 2. make an empty data frame, with just the samples and classes
amr_summarised <- expand.grid(sample = samples, Class = classes)
# 3. add the metadata
amr_summarised_meta <- left_join(amr_summarised, metadata, by = join_by(sample == Lane))

```

```{r ncbi_amr_info}
# 4. take the "found" genes
# we filter for 100% cov here, and these are the "present" genes
amr_data_with_ncbi_cln <- amr_data_with_ncbi %>% filter(`Average Percent Coverage` == 100) %>% filter(Class != 'NA')

# 5. summarise the data (by class)
amr_data_with_ncbi_cln_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))
# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)
# 6. join the data to the template (amr_summarised_meta)
amr_summarised_meta_with_data <- left_join(amr_summarised_meta, amr_data_with_ncbi_cln_summarised, by = join_by(sample == sample, Class == Class)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% mutate(Class = ifelse(is.na(Class), "None", Class))

# amr_data_with_ncbi_cln_meta <- left_join(amr_data_with_ncbi_cln, metadata, by = join_by(sample == Lane))
View(amr_summarised_meta_with_data %>% filter(Class == 'AMINOGLYCOSIDE') %>% select(sample, Class, n, Group, Country) %>% arrange(Country, Group))

```


```{r overall_graphs}
number_of_amr_genes_per_sample <- amr_summarised_meta_with_data %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
number_of_amr_genes_per_sample %>% group_by(Country) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  ylab('Number of AMR genes identified per sample') +
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_boxplot.png"), plot = p, width = 5, height = 4)

p

p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpkm_boxplot.png"), plot = p, width = 5, height = 4)

p
```



```{r ncbi_amr_graphs}

# amr_data_with_ncbi_cln_meta_summarised <- amr_data_with_ncbi_cln_meta %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
amr_summarised_meta_with_data$Group <- factor(amr_summarised_meta_with_data$Group, levels = c("Control_HealthySerosurvey", "Acute_Typhi", "Carrier"))

# classes <- unique(amr_summarised_meta_with_data$Class)
# cut down classes with some manual removal of low numbers
classes <- c("AMINOGLYCOSIDE", "BETA-LACTAM", "LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN", "TETRACYCLINE", "TRIMETHOPRIM", "EFFLUX", "SULFONAMIDE")

# ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs',  paste(gsub("/", "_", drug_class), "number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)


# by country by group, number of genes
for (drug_class in classes) {
  # print(drug_class)
  drug_class_data <- filter(amr_summarised_meta_with_data, Class == drug_class)
  # View(drug_class_data)
  p <- ggviolin(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group", draw_quantiles = 0.5) + 
    stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
    stat_compare_means(comparisons = list(c('Control_HealthySerosurvey', 'Acute_Typhi'), c('Control_HealthySerosurvey', 'Carrier'), c('Acute_Typhi', 'Carrier'))) + 
    rremove("x.text") + 
    rremove("xlab") + 
    rotate_x_text(angle = 45) +
    ggtitle(drug_class) +
    geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
  ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs',  paste(gsub("/", "_", drug_class), "number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}
# by country, number of genes
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class)
    p <- ggboxplot(drug_class_data, y = "n", x = "Country") + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs',  paste(gsub("/", "_", drug_class), "number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}
# by country, by group, rpkm
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class)
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    # print(drug_class)
    # View(drug_class_data)
    
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control_HealthySerosurvey', 'Acute_Typhi'), c('Control_HealthySerosurvey', 'Carrier'), c('Acute_Typhi', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) + scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', paste(gsub("/", "_", drug_class), "rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, rpkm
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class)
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = log10(max(drug_class_data$sum_reads_per_kb) * 0.9)) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0) +
      scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', paste(gsub("/", "_", drug_class), "rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```


```{r betalactam_graphs}

# 1. set the samples and classes to be analysed (samples set above)
subclasses <- c('BETA-LACTAM', 'CEPHALOSPORIN')
# 2. make an empty data frame, with just the samples and classes
amr_summarised_betalac <- expand.grid(sample = samples, Subclass = subclasses)
# 3. add the metadata
amr_summarised_betalac_meta <- left_join(amr_summarised_betalac, metadata, by = join_by(sample == Lane))
# 4. take the found genes (can just use amr_data_with_ncbi_cln from above)
# 5. summarise the data (by Subclass)
amr_data_with_ncbi_cln_betalac_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Subclass) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))

# 6. join the data to the template (amr_summarised_betalac_meta)

amr_summarised_meta_with_data_betalac <- left_join(amr_summarised_betalac_meta, amr_data_with_ncbi_cln_betalac_summarised, by = join_by(sample == sample, Subclass == Subclass)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb))

# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)

amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control_HealthySerosurvey", "Acute_Typhi", "Carrier"))

# amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control_HealthySerosurvey", "Acute_Typhi", "Carrier"))

# by country by group, boxplot of number of genes
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group") + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control_HealthySerosurvey', 'Acute_Typhi'), c('Control_HealthySerosurvey', 'Carrier'), c('Acute_Typhi', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of number of genes
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, y = "n", x = "Country") + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +       rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country by group, boxplot of rpkm
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control_HealthySerosurvey', 'Acute_Typhi'), c('Control_HealthySerosurvey', 'Carrier'), c('Acute_Typhi', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of rpkm
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    # View(drug_class_data)
    
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```

```{r amr_data_cleaning}

amr_data_cln <- amr_data %>% filter(`Average Percent Coverage` == 100)
amr_data_cln_meta <- left_join(amr_data_cln, metadata, by = join_by(sample == Lane))

amr_data_cln_w <- amr_data_cln %>% select(sample, `ARO Term`, reads_per_kb) %>% pivot_wider(names_from = sample, values_from = reads_per_kb) %>% replace(is.na(.), 0)
amr_data_cln_w <- amr_data_cln_w %>% column_to_rownames(var = "ARO Term")
metadata_r <- metadata
row.names(metadata_r) <-  metadata_r$Lane

# amr_data_cln

# amr_data_cln_w_binary <- amr_data_cln_w %>% replace(.>0, 1)

```

```{r basic_amr_plots}


amr_data_cln_meta_extended <- amr_data_cln_meta %>% select(`Drug Class`, Country, Group) %>% separate_longer_delim(`Drug Class`, '; ')
amr_data_cln_meta_extended_summarised <- amr_data_cln_meta_extended %>% group_by(`Drug Class`, Country, Group) %>% summarise(n = n())
```

```{r run_maaslin_amr, include=FALSE}
Maaslin2(input_data = amr_data_cln_w, input_metadata = metadata_r, analysis_method = "LM", min_prevalence = 0,
           output         = '~/Desktop/maaslin2_output_amr', 
           fixed_effects  = c('Country', 'Sex', 'Group', 'Age', 'Antibiotics_taken_before_sampling_yes_no_assumptions'),
           reference = c('Country,Malawi', 'Group,Control_HealthySerosurvey','Antibiotics_taken_before_sampling_yes_no_assumptions,Yes'))
```

Run kruskall on everything, and then do all the pairwise wilcoxon tests for every gene where the kruskal test is significant.

```{r kruskal}
# widen, then lengthen so that the matrix is complete. otherwise have lots of genes with no entry for a sample, when it should be 0.
# i.e. if you htink of the all by all matrix, the clean one has lots of NAs in.
amr_data_cln_w_l <- amr_data_cln_w %>% pivot_longer(!`ARO Term`, names_to = "sample", values_to = "reads_per_kb")
amr_data_cln_w_l_meta <- left_join(amr_data_cln_w_l, metadata, by = join_by(sample == Lane))

amr_country_kruskal_results <- amr_data_cln_w_l_meta %>% group_by(`ARO Term`) %>% do(tidy(kruskal.test(reads_per_kb ~ Country, data = .))) %>% ungroup() %>% mutate(p.adjust = p.adjust(p.value, method = "bonferroni")) %>% arrange(p.adjust)
amr_country_kruskal_results_sig <- amr_country_kruskal_results %>% filter(p.adjust < 0.05) %>% pull(`ARO Term`)

amr_data_cln_w_l_meta_sig <- amr_data_cln_w_l_meta %>% filter(`ARO Term` %in% amr_country_kruskal_results_sig)


pairwise_results <- list()
# Loop through each significant gene
for (gene_name in amr_country_kruskal_results_sig) {
    gene_data <- filter(amr_data_cln_w_l_meta_sig, `ARO Term` == gene_name)
    
    # Check if all countries are present for the gene
    if (length(unique(gene_data$Country)) == 3) {
        combinations <- combn(unique(gene_data$Country), 2, simplify = TRUE)
        
        # Loop through country combinations
        for (col in seq(ncol(combinations))) {
            country1 <- combinations[1, col]
            country2 <- combinations[2, col]
            
            test_result <- wilcox.test(reads_per_kb ~ Country, data = filter(gene_data, Country %in% c(country1, country2)))
            
            pairwise_results[[length(pairwise_results) + 1]] <- data.frame(
                gene = gene_name,
                comparison = paste(country1, "vs", country2),
                statistic = test_result$statistic,
                p.value = test_result$p.value
            )
        }
    }
}

# Convert list to dataframe and adjust p-values
pairwise_results_df <- do.call(rbind, pairwise_results)
pairwise_results_df$adj.p.value <- p.adjust(pairwise_results_df$p.value, method = "bonferroni")

# print(pairwise_results_df)
write_csv(pairwise_results_df, file.path('~/Desktop', "pairwise_results.csv"))

```

```{r amr_plots}


# Calculate pairwise comparisons for each gene and store results in a list
# multiply the pvalue by 156 as this is the number of comparisons to get the 
# bonferoni corrected p-value

x <- amr_data_cln_w_l_meta_sig %>% filter(`ARO Term` == "tet(M)")
xx <- wilcox.test(reads_per_kb ~ Country, data = filter(x, Country %in% c("Malawi", "Bangladesh")))

pairwise_p_values <- lapply(amr_country_kruskal_results_sig, function(gene_name) {
  gene_data <- amr_data_cln_w_l_meta_sig %>% 
    filter(`ARO Term` == gene_name)
  
  list(
    country1_vs_country2 = wilcox.test(reads_per_kb ~ Country, data = filter(gene_data, Country %in% c("Malawi", "Bangladesh")))$p.value * 156,
    country2_vs_country3 = wilcox.test(reads_per_kb ~ Country, data = filter(gene_data, Country %in% c("Bangladesh", "Nepal")))$p.value * 156,
    country1_vs_country3 = wilcox.test(reads_per_kb ~ Country, data = filter(gene_data, Country %in% c("Malawi", "Nepal")))$p.value * 156
  )
})

modify_values <- function(x) {
  if (is.na(x)) {
    return(NaN)
  } else if (x > 1) {
    return(1)
  } else {
    return(x)
  }
}

# Apply the modification function to the nested list
pairwise_p_values <- lapply(pairwise_p_values, function(gene) {
  lapply(gene, modify_values)
})


names(pairwise_p_values) <- amr_country_kruskal_results_sig


# Loop through each gene, create the boxplot, add p-values, and save
for (gene_name in amr_country_kruskal_results_sig) {
  
  gene_data <- amr_data_cln_w_l_meta_sig %>% dplyr::filter(`ARO Term` == gene_name)
  
  # Extract p-values for current gene
  current_p_values <- pairwise_p_values[[gene_name]]
  max_read <- max(gene_data$reads_per_kb)

  # Create boxplot
  p <- ggplot(gene_data, aes(x = Country, y = reads_per_kb, fill = Country)) +
    geom_boxplot() +

    geom_segment(aes(x = 1, xend = 2, y = max_read * 0.75, yend = max_read * 0.75)) +
    geom_segment(aes(x = 2, xend = 3, y = max_read * 0.85, yend = max_read * 0.85)) +
    geom_segment(aes(x = 1, xend = 3, y = max_read * 0.95, yend = max_read * 0.95)) +
    

    geom_text(aes(x = 1.5, y = max(reads_per_kb) * 0.8, label = paste0("p = ", round(current_p_values$country1_vs_country2, 3)))) +
    geom_text(aes(x = 2.5, y = max(reads_per_kb) * 0.9, label = paste0("p = ", round(current_p_values$country2_vs_country3, 3)))) +
    geom_text(aes(x = 2, y = max(reads_per_kb), label = paste0("p = ", round(current_p_values$country1_vs_country3, 3)))) +
    scale_y_log10() +
    theme_bw() +
    labs(title = paste0("Boxplot of Reads by Country for ", gene_name),
         y = "Number of Reads",
         x = "Country")
  
  # Save the plot
  ggsave(filename =  file.path('~/Desktop', paste(gene_name, "_log_boxplot.png")), plot = p, width = 10, height = 6)
}

```

```{r amr_stats}

```

```{r basic_amr_stats}
length(unique(amr_data_cln$`ARO Term`))
amr_data_cln %>% group_by(`ARO Term`) %>% summarise(n = n()) %>% arrange(desc(n))

```