---
title: "STRATAA Microbiome"
author: "Philip Ashton"
date: "`r Sys.Date()`"
output: 
  html_document:
      toc: true
      toc_float: true
      theme: united
      number_sections: true
---


```{r setup, error=FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)

```

# STRATAA microbiome analysis

## Imports
```{r imports, error=FALSE}
library(rbiom)
library(kableExtra)
library(ggplot2)
library(ggpubr)
library(ggforce)
library(patchwork)
library(vegan)
library(dplyr)
library(forcats)
library(RColorBrewer)

```


## Sources

The file handles are set in config.R as they're used by both this script and data_cleaning.

```{r source, error=FALSE}
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```


## Metadata basics

First do the plots with kruskall wallis comparison between groups to see if there's an overall difference, then do the pairwise tests because KW says there is a difference.

```{r metadata_read_in, error=FALSE}
metadata <- read_metadata(metadata_handle)
# putting this here so that the output files of maaslin get named accroding to the variable names in the metadata file.
metadata <- metadata %>% mutate(Group = if_else(Group == 'Control_HealthySerosurvey', 'Household contact', Group)) %>% mutate(Group = if_else(Group == 'Acute_Typhi', 'Acute typhoid', Group)) %>% mutate(Group = if_else(Group == 'Carrier', 'High Vi-titre', Group))


three_qual_colours <- brewer.pal(3, "Dark2")

participant_group_colours <- c("Acute typhoid" = three_qual_colours[[2]], "Household contact" = three_qual_colours[[1]], "High Vi-titre" = three_qual_colours[[3]])
```

```{r metadata_basics, error=FALSE}
number_per_country <- metadata %>% group_by(Country) %>% summarise(count = n())

number_per_country %>% kbl() %>% kable_styling()

print(sum(number_per_country$count))

metadata %>% group_by(Country, Group) %>% summarise(count = n()) %>% kbl() %>% kable_styling()

metadata %>% group_by(Group) %>% summarise(count = n(), median_age = median(Age)) %>% kbl() %>% kable_styling()
baseline_chars <- get_baseline_characteristics(metadata)
comparisons_groups <- list(c("Acute typhoid", "High Vi-titre"), c("Acute typhoid", "Household contact"), c("High Vi-titre", "Household contact"))

ggboxplot(metadata, y = "Age", x = "Group", color = "Group") + 
  stat_compare_means() + 
  stat_compare_means(comparisons = comparisons_groups) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rremove("x.ticks") +
  scale_color_manual(values = participant_group_colours)
   # +rotate_x_text(angle = 45)


baseline_chars_table <- baseline_chars %>% 
  rename(`Median age` = median_age, `Women (%)` = pct_fem, `Antibiotics in last 2 weeks (%)` = pct_anti) %>% 
  pivot_longer(!c(Country, Group)) %>% 
  rename(variable_name = name) %>% 
  pivot_wider(names_from = Country, values_from = value)

baseline_chars_table %>% kbl(digits = c(NA, NA, 1, 1, 1)) %>% kable_styling()

# the kruskal wallis test is positive, showing a difference between the groups, so we move onto pairwise tests.
# ggplot(metadata, aes(x = Country, y = Age, fill = Group)) + geom_boxplot() + stat_compare_means(method = 'kruskal.test', label = "p")
comparisons_countries <- list(c('Bangladesh', 'Malawi'), c('Bangladesh', 'Nepal'), c('Malawi', 'Nepal'))
# default stat method when doing pairwise tests in wilcoxon.
metadata_for_plotting <- metadata
metadata_for_plotting$Group <- factor(metadata_for_plotting$Group, levels = c("Household contact", "Acute typhoid", "High Vi-titre"))
ggboxplot(metadata_for_plotting, facet.by = "Country", y = "Age", x = "Group", color = "Group") + stat_compare_means(comparisons = comparisons_groups) + rremove("x.text") + rremove("xlab") + rremove("x.ticks") + scale_color_manual(values = participant_group_colours) # +rotate_x_text(angle = 45)
# ggboxplot(metadata_for_plotting, facet.by = "Group", y = "Age", x = "Country", color = "Country") + stat_compare_means(comparisons = comparisons_countries) + rotate_x_text(angle = 45)

```

```{r metadata_basic_plots, error=FALSE, fig.width = 4, fig.height= 8}
country_group_sex <- metadata %>% group_by(Country, Group, Sex) %>% summarise(count = n()) 
comparisons_groups <- list(c("Acute typhoid", "High Vi-titre"), c("Acute typhoid", "Household contact"), c("High Vi-titre", "Household contact"))
# country_group_sex %>% group_by(Country, Sex) %>% summarise(n = sum(count))
plot_sex <- function(eg1, c){
  d <- eg1 %>% filter(Country == c)
  p <- ggplot(d, aes(x = Group, y = count, fill = Sex)) + 
    geom_bar(stat ='identity', position = 'fill') + 
    ylab('Proportion') + 
    ggtitle(c)
    # +
    #theme(axis.text=element_text(size=34), axis.title=element_text(size=36,face="bold"), plot.title = element_text(size = 40, face = "bold"), legend.key.size = unit(4, 'cm'), legend.title = element_text(size = 34), legend.text = element_text(size = 28))
  
  return(p)
}

country_group_sex$Group <- factor(country_group_sex$Group, levels = c("Household contact", "Acute typhoid", "High Vi-titre"))

m_sex <- plot_sex(country_group_sex, 'Malawi')
b_sex <- plot_sex(country_group_sex, 'Bangladesh')
n_sex <- plot_sex(country_group_sex, 'Nepal')
m_sex / b_sex / n_sex 
```

## sequencing read stats

```{r sequencing_read_stats, error=FALSE}

# todo - replace with the read stats from the metadata file.
# read_stats_path <- '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/data_munging/2023.08.28/2023.08.28 strataa microbiome number of reads.csv'
# read_stats <- read_csv(read_stats_path) %>% filter(Country != '#N/A')

ggboxplot(metadata, facet.by = "Country", x = "Group", y = "number_of_reads", color = "Group") + rremove("x.text") + rremove("xlab") + rremove("x.ticks") + rotate_x_text(angle = 45) + stat_compare_means()

```

## read in metaphlan data
```{r read_in_metaphlan, error=FALSE}
strataa_metaphlan_data <- read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.all_strataa_metaphlan.csv'), header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_metaphlan_data$lowest_taxonomic_level <- sapply(str_split(row.names(strataa_metaphlan_data), "\\|"), function(x) x[length(x)])
strataa_metaphlan_data_species <- strataa_metaphlan_data %>% filter(str_starts(lowest_taxonomic_level, 's__'))

# metadata <- read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.strataa_metadata.metaphlan.csv'), header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)

```

## Healthy vs Typhi

### alpha diversity
Alpha diversity - all countries, healthy and acute

```{r alpha_diversity_healthy_acute, error=FALSE}

all_countries_healthy_acute_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, metadata, countries_of_interest = c('Bangladesh', 'Malawi', 'Nepal'), groups_of_interest = c('Acute typhoid', 'Household contact'), comparisons = list(c('Acute typhoid', 'Household contact')), participant_group_colours = participant_group_colours)
# all_countries_healthy_acute_alpha$alpha_by_country
all_countries_healthy_acute_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_acute_alpha$alpha_plot_group
# all_countries_healthy_acute_alpha$alpha_plot_antibiotics
```



### beta diversity 

Acute vs healthy.
```{r beta_diversity_healthy_acute, error=FALSE}

all_countries_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Bangladesh', 'Malawi', 'Nepal'), c('Acute typhoid', 'Household contact'), participant_group_colours)
all_countries_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

bgd_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Bangladesh'), c('Acute typhoid', 'Household contact'), participant_group_colours)
bgd_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi'), c('Acute typhoid', 'Household contact'), participant_group_colours)
mal_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

nep_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Nepal'), c('Acute typhoid', 'Household contact'), participant_group_colours)
nep_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

all_countries_beta_acute_healthy$pc12 + bgd_beta_acute_healthy$pc12 + mal_beta_acute_healthy$pc12 + nep_beta_acute_healthy$pc12 + plot_layout(guides = 'collect')
```

### maaslin2 taxonomy
Maaslin basics

```{r basics_for_maaslin_taxonomic, error=FALSE}
# the names here should be the full name from the metadata file, not the "presentation" name
# because this is used to read in the files, which are written using the full name.
groups_to_analyse <- c('Acute_Typhi', 'Control_HealthySerosurvey')
bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")
nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
```

```{r read_in_maaslin_taxonomic, error=FALSE}
bangladesh_taxonomic_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'metaphlan')
malawi_taxonomic_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'metaphlan')
nepal_taxonomic_maaslin <- read_in_maaslin('Nepal', groups_to_analyse, nep_variables_for_analysis, 'metaphlan')
```

```{r filter_maaslin_taxonomic, error=FALSE}
bangladesh_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(bangladesh_taxonomic_maaslin)
malawi_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(malawi_taxonomic_maaslin)
nepal_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(nepal_taxonomic_maaslin)
```

```{r basic_stats_maaslin_taxonomic, error=FALSE}
bangladesh_maaslin_stats <- basic_maaslin_stats(bangladesh_taxonomic_maaslin_filtered, 'Bangladesh', bang_variables_for_analysis, groups_to_analyse)
malawi_maaslin_stats <- basic_maaslin_stats(malawi_taxonomic_maaslin_filtered, 'Malawi', mwi_variables_for_analysis, groups_to_analyse)
nepal_maaslin_stats <- basic_maaslin_stats(nepal_taxonomic_maaslin_filtered, 'Nepal', nep_variables_for_analysis, groups_to_analyse)
```

```{r volcano_plots_maaslin_taxonomic, error=FALSE}

malawi_maaslin_stats$volcano_plot / bangladesh_maaslin_stats$volcano_plot / nepal_maaslin_stats$volcano_plot
nrow(malawi_maaslin_stats$maaslin_results_sig)
nrow(bangladesh_maaslin_stats$maaslin_results_sig)
nrow(nepal_maaslin_stats$maaslin_results_sig)
```

There were `r nrow(malawi_maaslin_stats$maaslin_results_sig)` species significantly (q-val < 0.05) associated with health/disease in Malawi, `r nrow(bangladesh_maaslin_stats$maaslin_results_species_group_sig)` in Bangladesh, and `r nrow(nepal_maaslin_stats$maaslin_results_species_group_sig)` in Nepal.

Combine the taxonomic maaslins, and print out the species that are sig in both and share directions.

Because sequencing run and participant type are totally confounded for Bangladesh, need to exclude sequencing run from the final model for Bangladesh (otherwise, wipes out the signals).

associated at both sites
```{r combine_maaslins_taxonomic, error=FALSE}

bang_mal <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered)
combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# View(combined_results$positive_coef)
# View(combined_results$negative_coef)


combined_results$positive_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

combined_results$negative_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

# todo - refactoring the run_combine_maaslins means that we dont get the species that are only associated at one site. need to fix that.

# nrow(combined_results$mwi_maaslin_only)
# nrow(combined_results$bang_maaslin_only)
```

There were `r nrow(combined_results$mwi_maaslin_only)` species significantly (q-val < 0.05) associated with health/disease in Malawi only and `r nrow(combined_results$bang_maaslin_only)` in Bangladesh only.

The ones associated at only one site are written out to a file, you can look at them manually there.

### Forest plot

Setting up the forest plots. Need to combine all three maaslin outputs, and then add in the patch data.

```{r combine_maaslins_for_strataa_and_patch, error=FALSE}
# combine the maaslines for bgd and mal, and then add in the nepal ones.
combined_maaslins <- combine_maaslins(bangladesh_taxonomic_maaslin, malawi_taxonomic_maaslin, '_bang', '_mal', 'metaphlan')
combined_maaslins <- combine_maaslins(combined_maaslins, nepal_taxonomic_maaslin, 'not', 'used', 'metaphlan')
# for the nepal columns, give them a "_nep" suffix
combined_maaslins <- combined_maaslins %>%
  rename_with(~ paste0(.x, '_nep'), c("coef", "stderr", "N", "N.not.0", "pval", "qval"))

# get the patch data
patch_taxonomic_maaslin <- read_tsv(file.path(patch_maaslin_taxonomic_output_root_folder, 'baseline_typhi_species', 'all_results.tsv'))
patch_taxonomic_maaslin$lowest_taxonomic_level <- sapply(str_split(patch_taxonomic_maaslin$feature, "\\."), function(x) x[length(x)])
patch_taxonomic_maaslin <- patch_taxonomic_maaslin %>% relocate(lowest_taxonomic_level, .after = feature)
patch_taxonomic_maaslin <- patch_taxonomic_maaslin %>%
  rename_with(~ paste0(.x, '_patch'), c("coef", "stderr", "N", "N.not.0", "pval", "qval"))

# in patch_taxonomic_maaslin, change all Diagnosis in metadata column to Group, and all 'no_disease' in value column to 'ControlHealthySerosurvey'
# this is to match the other maaslin outputs
patch_taxonomic_maaslin <- patch_taxonomic_maaslin %>% filter(metadata == 'Diagnosis')
# in patch_taxonomic_maaslin, change all Diagnosis in metadata column to Group, and all 'no_disease' in value column to 'ControlHealthySerosurvey'
patch_taxonomic_maaslin$metadata <- ifelse(patch_taxonomic_maaslin$metadata == 'Diagnosis', 'Group', patch_taxonomic_maaslin$metadata)
patch_taxonomic_maaslin$value <- ifelse(patch_taxonomic_maaslin$value == 'no_disease', 'Control_HealthySerosurvey', patch_taxonomic_maaslin$value)

# join the patch data to the combined maaslin data
combined_maaslins <- combined_maaslins %>% left_join(patch_taxonomic_maaslin, by = c("feature", 'metadata', 'value', 'lowest_taxonomic_level'))

```

for the forest plot, i also want to include the per-cohort abundance medians for the species of interest.

```{r species_of_interest_prevalence, error=FALSE}

groups_to_analyse <- c('Acute typhoid', 'Household contact')
prevalence <- get_prevalence(strataa_metaphlan_data_species, groups_to_analyse)

```
do the forest plot

```{r per_species_data_for_forest_plot, error=FALSE}
# prevalence is the relative abundance data, the species of interest are the ones that are significantly associated with health/disease in two countries, and combined_maaslins is the maaslin output with the associations
run_forest_plot(prevalence, c('s__Prevotella_copri_clade_A', 's__Clostridium_SGB6179', 's__GGB4266_SGB5809', 's__Haemophilus_parainfluenzae'), combined_maaslins)

```


### plots of species of interest

```{r plot_species_of_interest_taxonomic, error=FALSE}

strataa_metaphlan_data_longer <- strataa_metaphlan_data %>% mutate(feature = rownames(strataa_metaphlan_data)) %>% pivot_longer(!c(feature, lowest_taxonomic_level), names_to = "SampleID", values_to = "prevalence")
# View(head(strataa_metaphlan_data_longer))
strataa_metaphlan_data_longer_meta <- strataa_metaphlan_data_longer %>% left_join(metadata, by = c("SampleID" = "SampleID"))

pc <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Prevotella_copri_clade_A', participant_group_colours)
cs <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Clostridium_SGB6179', participant_group_colours)
SGB5809 <-run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__GGB4266_SGB5809', participant_group_colours)
hp <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Haemophilus_parainfluenzae', participant_group_colours)
# rt <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Romboutsia_timonensis', participant_group_colours)
# lb <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Lachnospiraceae_bacterium', participant_group_colours)

# pc
  # pc + cs + SGB5809 + hp + rt + lb

```

### maaslin functional groups 

Functional groups associated with health

```{r read_in_maaslin_functional, error=FALSE}

bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")
# nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
groups_to_analyse <- c('Acute_typhi', 'Control_HealthySerosurvey')

bang_func_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'bigmap')
mal_func_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'bigmap')
```

```{r maaslin_functional_groups, error=FALSE}

positive_table <- combined_results$positive_coef %>%
  dplyr::select(-metadata, -value, -N_bang, -N.not.0_bang, -pval_bang, -N_mal, -N.not.0_mal, -pval_mal) %>% 
  rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, 
         `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, ~as.character(signif(., 3))) %>% 
  kbl() %>% 
  kable_styling()

# Process negative coefficients
negative_table <- combined_results$negative_coef %>%
  dplyr::select(-metadata, -value, -N_bang, -N.not.0_bang, -pval_bang, -N_mal, -N.not.0_mal, -pval_mal) %>% 
  rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, 
         `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, ~as.character(signif(., 3))) %>% 
  kbl() %>% 
  kable_styling()

# Split species and clean data for positive coefficients
positive_cleaned <- combined_results$positive_coef %>% 
  mutate(split_species = str_split(Species, '_')) %>% 
  mutate(genus = sapply(split_species, "[[", 1)) %>% 
  mutate(specific = sapply(split_species, "[[", 2)) %>% 
  mutate(clean_species = paste(genus, specific, sep = ' ')) %>% 
  dplyr::select(-split_species, -genus, -specific) %>% 
  relocate(clean_species, .after = Species) %>% 
  group_by(MGC_class) %>% 
  arrange(Species) %>% 
  summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% 
  arrange(desc(n), MGC_class) %>%  
  kbl() %>% 
  kable_styling()

# Split species and clean data for negative coefficients
negative_cleaned <- combined_results$negative_coef %>% 
  mutate(split_species = str_split(Species, '_')) %>% 
  mutate(genus = sapply(split_species, "[[", 1)) %>% 
  mutate(specific = sapply(split_species, "[[", 2)) %>% 
  mutate(clean_species = paste(genus, specific, sep = ' ')) %>% 
  dplyr::select(-split_species, -genus, -specific) %>% 
  relocate(clean_species, .after = Species) %>% 
  group_by(MGC_class) %>% 
  arrange(Species) %>% 
  summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% 
  arrange(desc(n), MGC_class) %>%  
  kbl() %>% 
  kable_styling()

# Display the tables
positive_table
negative_table
positive_cleaned
negative_cleaned

```

## healthy vs carrier
### phylum plots

```{r phylum_plots_carrier_healthy, error=FALSE}
order_of_groups <- c("High Vi-titre", "Household contacts")
# bangladesh_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Bangladesh", group_order = order_of_groups)
# bangladesh_phyla_plot
malawi_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Malawi", group_order = order_of_groups)
nepal_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Nepal", group_order = order_of_groups)

# bangladesh_phyla_plot / 
malawi_phyla_plot / nepal_phyla_plot


```

### alpha diversity

Alpha diversity - all countries, healthy and carrier

```{r alpha_diversity_healthy_carrier, error=FALSE}

all_countries_healthy_carrier_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, metadata, countries_of_interest = c('Malawi', 'Nepal'), groups_of_interest = c('High Vi-titre', 'Household contact'), comparisons = list(c('High Vi-titre', 'Household contact')), participant_group_colours)
all_countries_healthy_carrier_alpha$alpha_by_country
all_countries_healthy_carrier_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_carrier_alpha$alpha_plot_group
```

### beta diversity

High Vi-titre vs healthy.
```{r beta_diversity_healthy_carrier, error=FALSE}

all_countries_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi', 'Nepal'), c('High Vi-titre', 'Household contact'), participant_group_colours)
all_countries_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

# bgd_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Bangladesh'), c('High Vi-titre', 'Household contact'))
# bgd_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi'), c('High Vi-titre', 'Household contact'), participant_group_colours)
mal_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

nep_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Nepal'), c('High Vi-titre', 'Household contact'), participant_group_colours)
nep_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_carrier_healthy$pc12 + nep_beta_carrier_healthy$pc12 + plot_layout(guides = 'collect')
```

### maaslin taxonomic groups

```{r basics_for_maaslin_taxonomic_carrier_healthy, error=FALSE}
# groups_to_analyse <- c('Acute_typhi', 'Control_HealthySerosurvey')

groups_to_analyse <- c('Carrier', 'Control_HealthySerosurvey')
# bang_variables_for_analysis <- c("Group", "Sex", "Age")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
nep_variables_for_analysis <- c("Group", "Sex", "Age")

```

```{r read_in_maaslin_taxonomic_carrier_healthy, error=FALSE}
# bangladesh_taxonomic_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'metaphlan')
malawi_taxonomic_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'metaphlan')
nepal_taxonomic_maaslin <- read_in_maaslin('Nepal', groups_to_analyse, nep_variables_for_analysis, 'metaphlan')
```

```{r filter_maaslin_taxonomic_carrier_healthy, error=FALSE}
# bangladesh_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(bangladesh_taxonomic_maaslin)
malawi_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(malawi_taxonomic_maaslin)
nepal_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(nepal_taxonomic_maaslin)
```

```{r basic_stats_maaslin_taxonomic_carrier_healthy, error=FALSE}
# bangladesh_maaslin_stats <- basic_maaslin_stats(bangladesh_taxonomic_maaslin_filtered, 'Bangladesh', bang_variables_for_analysis, groups_to_analyse)
malawi_maaslin_stats <- basic_maaslin_stats(malawi_taxonomic_maaslin_filtered, 'Malawi', mwi_variables_for_analysis, groups_to_analyse)
nepal_maaslin_stats <- basic_maaslin_stats(nepal_taxonomic_maaslin_filtered, 'Nepal', nep_variables_for_analysis, groups_to_analyse)
```

```{r volcano_plots_maaslin_taxonomic_carrier_healthy, error=FALSE}
# / bangladesh_maaslin_stats$volcano_plot 
malawi_maaslin_stats$volcano_plot / nepal_maaslin_stats$volcano_plot
nrow(malawi_maaslin_stats$maaslin_results_sig)
# nrow(bangladesh_maaslin_stats$maaslin_results_sig)
nrow(nepal_maaslin_stats$maaslin_results_sig)
```

We're not including the bangladesh samples in this analysis, because the bangladesh carriers were processed differently (extracted without being frozen).

There are no species significantly associated with carrier status in all Mal and Nep, so not including the combine.

```{r combine_maaslins_taxonomic_carrier_healthy, error=FALSE}

# combined_results <- run_combine_maaslins(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered, bang_variables_for_analysis, mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# only one species significantly associated in Nepal.
# there are no species consistently associated across all three sites
# bang_mal_nep <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_mal_nep, c('_bang', '_mwi', '_nep'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# there are no species associated in both bang and nep
# bang_nep <- list(bangladesh_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_nep, c('_bang', '_nep'), bang_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# there are no species associated in both mal and nep
# mal_nep <- list(malawi_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_nep, c('_mal', '_nep'), bang_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# bang_mal <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# # positive is associated with health
# combined_results$positive_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()

# # negative is associated with carrier
# combined_results$negative_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()

# nrow(combined_results$mwi_maaslin_only)
# nrow(combined_results$bang_maaslin_only)
```


### functional groups

```{r read_in_maaslin_functional_healthy_carrier, error=FALSE}
# bang_variables_for_analysis <- c("Group", "Sex", "Age")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
# nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
groups_to_analyse <- c('High Vi-titre', 'Household contact')

# bang_func_carr_health_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'bigmap')
mal_func_carr_health_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'bigmap')
```


```{r maaslin_functional_groups_healthy_carrier, error=FALSE}
# combined_results <- run_combine_maaslins(bang_func_carr_health_maaslin, mal_func_carr_health_maaslin, bang_variables_for_analysis, mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)

# bang_mal <- list(bang_func_carr_health_maaslin, mal_func_carr_health_maaslin)
# combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)

# # positive is associated with health
# # combined_results$positive_coef %>%  kbl() %>% kable_styling()
# combined_results$positive_coef %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()
# # negative is associated with carrier
# # combined_results$negative_coef %>%  kbl() %>% kable_styling()
# combined_results$negative_coef %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()


# # print out a nicely formatted table of the count of the MGCs
# combined_results$positive_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()
# combined_results$negative_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()
```

## healthy vs acute vs carriers 

### alpha

```{r alpha_diversity_healthy_acute_carrier, error=FALSE}

all_countries_healthy_acute_carrier_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, metadata, countries_of_interest = c('Malawi', 'Nepal'), groups_of_interest = c('High Vi-titre', 'Household contact', 'Acute typhoid'), comparisons = list(c('Acute typhoid', 'High Vi-titre'), c('Acute typhoid', 'Household contact'), c('High Vi-titre', 'Household contact')), participant_group_colours)

all_countries_healthy_acute_carrier_alpha$alpha_by_country
all_countries_healthy_acute_carrier_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_acute_carrier_alpha$alpha_plot_group
```

### beta

```{r beta_diversity_healthy_acute_carrier, error=FALSE}

all_countries_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi', 'Nepal'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
all_countries_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

# bgd_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data, metadata, c('Bangladesh'), c('High Vi-titre', 'Household contact'))
# bgd_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_acute_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
mal_beta_acute_carrier_healthy$pn_res %>% kbl %>% kable_styling()
mal_beta_acute_carrier_healthy$pc12
# mal_beta_acute_carrier_healthy$pc34

nep_beta_acute_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Nepal'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
nep_beta_acute_carrier_healthy$pn_res %>% kbl %>% kable_styling()
nep_beta_acute_carrier_healthy$pc12
# nep_beta_acute_carrier_healthy$pc34

mal_beta_acute_carrier_healthy$pc12 / nep_beta_acute_carrier_healthy$pc12  + plot_layout(guides = 'collect')


### maaslin2 taxonomic

```{r read_in_heal_ac_car_taxonomic, error=FALSE}
groups_to_analyse <- c('High Vi-titre', 'Household contact', 'Acute typhoid')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions")
heal_ac_car_taxonomic_np <- read_in_maaslin('Nepal', groups_to_analyse, variables_for_analysis, 'metaphlan')
heal_ac_car_taxonomic_np_sig <- filter_taxonomic_maaslin(heal_ac_car_taxonomic_np) %>% filter(qval <= 0.05) 
print(nrow(heal_ac_car_taxonomic_np_sig))
```


```{r heal_ac_car_taxonomic, error=FALSE}
groups_to_analyse <- c('High Vi-titre', 'Household contact', 'Acute typhoid')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_assumptions", "sequencing_lane")
heal_ac_car_taxonomic <- read_in_maaslin('Malawi', groups_to_analyse, variables_for_analysis, 'metaphlan')
heal_ac_car_taxonomic_cln <- filter_taxonomic_maaslin(heal_ac_car_taxonomic) #%>% filter(qval <= 0.05) 
# spread heal_ac_car_taxonomic_cln wider so that the coef is on the same row for each feature
heal_ac_car_taxonomic_cln_w <- heal_ac_car_taxonomic_cln %>% pivot_wider(names_from = value, values_from = c(coef, qval), id_cols = c(feature, lowest_taxonomic_level))

heal_ac_car_taxonomic_mwi_sig <- filter_taxonomic_maaslin(heal_ac_car_taxonomic) %>% filter(qval <= 0.05) 
print(nrow(heal_ac_car_taxonomic_mwi_sig))
```
changes relative to healthy state
```{r maaslin_heal_ac_car_taxonomic, error=FALSE}
# same_same cant do from here because filtered for qvalue
# 1
up_up <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & coef_High Vi-titre > 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre < 0.05) %>% mutate(type = '1.up_up')
print(nrow(up_up))
# write_csv(up_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.1.up_up.csv'))
# 2
up_down <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & coef_High Vi-titre < 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre < 0.05) %>% mutate(type = '2.up_down')
print(nrow(up_down))
# write_csv(up_down, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.2.up_down.csv'))
# 3
up_same <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre >= 0.05) %>% mutate(type = '3.up_same')
print(nrow(up_same))
# write_csv(up_same, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.3.up_same.csv'))
# 4
down_up <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & coef_High Vi-titre > 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre < 0.05) %>% mutate(type = '4.down_up')
print(nrow(down_up))
# write_csv(down_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.4.down_up.csv'))
# 5
down_down <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & coef_High Vi-titre < 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre < 0.05) %>% mutate(type = '5.down_down')
print(nrow(down_down))
# write_csv(down_down, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.5.down_down.csv'))
# 6
down_same <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & qval_Acute typhoid < 0.05 & qval_High Vi-titre > 0.05) %>% mutate(type = '6.down_same')
print(nrow(down_same))
# write_csv(down_same, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.6.down_same.csv'))
# 7
same_up <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid > 0.05 & qval_High Vi-titre < 0.05 & coef_High Vi-titre > 0) %>% mutate(type = '7.same_up')
print(nrow(same_up))
# write_csv(same_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.7.same_up.csv'))
# 8
same_down <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid > 0.05 & qval_High Vi-titre < 0.05 & coef_High Vi-titre < 0) %>% mutate(type = '8.same_down')
print(nrow(same_down))
# 9
same_same <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid >= 0.05 & qval_High Vi-titre >= 0.05) %>% mutate(type = '9.same_same')
print(nrow(same_same))

all_patterns <- rbind(up_up, up_down, up_same, down_up, down_down, down_same, same_up, same_down, same_same)

write_csv(all_patterns, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.all_patterns.csv')) 

up_in_ac_car <- all_patterns %>% filter(type %in% c('1.up_up', '2.up_down', '3.up_same', '4.down_up', '7.same_up')) 
down_in_ac_car <- all_patterns %>% filter(type %in% c('5.down_down', '6.down_same', '8.same_down'))
up_in_health_only <- all_patterns %>% filter(type == '5.down_down')
up_in_car_only <- all_patterns %>% filter(type == '7.same_up')

prop_that_is_sgb <- function(input_data) {
  x <- input_data %>% filter(str_detect(feature, 'SGB')) %>% summarise(prop = n()/nrow(input_data), n_sgb = n(), total_n = nrow(input_data))
  print(x)
  return(x)
}
prop_up_in_ac_car <- prop_that_is_sgb(up_in_ac_car)
prop_down_in_ac_car <- prop_that_is_sgb(down_in_ac_car)
prop_up_in_health_only <- prop_that_is_sgb(up_in_health_only)
prop_up_in_car_only <- prop_that_is_sgb(up_in_car_only)


prop.test(x = c(prop_up_in_ac_car$n_sgb, prop_down_in_ac_car$n_sgb), n = c(prop_up_in_ac_car$total_n, prop_down_in_ac_car$total_n), alternative = 'two.sided', correct = FALSE)
prop.test(x = c(prop_up_in_health_only$n_sgb, prop_up_in_car_only$n_sgb), n = c(prop_up_in_health_only$total_n, prop_up_in_car_only$total_n), alternative = 'two.sided', correct = FALSE)
```

```{r maaslin_heal_ac_car_sgb_plot, error=FALSE}
# make a new strataa_metaphlan_data object where the rowname column is an actual column

strataa_metaphlan_data_cln <- strataa_metaphlan_data %>% mutate(feature = rownames(strataa_metaphlan_data)) 
rownames(strataa_metaphlan_data_cln) <- NULL
strataa_metaphlan_data_l <- strataa_metaphlan_data_cln %>% pivot_longer(!c(feature, lowest_taxonomic_level), names_to = 'sample', values_to = 'abundance') %>% left_join(metadata %>% mutate(sample = rownames(metadata)), by = 'sample')
# head(strataa_metaphlan_data_l)
strataa_metaphlan_data_mwi_sp <- strataa_metaphlan_data_l %>% filter(Country == 'Malawi', str_detect(feature, 's__'), str_detect(feature, 't__', negate = TRUE)) %>% mutate(SGB = str_detect(feature, '_SGB'))
head(strataa_metaphlan_data_mwi_sp)

mwi_sgb_abundance <- strataa_metaphlan_data_mwi_sp %>% group_by(sample, SGB) %>% summarise(total_abundance = sum(abundance)) %>% left_join(metadata %>% mutate(sample = rownames(metadata)), by = 'sample')  %>% mutate(SGB = ifelse(SGB, 'SGB', 'non-SGB')) %>% mutate(Group = factor(Group, levels = c('Household contact', 'Acute typhoid', 'High Vi-titre')))
mwi_sgb_abundance %>% group_by(SGB, Group) %>% summarise(median_abundance = median(total_abundance), n = n()) %>% kbl() %>% kable_styling()
# mwi_sgb_abundance <- mwi_sgb_abundance 
# do a scatter plot of the abundance of sgb in each group
my_comparisons <- list(c('Acute typhoid', 'High Vi-titre'), c('Acute typhoid', 'Household contact'), c('High Vi-titre', 'Household contact'))
mwi_sgb_abundance_plot <- ggplot(mwi_sgb_abundance %>% filter(SGB == 'SGB'), aes(x = Group, y = total_abundance, color = Group)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons) + 
  stat_compare_means(label.y = 70, size = 5) + 
  theme(legend.position = "none") + 
  labs(y = 'SGB abundance in metagenome (%)', x = '') + 
  theme(text = element_text(size=18)) + 
  scale_x_discrete(labels= c('Household contact', 'Acute typhoid', 'Presumptive carrier'))

mwi_sgb_abundance_plot$layers[[2]]$aes_params$textsize <- 5
# mwi_sgb_abundance_plot$layers[[3]]$aes_params$textsize <- 8
# which_layers(mwi_sgb_abundance_plot, "GeomSignif")
mwi_sgb_abundance_plot
ggsave(file.path(metaphlan_input_folder, 'mwi_sgb_abundance.png'), mwi_sgb_abundance_plot, width = 8, height = 4.8, units = 'in', dpi = 300)

```

```{r make_health_acc_carr_table, error=FALSE}

to_get_list <- c('k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Oscillospiraceae.g__Faecalibacterium.s__Faecalibacterium_prausnitzii', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Barnesiellaceae.g__Barnesiella.s__Barnesiella_intestinihominis', 'k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Limosilactobacillus.s__Limosilactobacillus_reuteri', 'k__Bacteria.p__Proteobacteria.c__Gammaproteobacteria.o__Pasteurellales.f__Pasteurellaceae.g__Haemophilus.s__Haemophilus_parainfluenzae', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Erysipelotrichaceae.g__Holdemanella.s__Holdemanella_biformis', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Mediterraneibacter.s__Ruminococcus_gnavus', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptostreptococcaceae.g__Paeniclostridium.s__Eubacterium_tenue', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Turicibacteraceae.g__Turicibacter.s__Turicibacter_bilis', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Turicibacteraceae.g__Turicibacter.s__Turicibacter_sanguinis', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Odoribacteraceae.g__Odoribacter.s__Odoribacter_splanchnicus', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Blautia.s__Blautia_faecis', 'k__Bacteria.p__Proteobacteria.c__Gammaproteobacteria.o__Enterobacterales.f__Enterobacteriaceae.g__Escherichia.s__Escherichia_coli', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Lachnospiraceae_unclassified.s__Eubacterium_rectale', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Bacteroidaceae.g__Bacteroides.s__Bacteroides_fragilis', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Erysipelotrichaceae.g__Holdemanella.s__Holdemanella_porci', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Prevotellamassilia.s__Prevotellamassilia_timonensis')
# this one doesn't include the Oscillospiraceae, so will just have to point towards the supplementary table for those.
heal_acc_carr_table <- all_patterns %>% filter(feature %in% to_get_list)
write_csv(heal_acc_carr_table, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.health_acc_carr_table.csv'))

```



## random forest analysis

Pre-processing.

```{r}
strataa_metaphlan_data_t <- as.data.frame(t(strataa_metaphlan_data))
names(strataa_metaphlan_data_t) <- gsub("\\|", "_", names(strataa_metaphlan_data_t))
strataa_metaphlan_data_t <- strataa_metaphlan_data_t %>% select(contains('_s__')) %>% select(!contains('_t__'))
strataa_metaphlan_data_t$sample_name <- rownames(strataa_metaphlan_data_t)
metadata_copy <- metadata %>% select(Group, Country) %>% mutate(sample_name = rownames(.))

# join the Group column from metadata to the metaphlan data
strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_t %>% left_join(metadata_copy, by = "sample_name") %>% filter(Group != "Acute typhoid", Country != 'Bangladesh') %>% mutate(Group = as.factor(Group))
row.names(strataa_metaphlan_data_heal_carr_rf) <- strataa_metaphlan_data_heal_carr_rf$sample_name
strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_heal_carr_rf %>% select(-sample_name) %>% filter(!is.na(Group))
# remove rows where the group is acute typhi and remove any column that sums to 0
# strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_rf %>% filter(G #%>% select(where(~ !is.numeric(.) || sum(.) != 0))

```

RF - Using a strategy from one of Kevin Bonham's papers https://www.biorxiv.org/content/10.1101/2020.02.13.944181v5.full


# this creates three folds, each of which has 2/3rds of the data in it.
# this is because for a 3 fold CV, we use 2/3 as training, and 1/3 as test.

```{r}

library(randomForest)
library(caret)
# library(dplyr)
# library(tidyr)

# Assuming strataa_metaphlan_data_heal_carr_rf is your dataset
data <- strataa_metaphlan_data_heal_carr_rf %>% select(-Country)

# Parameters for the CV process
n_repeats <- 100
n_folds <- 3
n_rng_states <- 10

# Data frames to store the results
accuracy_df <- data.frame(rep = integer(), 
                          fold = integer(), 
                          rng_state = integer(), 
                          data_type = character(),  # Indicates train or test data
                          accuracy = numeric())

importance_df <- data.frame(feature = character(), 
                            rep = integer(), 
                            fold = integer(), 
                            rng_state = integer(), 
                            importance = numeric())

# Start the cross-validation process
for (rep in 1:n_repeats) {
  fold_indices <- createFolds(data$Group, k = n_folds, list = TRUE)
  
  for (fold in 1:n_folds) {
    train_indices <- fold_indices[[fold]]
    test_indices <- setdiff(1:nrow(data), train_indices)
    train_data <- data[train_indices, ]
    test_data <- data[test_indices, ]
    
    for (rng_state in 1:n_rng_states) {
      set.seed(1000 * rep + 100 * fold + rng_state)
      rf <- randomForest(Group ~ ., data = train_data, ntree = 500)
      
      # Calculate accuracy on training data
      train_pred <- predict(rf, train_data)
      train_accuracy <- sum(train_pred == train_data$Group) / nrow(train_data)
      
      # Calculate accuracy on test data
      test_pred <- predict(rf, test_data)
      test_accuracy <- sum(test_pred == test_data$Group) / nrow(test_data)
      
      # Add to accuracy dataframe
      accuracy_df <- rbind(accuracy_df, 
                           data.frame(rep = rep, 
                                      fold = fold, 
                                      rng_state = rng_state, 
                                      data_type = "Train",
                                      accuracy = train_accuracy),
                           data.frame(rep = rep, 
                                      fold = fold, 
                                      rng_state = rng_state, 
                                      data_type = "Test",
                                      accuracy = test_accuracy))
      
      # Capture feature importance
      imp <- as.data.frame(importance(rf))
      imp$feature <- row.names(imp)
      imp$rep <- rep
      imp$fold <- fold
      imp$rng_state <- rng_state
      
      # Reshape for tidy data and bind to the main importance dataframe
      imp_long <- pivot_longer(imp, 
                               cols = -c(feature, rep, fold, rng_state), 
                               names_to = "metric", 
                               values_to = "importance")
      
      # We only need the MeanDecreaseGini (or MeanDecreaseAccuracy) column here
      imp_filtered <- imp_long %>% 
                      filter(metric == "MeanDecreaseGini") %>% 
                      select(-metric)
      
      importance_df <- rbind(importance_df, imp_filtered)
    }
  }
}

# Clean the importance_df to have columns as specified
importance_df <- importance_df %>% 
                 select(feature, rep, fold, rng_state, importance)

# Print the data frames to check
print(head(accuracy_df))
print(head(importance_df))

# save(accuracy_df, importance_df, file = "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.RData")

# test_accuracy_df <- accuracy_df
# test_importance_df <- importance_df
```

```{r}

accuracy_df_w <- accuracy_df %>%
             pivot_wider(names_from = data_type, values_from = accuracy, 
                         names_prefix = "accuracy_") %>%
             mutate(sqrt_product_accuracy = sqrt(accuracy_Train * accuracy_Test)) %>%
             select(rep, fold, rng_state, sqrt_product_accuracy)


joined_df <- left_join(importance_df, accuracy_df_w, by = c("rep", "fold", "rng_state"))

# Calculate sqrt_product_accuracy weighted importance
joined_df <- joined_df %>%
             mutate(weighted_importance = sqrt_product_accuracy * importance)

# Group by feature and calculate the average of the weighted importances
average_importance_df <- joined_df %>%
                         group_by(feature) %>%
                         summarise(average_weighted_importance = mean(weighted_importance, na.rm = TRUE)) %>%
                         arrange(desc(average_weighted_importance))

# report the median and second and third quartiles

average_accuracy <- accuracy_df %>% group_by(data_type) %>% summarise(average_accuracy = median(accuracy, na.rm = TRUE), second_quantile = quantile(accuracy, 0.25, na.rm = TRUE), third_quantile = quantile(accuracy, 0.75, na.rm = TRUE))
View(average_accuracy)


# accuracy_df %>% filter(data_type == 'Test') %>% ggplot(aes(x = accuracy)) + geom_histogram(bins = 20)

# View the result
print(head(average_importance_df, 10))
save(average_importance_df, file = "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.final.RData")
write_csv(average_importance_df, '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.feature_importance.csv')
```

## all countries, all groups

### Alpha diversity - metaphlan

Alpha diversity - all countries, all groups

```{r alpha_diversity_all_countries_all_groups, error=FALSE}
all_countries_all_groups_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, metadata, countries_of_interest = c('Bangladesh', 'Malawi', 'Nepal'), groups_of_interest = c('Acute typhoid', 'High Vi-titre', 'Household contact'), comparisons = list(c('Acute typhoid', 'High Vi-titre'), c('Acute typhoid', 'Household contact'), c('Household contact', 'High Vi-titre')), participant_group_colours)
all_countries_all_groups_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_all_groups_alpha$alpha_plot_group
all_countries_all_groups_alpha$alpha_plot_antibiotics
all_countries_all_groups_alpha$alpha_by_country
```

### Beta diversity - metaphlan

All groups.
```{r beta_diversity_all_groups, error=FALSE}
all_countries_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Bangladesh', 'Malawi', 'Nepal'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
all_countries_beta_all_groups$pn_res %>% kbl %>% kable_styling()

bgd_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Bangladesh'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
bgd_beta_all_groups$pn_res %>% kbl %>% kable_styling()
# bgd_beta_all_groups$pc12
# bgd_beta_all_groups$pc34

mal_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Malawi'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
mal_beta_all_groups$pn_res %>% kbl %>% kable_styling()
mal_beta_all_groups$pc12
# mal_beta_all_groups$pc34

nep_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, metadata, c('Nepal'), c('Acute typhoid', 'High Vi-titre', 'Household contact'), participant_group_colours)
nep_beta_all_groups$pn_res %>% kbl %>% kable_styling()

all_countries_beta_all_groups$pc12 / bgd_beta_all_groups$pc12 / nep_beta_all_groups$pc12

```

### P. copri clades

Print the results for all the P. copri clades

```{r print_pcopri_results, error=FALSE, eval = FALSE}
# all_taxa_maaslin <- combined_maaslins$all_features
# p_copri_maaslin <- all_taxa_maaslin %>% filter(grepl('_Prevotella_copri_', feature)) %>% filter(qval_bang < 0.05 | qval_mal < 0.05)
# write_csv(p_copri_maaslin, file.path(maaslin_taxonomic_output_root_folder, 'combined_maaslins_p_copri.csv'))
```

