---
title: "STRATAA Microbiome"
author: "Philip Ashton"
date: "`r Sys.Date()`"
output: 
  html_document:
      toc: true
      toc_float: true
      theme: united
      number_sections: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)

```

# STRATAA microbiome analysis

## Imports
```{r}
library(rbiom)
library(kableExtra)
library(ggplot2)
library(ggpubr)
library(ggforce)
library(patchwork)
library(vegan)
library(dplyr)
library(forcats)

```


## Sources

The file handles are set in config.R as they're used by both this script and data_cleaning.

```{r source}
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/core_functions.R")
source("/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_Leo/Leonardos_analysis/bin/config.R")
```


## Metadata basics

First do the plots with kruskall wallis comparison between groups to see if there's an overall difference, then do the pairwise tests because KW says there is a difference.


```{r metadata_read_in}
metadata <- read_metadata(metadata_handle)

metadata <- metadata %>% mutate(Group = if_else(Group == 'Control', 'Control', Group))
metadata <- metadata %>% mutate(Group = if_else(Group == 'Acute typhoid', 'Acute typhoid', Group))

```

```{r metadata_basics}

number_per_country <- metadata %>% group_by(Country) %>% summarise(count = n())

number_per_country %>% kbl() %>% kable_styling()

print(sum(number_per_country$count))

metadata %>% group_by(Country, Group) %>% summarise(count = n()) %>% kbl() %>% kable_styling()

metadata %>% group_by(Group) %>% summarise(count = n(), median_age = median(Age)) %>% kbl() %>% kable_styling()
baseline_chars <- get_baseline_characteristics(metadata)
comparisons_groups <- list(c("Acute typhoid", "Carrier"), c("Acute typhoid", "Control"), c("Carrier", "Control"))

ggboxplot(metadata, y = "Age", x = "Group", color = "Group") + 
  stat_compare_means() + 
  stat_compare_means(comparisons = comparisons_groups) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rremove("x.ticks") # +rotate_x_text(angle = 45)


# baseline_chars$pct_anti
# baseline_chars$pct_anti %>% na_if(0)


# BEWARE - na_if(0.0) will convert ALL 0s to NAs, this is ok as they're only in the antibiotics of carriers and controls
# at the moment, but if others get added in, will screw with it.
# baseline_chars %>% rename(`Median age` = median_age, `Women (%)` = pct_fem, `Antibiotics in last 2 weeks (%)` = pct_anti) %>% pivot_longer(!c(Country, Group)) %>% rename(variable_name = name) %>% pivot_wider(names_from = Group, values_from = value) %>% filter(variable_name != 'number') %>% na_if(0.0) %>% kbl(digits = c(NA, NA, 1, 1, 1)) %>% kable_styling()

# na_if stopped working, so had to do this instead
baseline_chars_table <- baseline_chars %>% rename(`Median age` = median_age, `Women (%)` = pct_fem, `Antibiotics in last 2 weeks (%)` = pct_anti) %>% pivot_longer(!c(Country, Group)) %>% rename(variable_name = name) %>% pivot_wider(names_from = Country, values_from = value)

# baseline_chars_table$Carrier <- replace(baseline_chars_table$Carrier,which(baseline_chars_table$Carrier==0),NA)
# baseline_chars_table$`Control` <- replace(baseline_chars_table$`Control`,which(baseline_chars_table$`Control`==0),NA)
baseline_chars_table %>% kbl(digits = c(NA, NA, 1, 1, 1)) %>% kable_styling()

# the kruskal wallis test is positive, showing a difference between the groups, so we move onto pairwise tests.
# ggplot(metadata, aes(x = Country, y = Age, fill = Group)) + geom_boxplot() + stat_compare_means(method = 'kruskal.test', label = "p")
comparisons_countries <- list(c('Bangladesh', 'Malawi'), c('Bangladesh', 'Nepal'), c('Malawi', 'Nepal'))
# default stat method when doing pairwise tests in wilcoxon.
metadata_for_plotting <- metadata
metadata_for_plotting$Group <- factor(metadata_for_plotting$Group, levels = c("Control", "Acute typhoid", "Carrier"))
ggboxplot(metadata_for_plotting, facet.by = "Country", y = "Age", x = "Group", color = "Group") + stat_compare_means(comparisons = comparisons_groups) + rremove("x.text") + rremove("xlab") + rremove("x.ticks") # +rotate_x_text(angle = 45)
ggboxplot(metadata_for_plotting, facet.by = "Group", y = "Age", x = "Country", color = "Country") + stat_compare_means(comparisons = comparisons_countries) + rotate_x_text(angle = 45)

```

```{r metadata_basic_plots, fig.width = 4, fig.height= 8}
country_group_sex <- metadata %>% group_by(Country, Group, Sex) %>% summarise(count = n()) 
comparisons_groups <- list(c("Acute typhoid", "Carrier"), c("Acute typhoid", "Control"), c("Carrier", "Control"))
country_group_sex %>% group_by(Country, Sex) %>% summarise(n = sum(count))
plot_sex <- function(eg1, c){
  d <- eg1 %>% filter(Country == c)
  p <- ggplot(d, aes(x = Group, y = count, fill = Sex)) + 
    geom_bar(stat ='identity', position = 'fill') + 
    ylab('Proportion') + 
    ggtitle(c)
    # +
    #theme(axis.text=element_text(size=34), axis.title=element_text(size=36,face="bold"), plot.title = element_text(size = 40, face = "bold"), legend.key.size = unit(4, 'cm'), legend.title = element_text(size = 34), legend.text = element_text(size = 28))
  
  return(p)
}

m_sex <- plot_sex(country_group_sex, 'Malawi')
b_sex <- plot_sex(country_group_sex, 'Bangladesh')
n_sex <- plot_sex(country_group_sex, 'Nepal')
m_sex / b_sex / n_sex 
```

## sequencing read stats

```{r sequencing_read_stats}

read_stats_path <- '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/data_munging/2023.08.28/2023.08.28 strataa microbiome number of reads.csv'
read_stats <- read_csv(read_stats_path) %>% filter(Country != '#N/A')

ggboxplot(read_stats, facet.by = "Country", x = "Group", y = "num_seqs", color = "Group") + rremove("x.text") + rremove("xlab") + rremove("x.ticks") + rotate_x_text(angle = 45) + stat_compare_means()

```

## read in metaphlan data
```{r read_in_metaphlan}
strataa_metaphlan_data <- read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.all_strataa_metaphlan.csv'), header= TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE, check.names=FALSE)
strataa_metaphlan_data$lowest_taxonomic_level <- sapply(str_split(row.names(strataa_metaphlan_data), "\\|"), function(x) x[length(x)])
strataa_metaphlan_data_species <- strataa_metaphlan_data %>% filter(str_starts(lowest_taxonomic_level, 's__'))

strataa_metaphlan_metadata <- read.csv(file = file.path(metaphlan_input_folder, '2023.05.11.strataa_metadata.metaphlan.csv'), header = TRUE, sep = ",", row.names = 1, stringsAsFactors = FALSE)
strataa_metaphlan_metadata <- strataa_metaphlan_metadata %>% mutate(Group = if_else(Group == 'Control', 'Control', Group)) %>% mutate(Group = if_else(Group == 'Acute typhoid', 'Acute typhoid', Group))
strataa_metaphlan_metadata <- strataa_metaphlan_metadata %>% mutate(SampleID = rownames(strataa_metaphlan_metadata))
strataa_metaphlan_metadata <- strataa_metaphlan_metadata %>% mutate(age_bracket=cut(Age, breaks=c(0, 1, 5, 15, Inf), labels=c("<1", "1-5", "6-15", ">15")))
```

## all countries, all groups

### Alpha diversity - metaphlan

Alpha diversity - all countries, all groups

```{r alpha_diversity_all_countries_all_groups}
all_countries_all_groups_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, strataa_metaphlan_metadata, countries_of_interest = c('Bangladesh', 'Malawi', 'Nepal'), groups_of_interest = c('Acute typhoid', 'Carrier', 'Control'), comparisons = list(c('Acute typhoid', 'Carrier'), c('Acute typhoid', 'Control'), c('Control', 'Carrier')))
all_countries_all_groups_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_all_groups_alpha$alpha_plot_group
all_countries_all_groups_alpha$alpha_plot_antibiotics
all_countries_all_groups_alpha$alpha_by_country

```
### Beta diversity - metaphlan

All groups.
```{r beta_diversity_all_groups}
all_countries_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Bangladesh', 'Malawi', 'Nepal'), c('Acute typhoid', 'Carrier', 'Control'))
all_countries_beta_all_groups$pn_res %>% kbl %>% kable_styling()

bgd_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Bangladesh'), c('Acute typhoid', 'Carrier', 'Control'))
bgd_beta_all_groups$pn_res %>% kbl %>% kable_styling()
# bgd_beta_all_groups$pc12
# bgd_beta_all_groups$pc34

mal_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi'), c('Acute typhoid', 'Carrier', 'Control'))
mal_beta_all_groups$pn_res %>% kbl %>% kable_styling()
# mal_beta_all_groups$pc12
# mal_beta_all_groups$pc34

nep_beta_all_groups <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Nepal'), c('Acute typhoid', 'Carrier', 'Control'))
nep_beta_all_groups$pn_res %>% kbl %>% kable_styling()

all_countries_beta_all_groups$pc12 / bgd_beta_all_groups$pc12 / nep_beta_all_groups$pc12

```

## Healthy vs Typhi

### Plot phyla bar plots

```{r phyla_barplots}


phyla <- strataa_metaphlan_data %>% mutate(clade_name = rownames(strataa_metaphlan_data)) %>% filter(grepl("p__", clade_name)) %>% filter(!grepl("c__", clade_name)) %>% pivot_longer(!c(clade_name, lowest_taxonomic_level), names_to = "sample", values_to = "relative_abundance")
metadata_for_phyla_plots <- strataa_metaphlan_metadata %>% dplyr::select(SampleID, Group, Country)

phyla_clean_metadata <- prep_data_to_plot_phyla(phyla, strataa_metaphlan_metadata)

order_of_groups <- c("Typhi", "Healthy")
bangladesh_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Bangladesh", group_order = order_of_groups)
# bangladesh_phyla_plot
malawi_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Malawi", group_order = order_of_groups)
nepal_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Nepal", group_order = order_of_groups)

bangladesh_phyla_plot / malawi_phyla_plot / nepal_phyla_plot
```

```{r phyla_table}

# going back to phyla so that get all the weird rare ones.
# need to join to meta-data
# remove carriers
# group by clade, group, country
# summarise to get the median.
# pivot wider to get the table.
# dplyr::select(order(colnames(.))) rearranges the columns in alphabetical order
# relocate moves the clade_name column to the first column.
phyla %>% left_join(metadata_for_phyla_plots, by = c("sample" = "SampleID")) %>% filter(Group != 'Carrier') %>% group_by(clade_name, Group, Country) %>% summarise(median_prevalence = median(relative_abundance)) %>% pivot_wider(names_from = c('Country', 'Group'), values_from =  'median_prevalence') %>% dplyr::select(order(colnames(.))) %>% relocate(clade_name, .before = 1) %>% kbl() %>% kable_styling()


```

### alpha diversity
Alpha diversity - all countries, healthy and acute

```{r alpha_diversity_healthy_acute}

all_countries_healthy_acute_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, strataa_metaphlan_metadata, countries_of_interest = c('Bangladesh', 'Malawi', 'Nepal'), groups_of_interest = c('Acute typhoid', 'Control'), comparisons = list(c('Acute typhoid', 'Control')))
# all_countries_healthy_acute_alpha$alpha_by_country
all_countries_healthy_acute_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_acute_alpha$alpha_plot_group
all_countries_healthy_acute_alpha$alpha_plot_antibiotics

```

alpha diversity, malawi only

```{r alpha_diversity_malawi}
malawi_healthy_acute_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, strataa_metaphlan_metadata, countries_of_interest = c('Malawi'), groups_of_interest = c('Acute typhoid', 'Control'), comparisons = list(c('Acute typhoid', 'Control')))
malawi_healthy_acute_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
malawi_healthy_acute_alpha$alpha_plot_group
malawi_healthy_acute_alpha$alpha_plot_antibiotics
```

### beta diversity 

Acute vs healthy.
```{r beta_diversity_healthy_acute}

all_countries_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Bangladesh', 'Malawi', 'Nepal'), c('Acute typhoid', 'Control'))
all_countries_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

bgd_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Bangladesh'), c('Acute typhoid', 'Control'))
bgd_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi'), c('Acute typhoid', 'Control'))
mal_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

nep_beta_acute_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Nepal'), c('Acute typhoid', 'Control'))
nep_beta_acute_healthy$pn_res %>% kbl %>% kable_styling()

all_countries_beta_acute_healthy$pc12 + bgd_beta_acute_healthy$pc12 + mal_beta_acute_healthy$pc12 + nep_beta_acute_healthy$pc12 + plot_layout(guides = 'collect')
```

### maaslin2 taxonomy
Maaslin basics

```{r basics_for_maaslin_taxonomic}

groups_to_analyse <- c('Acute typhoid', 'Control')
bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")

```

```{r read_in_maaslin_taxonomic}
bangladesh_taxonomic_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'metaphlan')
malawi_taxonomic_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'metaphlan')
nepal_taxonomic_maaslin <- read_in_maaslin('Nepal', groups_to_analyse, nep_variables_for_analysis, 'metaphlan')
```

```{r filter_maaslin_taxonomic}
bangladesh_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(bangladesh_taxonomic_maaslin)
malawi_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(malawi_taxonomic_maaslin)
nepal_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(nepal_taxonomic_maaslin)
```

```{r basic_stats_maaslin_taxonomic}
bangladesh_maaslin_stats <- basic_maaslin_stats(bangladesh_taxonomic_maaslin_filtered, 'Bangladesh', bang_variables_for_analysis, groups_to_analyse)
malawi_maaslin_stats <- basic_maaslin_stats(malawi_taxonomic_maaslin_filtered, 'Malawi', mwi_variables_for_analysis, groups_to_analyse)
nepal_maaslin_stats <- basic_maaslin_stats(nepal_taxonomic_maaslin_filtered, 'Nepal', nep_variables_for_analysis, groups_to_analyse)
```

```{r volcano_plots_maaslin_taxonomic}

malawi_maaslin_stats$volcano_plot / bangladesh_maaslin_stats$volcano_plot / nepal_maaslin_stats$volcano_plot
nrow(malawi_maaslin_stats$maaslin_results_sig)
nrow(bangladesh_maaslin_stats$maaslin_results_sig)
nrow(nepal_maaslin_stats$maaslin_results_sig)
```

There were `r nrow(malawi_maaslin_stats$maaslin_results_sig)` species significantly (q-val < 0.05) associated with health/disease in Malawi, `r nrow(bangladesh_maaslin_stats$maaslin_results_species_group_sig)` in Bangladesh, and `r nrow(nepal_maaslin_stats$maaslin_results_species_group_sig)` in Nepal.

Combine the taxonomic maaslins, and print out the species that are sig in both and share directions.

Because sequencing run and participant type are totally confounded for Bangladesh, need to exclude sequencing run from the final model for Bangladesh (otherwise, wipes out the signals).

associated at both sites
```{r combine_maaslins_taxonomic}

bang_mal <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered)
combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# View(combined_results$positive_coef)
# View(combined_results$negative_coef)


combined_results$positive_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

combined_results$negative_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

# todo - refactoring the run_combine_maaslins means that we dont get the species that are only associated at one site. need to fix that.

# nrow(combined_results$mwi_maaslin_only)
# nrow(combined_results$bang_maaslin_only)
```

There were `r nrow(combined_results$mwi_maaslin_only)` species significantly (q-val < 0.05) associated with health/disease in Malawi only and `r nrow(combined_results$bang_maaslin_only)` in Bangladesh only.

The ones associated at only one site are written out to a file, you can look at them manually there.


### plots of species of interest

```{r plot_species_of_interest_taxonomic}

strataa_metaphlan_data_longer <- strataa_metaphlan_data %>% mutate(feature = rownames(strataa_metaphlan_data)) %>% pivot_longer(!c(feature, lowest_taxonomic_level), names_to = "SampleID", values_to = "prevalence")
strataa_metaphlan_data_longer_meta <- strataa_metaphlan_data_longer %>% left_join(strataa_metaphlan_metadata, by = c("SampleID" = "SampleID"))


pc <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Prevotella_copri_clade_A')
cs <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Clostridium_SGB6179')
SGB5809 <-run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__GGB4266_SGB5809')
hp <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Haemophilus_parainfluenzae')
rt <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Romboutsia_timonensis')
lb <- run_plot_species_of_interest(strataa_metaphlan_data_longer_meta, 's__Lachnospiraceae_bacterium')

pc
  # pc + cs + SGB5809 + hp + rt + lb

```

### P. copri clades

Print the results for all the P. copri clades

```{r, eval = FALSE}
# all_taxa_maaslin <- combined_maaslins$all_features
# p_copri_maaslin <- all_taxa_maaslin %>% filter(grepl('_Prevotella_copri_', feature)) %>% filter(qval_bang < 0.05 | qval_mal < 0.05)
# write_csv(p_copri_maaslin, file.path(maaslin_taxonomic_output_root_folder, 'combined_maaslins_p_copri.csv'))
```



### maaslin functional groups 

Functional groups associated with health

```{r read_in_maaslin_functional}

bang_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
# nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
groups_to_analyse <- c('Acute typhoid', 'Control')

bang_func_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'bigmap')
mal_func_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'bigmap')
```

```{r maaslin_functional_groups}

# combined_results <- run_combine_maaslins(bang_func_maaslin, mal_func_maaslin, bang_variables_for_analysis, mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)

bang_mal <- list(bang_func_maaslin, mal_func_maaslin)
combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)


combined_results$positive_coef %>%
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

combined_results$negative_coef %>%
  select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
  rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
  dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
  kbl() %>% kable_styling()

combined_results$positive_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()
combined_results$negative_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()

```


## healthy vs carrier
### phylum plots

```{r phylum_plots_carrier_healthy}
order_of_groups <- c("Carrier", "Healthy")
# bangladesh_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Bangladesh", group_order = order_of_groups)
# bangladesh_phyla_plot
malawi_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Malawi", group_order = order_of_groups)
nepal_phyla_plot <- plot_per_country_abundance(phyla_clean_metadata = phyla_clean_metadata, country = "Nepal", group_order = order_of_groups)

# bangladesh_phyla_plot / 
malawi_phyla_plot / nepal_phyla_plot


```

### alpha diversity

Alpha diversity - all countries, healthy and carrier

```{r alpha_diversity_healthy_carrier}

all_countries_healthy_carrier_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, strataa_metaphlan_metadata, countries_of_interest = c('Malawi', 'Nepal'), groups_of_interest = c('Carrier', 'Control'), comparisons = list(c('Carrier', 'Control')))
all_countries_healthy_carrier_alpha$alpha_by_country
all_countries_healthy_carrier_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_carrier_alpha$alpha_plot_group
```

### beta diversity

Carrier vs healthy.
```{r beta_diversity_healthy_carrier}

all_countries_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi', 'Nepal'), c('Carrier', 'Control'))
all_countries_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

# bgd_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Bangladesh'), c('Carrier', 'Control'))
# bgd_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi'), c('Carrier', 'Control'))
mal_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

nep_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Nepal'), c('Carrier', 'Control'))
nep_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_carrier_healthy$pc12 + nep_beta_carrier_healthy$pc12 + plot_layout(guides = 'collect')
```

### maaslin taxonomic groups

```{r basics_for_maaslin_taxonomic_carrier_healthy}

groups_to_analyse <- c('Carrier', 'Control')
# bang_variables_for_analysis <- c("Group", "Sex", "Age")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
nep_variables_for_analysis <- c("Group", "Sex", "Age")

```

```{r read_in_maaslin_taxonomic_carrier_healthy}
# bangladesh_taxonomic_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'metaphlan')
malawi_taxonomic_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'metaphlan')
nepal_taxonomic_maaslin <- read_in_maaslin('Nepal', groups_to_analyse, nep_variables_for_analysis, 'metaphlan')
```

```{r filter_maaslin_taxonomic_carrier_healthy}
# bangladesh_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(bangladesh_taxonomic_maaslin)
malawi_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(malawi_taxonomic_maaslin)
nepal_taxonomic_maaslin_filtered <- filter_taxonomic_maaslin(nepal_taxonomic_maaslin)
```

```{r basic_stats_maaslin_taxonomic_carrier_healthy}
# bangladesh_maaslin_stats <- basic_maaslin_stats(bangladesh_taxonomic_maaslin_filtered, 'Bangladesh', bang_variables_for_analysis, groups_to_analyse)
malawi_maaslin_stats <- basic_maaslin_stats(malawi_taxonomic_maaslin_filtered, 'Malawi', mwi_variables_for_analysis, groups_to_analyse)
nepal_maaslin_stats <- basic_maaslin_stats(nepal_taxonomic_maaslin_filtered, 'Nepal', nep_variables_for_analysis, groups_to_analyse)
```

```{r volcano_plots_maaslin_taxonomic_carrier_healthy}
# / bangladesh_maaslin_stats$volcano_plot 
malawi_maaslin_stats$volcano_plot / nepal_maaslin_stats$volcano_plot
nrow(malawi_maaslin_stats$maaslin_results_sig)
# nrow(bangladesh_maaslin_stats$maaslin_results_sig)
nrow(nepal_maaslin_stats$maaslin_results_sig)
```

We're not including the bangladesh samples in this analysis, because the bangladesh carriers were processed differently (extracted without being frozen).

There are no species significantly associated with carrier status in all Mal and Nep, so not including the combine.

```{r combine_maaslins_taxonomic_carrier_healthy}

# combined_results <- run_combine_maaslins(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered, bang_variables_for_analysis, mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# only one species significantly associated in Nepal.
# there are no species consistently associated across all three sites
# bang_mal_nep <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_mal_nep, c('_bang', '_mwi', '_nep'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# there are no species associated in both bang and nep
# bang_nep <- list(bangladesh_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_nep, c('_bang', '_nep'), bang_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# there are no species associated in both mal and nep
# mal_nep <- list(malawi_taxonomic_maaslin_filtered, nepal_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_nep, c('_mal', '_nep'), bang_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)
# View(combined_results$positive_coef)
# View(combined_results$negative_coef)

# bang_mal <- list(bangladesh_taxonomic_maaslin_filtered, malawi_taxonomic_maaslin_filtered)
# combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'metaphlan', maaslin_taxonomic_output_root_folder)

# # positive is associated with health
# combined_results$positive_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()

# # negative is associated with carrier
# combined_results$negative_coef %>% filter(grepl('^s', lowest_taxonomic_level)) %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(Species = lowest_taxonomic_level, `Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()

# nrow(combined_results$mwi_maaslin_only)
# nrow(combined_results$bang_maaslin_only)
```


### functional groups

```{r read_in_maaslin_functional_healthy_carrier}
# bang_variables_for_analysis <- c("Group", "Sex", "Age")
mwi_variables_for_analysis <- c("Group", "Sex", "Age", "sequencing_lane")
# nep_variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
groups_to_analyse <- c('Carrier', 'Control')

# bang_func_carr_health_maaslin <- read_in_maaslin('Bangladesh', groups_to_analyse, bang_variables_for_analysis, 'bigmap')
mal_func_carr_health_maaslin <- read_in_maaslin('Malawi', groups_to_analyse, mwi_variables_for_analysis, 'bigmap')
```


```{r maaslin_functional_groups_healthy_carrier}
# combined_results <- run_combine_maaslins(bang_func_carr_health_maaslin, mal_func_carr_health_maaslin, bang_variables_for_analysis, mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)

# bang_mal <- list(bang_func_carr_health_maaslin, mal_func_carr_health_maaslin)
# combined_results <- run_combine_maaslins(bang_mal, c('_bang', '_mal'), mwi_variables_for_analysis, groups_to_analyse, 'bigmap', maaslin_functional_output_root_folder)

# # positive is associated with health
# # combined_results$positive_coef %>%  kbl() %>% kable_styling()
# combined_results$positive_coef %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()
# # negative is associated with carrier
# # combined_results$negative_coef %>%  kbl() %>% kable_styling()
# combined_results$negative_coef %>% 
#   select(!c(metadata, value, N_bang, N.not.0_bang, pval_bang, N_mal, N.not.0_mal, pval_mal)) %>% 
#   rename(`Coefficient Bangladesh` = coef_bang, `Standard Error Bangladesh` = stderr_bang, `Q-value Bangladesh` = qval_bang, `Coefficient Malawi` = coef_mal, `Standard Error Malawi` = stderr_mal, `Q-value Malawi` = qval_mal) %>% 
#   dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% 
#   kbl() %>% kable_styling()


# # print out a nicely formatted table of the count of the MGCs
# combined_results$positive_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()
# combined_results$negative_coef %>% mutate(split_species = str_split(Species, '_')) %>% mutate(genus = sapply(split_species, "[[", 1)) %>% mutate(specific = sapply(split_species, "[[", 2)) %>% mutate(clean_species = paste(genus, specific, sep = ' ')) %>% select(-split_species, -genus, -specific) %>% relocate(clean_species, .after = Species) %>% group_by(MGC_class) %>% arrange(Species) %>% summarise(n = n(), Species = paste(clean_species, collapse = ', ')) %>% arrange(desc(n), MGC_class) %>%  kbl() %>% kable_styling()
```

## healthy vs acute vs carriers 

### alpha

```{r alpha_diversity_healthy_acute_carrier}

all_countries_healthy_acute_carrier_alpha <- metaphlan_alpha(strataa_metaphlan_data_species, strataa_metaphlan_metadata, countries_of_interest = c('Malawi', 'Nepal'), groups_of_interest = c('Carrier', 'Control', 'Acute typhoid'), comparisons = list(c('Acute typhoid', 'Carrier'), c('Acute typhoid', 'Control'), c('Carrier', 'Control')))

all_countries_healthy_acute_carrier_alpha$alpha_by_country
all_countries_healthy_acute_carrier_alpha$alpha_anova_summary_with_var_names %>% dplyr::mutate_if(is.numeric, funs(as.character(signif(., 3)))) %>% kbl() %>% kable_styling()
all_countries_healthy_acute_carrier_alpha$alpha_plot_group
```

### beta

```{r beta_diversity_healthy_acute_carrier}

all_countries_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi', 'Nepal'), c('Acute typhoid', 'Carrier', 'Control'))
all_countries_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

# bgd_beta_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data, strataa_metaphlan_metadata, c('Bangladesh'), c('Carrier', 'Control'))
# bgd_beta_carrier_healthy$pn_res %>% kbl %>% kable_styling()

mal_beta_acute_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Malawi'), c('Acute typhoid', 'Carrier', 'Control'))
mal_beta_acute_carrier_healthy$pn_res %>% kbl %>% kable_styling()
mal_beta_acute_carrier_healthy$pc12
# mal_beta_acute_carrier_healthy$pc34

nep_beta_acute_carrier_healthy <- strataa_metaphlan_beta(strataa_metaphlan_data_species, strataa_metaphlan_metadata, c('Nepal'), c('Acute typhoid', 'Carrier', 'Control'))
nep_beta_acute_carrier_healthy$pn_res %>% kbl %>% kable_styling()
nep_beta_acute_carrier_healthy$pc12
# nep_beta_acute_carrier_healthy$pc34

mal_beta_acute_carrier_healthy$pc12 / nep_beta_acute_carrier_healthy$pc12  + plot_layout(guides = 'collect')


### maaslin2 taxonomic

```{r read_in_heal_ac_car_taxonomic}
groups_to_analyse <- c('Carrier', 'Control', 'Acute typhoid')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions")
heal_ac_car_taxonomic_np <- read_in_maaslin('Nepal', groups_to_analyse, variables_for_analysis, 'metaphlan')
heal_ac_car_taxonomic_np_sig <- filter_taxonomic_maaslin(heal_ac_car_taxonomic_np) %>% filter(qval <= 0.05) 
print(nrow(heal_ac_car_taxonomic_np_sig))
```


```{r heal_ac_car_taxonomic}
groups_to_analyse <- c('Carrier', 'Control', 'Acute typhoid')
variables_for_analysis <- c("Group", "Sex", "Age", "Antibiotics_taken_before_sampling_yes_no_assumptions", "sequencing_lane")
heal_ac_car_taxonomic <- read_in_maaslin('Malawi', groups_to_analyse, variables_for_analysis, 'metaphlan')
heal_ac_car_taxonomic_cln <- filter_taxonomic_maaslin(heal_ac_car_taxonomic) #%>% filter(qval <= 0.05) 
# spread heal_ac_car_taxonomic_cln wider so that the coef is on the same row for each feature
heal_ac_car_taxonomic_cln_w <- heal_ac_car_taxonomic_cln %>% pivot_wider(names_from = value, values_from = c(coef, qval), id_cols = c(feature, lowest_taxonomic_level))

heal_ac_car_taxonomic_mwi_sig <- filter_taxonomic_maaslin(heal_ac_car_taxonomic) %>% filter(qval <= 0.05) 
print(nrow(heal_ac_car_taxonomic_mwi_sig))
```
changes relative to healthy state
```{r maaslin_heal_ac_car_taxonomic}
# same_same cant do from here because filtered for qvalue
# 1
up_up <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & coef_Carrier > 0 & qval_Acute typhoid < 0.05 & qval_Carrier < 0.05) %>% mutate(type = '1.up_up')
print(nrow(up_up))
# write_csv(up_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.1.up_up.csv'))
# 2
up_down <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & coef_Carrier < 0 & qval_Acute typhoid < 0.05 & qval_Carrier < 0.05) %>% mutate(type = '2.up_down')
print(nrow(up_down))
# write_csv(up_down, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.2.up_down.csv'))
# 3
up_same <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid > 0 & qval_Acute typhoid < 0.05 & qval_Carrier >= 0.05) %>% mutate(type = '3.up_same')
print(nrow(up_same))
# write_csv(up_same, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.3.up_same.csv'))
# 4
down_up <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & coef_Carrier > 0 & qval_Acute typhoid < 0.05 & qval_Carrier < 0.05) %>% mutate(type = '4.down_up')
print(nrow(down_up))
# write_csv(down_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.4.down_up.csv'))
# 5
down_down <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & coef_Carrier < 0 & qval_Acute typhoid < 0.05 & qval_Carrier < 0.05) %>% mutate(type = '5.down_down')
print(nrow(down_down))
# write_csv(down_down, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.5.down_down.csv'))
# 6
down_same <- heal_ac_car_taxonomic_cln_w %>% filter(coef_Acute typhoid < 0 & qval_Acute typhoid < 0.05 & qval_Carrier > 0.05) %>% mutate(type = '6.down_same')
print(nrow(down_same))
# write_csv(down_same, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.6.down_same.csv'))
# 7
same_up <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid > 0.05 & qval_Carrier < 0.05 & coef_Carrier > 0) %>% mutate(type = '7.same_up')
print(nrow(same_up))
# write_csv(same_up, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.7.same_up.csv'))
# 8
same_down <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid > 0.05 & qval_Carrier < 0.05 & coef_Carrier < 0) %>% mutate(type = '8.same_down')
print(nrow(same_down))
# 9
same_same <- heal_ac_car_taxonomic_cln_w %>% filter(qval_Acute typhoid >= 0.05 & qval_Carrier >= 0.05) %>% mutate(type = '9.same_same')
print(nrow(same_same))

all_patterns <- rbind(up_up, up_down, up_same, down_up, down_down, down_same, same_up, same_down, same_same)

write_csv(all_patterns, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.all_patterns.csv')) 

up_in_ac_car <- all_patterns %>% filter(type %in% c('1.up_up', '2.up_down', '3.up_same', '4.down_up', '7.same_up')) 
down_in_ac_car <- all_patterns %>% filter(type %in% c('5.down_down', '6.down_same', '8.same_down'))
up_in_health_only <- all_patterns %>% filter(type == '5.down_down')
up_in_car_only <- all_patterns %>% filter(type == '7.same_up')

prop_that_is_sgb <- function(input_data) {
  x <- input_data %>% filter(str_detect(feature, 'SGB')) %>% summarise(prop = n()/nrow(input_data), n_sgb = n(), total_n = nrow(input_data))
  print(x)
  return(x)
}
prop_up_in_ac_car <- prop_that_is_sgb(up_in_ac_car)
prop_down_in_ac_car <- prop_that_is_sgb(down_in_ac_car)
prop_up_in_health_only <- prop_that_is_sgb(up_in_health_only)
prop_up_in_car_only <- prop_that_is_sgb(up_in_car_only)


prop.test(x = c(prop_up_in_ac_car$n_sgb, prop_down_in_ac_car$n_sgb), n = c(prop_up_in_ac_car$total_n, prop_down_in_ac_car$total_n), alternative = 'two.sided', correct = FALSE)
prop.test(x = c(prop_up_in_health_only$n_sgb, prop_up_in_car_only$n_sgb), n = c(prop_up_in_health_only$total_n, prop_up_in_car_only$total_n), alternative = 'two.sided', correct = FALSE)
```

```{r maaslin_heal_ac_car_sgb_plot}
# make a new strataa_metaphlan_data object where the rowname column is an actual column

strataa_metaphlan_data_cln <- strataa_metaphlan_data %>% mutate(feature = rownames(strataa_metaphlan_data)) 
rownames(strataa_metaphlan_data_cln) <- NULL
strataa_metaphlan_data_l <- strataa_metaphlan_data_cln %>% pivot_longer(!c(feature, lowest_taxonomic_level), names_to = 'sample', values_to = 'abundance') %>% left_join(strataa_metaphlan_metadata %>% mutate(sample = rownames(strataa_metaphlan_metadata)), by = 'sample')
# head(strataa_metaphlan_data_l)
strataa_metaphlan_data_mwi_sp <- strataa_metaphlan_data_l %>% filter(Country == 'Malawi', str_detect(feature, 's__'), str_detect(feature, 't__', negate = TRUE)) %>% mutate(SGB = str_detect(feature, '_SGB'))
head(strataa_metaphlan_data_mwi_sp)

mwi_sgb_abundance <- strataa_metaphlan_data_mwi_sp %>% group_by(sample, SGB) %>% summarise(total_abundance = sum(abundance)) %>% left_join(strataa_metaphlan_metadata %>% mutate(sample = rownames(strataa_metaphlan_metadata)), by = 'sample')  %>% mutate(SGB = ifelse(SGB, 'SGB', 'non-SGB')) %>% mutate(Group = factor(Group, levels = c('Control', 'Acute typhoid', 'Carrier')))
mwi_sgb_abundance %>% group_by(SGB, Group) %>% summarise(median_abundance = median(total_abundance), n = n()) %>% kbl() %>% kable_styling()
# mwi_sgb_abundance <- mwi_sgb_abundance 
# do a scatter plot of the abundance of sgb in each group
my_comparisons <- list(c('Acute typhoid', 'Carrier'), c('Acute typhoid', 'Control'), c('Carrier', 'Control'))
mwi_sgb_abundance_plot <- ggplot(mwi_sgb_abundance %>% filter(SGB == 'SGB'), aes(x = Group, y = total_abundance, color = Group)) +
  geom_boxplot() +
  stat_compare_means(comparisons = my_comparisons) + 
  stat_compare_means(label.y = 70) + 
  theme(legend.position = "none") + 
  labs(y = 'SGB abundance in metagenome (%)', x = '') + 
  theme(text = element_text(size=18)) + 
  scale_x_discrete(labels= c('Control', 'Acute', 'Carrier'))

mwi_sgb_abundance_plot$layers[[2]]$aes_params$textsize <- 4
# which_layers(mwi_sgb_abundance_plot, "GeomSignif")

ggsave(file.path(metaphlan_input_folder, 'mwi_sgb_abundance.png'), mwi_sgb_abundance_plot, width = 8, height = 4.8, units = 'in', dpi = 300)

```

```{r get_same_same}



# heal_ac_car_taxonomic_all_w %>% filter(is.na(qval_Carrier))
```

```{r make_health_acc_carr_table}

to_get_list <- c('k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Oscillospiraceae.g__Faecalibacterium.s__Faecalibacterium_prausnitzii', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Barnesiellaceae.g__Barnesiella.s__Barnesiella_intestinihominis', 'k__Bacteria.p__Firmicutes.c__Bacilli.o__Lactobacillales.f__Lactobacillaceae.g__Limosilactobacillus.s__Limosilactobacillus_reuteri', 'k__Bacteria.p__Proteobacteria.c__Gammaproteobacteria.o__Pasteurellales.f__Pasteurellaceae.g__Haemophilus.s__Haemophilus_parainfluenzae', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Erysipelotrichaceae.g__Holdemanella.s__Holdemanella_biformis', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Mediterraneibacter.s__Ruminococcus_gnavus', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Peptostreptococcaceae.g__Paeniclostridium.s__Eubacterium_tenue', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Turicibacteraceae.g__Turicibacter.s__Turicibacter_bilis', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Turicibacteraceae.g__Turicibacter.s__Turicibacter_sanguinis', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Odoribacteraceae.g__Odoribacter.s__Odoribacter_splanchnicus', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Blautia.s__Blautia_faecis', 'k__Bacteria.p__Proteobacteria.c__Gammaproteobacteria.o__Enterobacterales.f__Enterobacteriaceae.g__Escherichia.s__Escherichia_coli', 'k__Bacteria.p__Firmicutes.c__Clostridia.o__Eubacteriales.f__Lachnospiraceae.g__Lachnospiraceae_unclassified.s__Eubacterium_rectale', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Bacteroidaceae.g__Bacteroides.s__Bacteroides_fragilis', 'k__Bacteria.p__Firmicutes.c__Erysipelotrichia.o__Erysipelotrichales.f__Erysipelotrichaceae.g__Holdemanella.s__Holdemanella_porci', 'k__Bacteria.p__Bacteroidetes.c__Bacteroidia.o__Bacteroidales.f__Prevotellaceae.g__Prevotellamassilia.s__Prevotellamassilia_timonensis')
# this one doesn't include the Oscillospiraceae, so will just have to point towards the supplementary table for those.
heal_acc_carr_table <- all_patterns %>% filter(feature %in% to_get_list)
write_csv(heal_acc_carr_table, file.path(maaslin_taxonomic_output_root_folder, 'blantyre.health_acute_carrier.health_acc_carr_table.csv'))

```

## amr analysis

Code here is not great, maybe come back and move some into functions/utils.

```{r amr_paths}
amr_data_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/2023.08.30.strataa_microbiome.subsamp.rgi.txt"
ncbi_amr_info_path <- "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/ncbi_amr/2023.08.31/refgenes.modified.csv"

```

Need to setup a data frame with every sample/drug combination so that capture the zeroes. 

```{r read_in_amr, include=FALSE}

ncbi_amr_info <- read_csv(ncbi_amr_info_path)

amr_data <- read_delim(amr_data_path, delim = "\t") %>% separate_wider_delim(`ARO Term`, names = c("sample", "ARO Term"), delim = ".gene_mapping_data.txt:") %>% mutate(reads_per_kb = `Completely Mapped Reads` / (`Reference Length` / 1000)) 

amr_data_with_ncbi <- left_join(amr_data, ncbi_amr_info, by = join_by(`ARO Term` == combined_name))

write_csv(amr_data_with_ncbi, file.path('~/Desktop', "amr_data_with_ncbi.csv"))

# 1. set the samples and classes to be analysed
samples <- metadata %>% pull(Lane)
classes <- distinct(amr_data_with_ncbi, Class) %>% filter(!is.na(Class)) %>% pull(Class) 

# 2. make an empty data frame, with just the samples and classes
amr_summarised <- expand.grid(sample = samples, Class = classes)
# 3. add the metadata
amr_summarised_meta <- left_join(amr_summarised, metadata, by = join_by(sample == Lane))

```

```{r ncbi_amr_info}
# 4. take the "found" genes
# we filter for 100% cov here, and these are the "present" genes
amr_data_with_ncbi_cln <- amr_data_with_ncbi %>% filter(`Average Percent Coverage` == 100) %>% filter(Class != 'NA') %>% filter(Scope == 'core')

# 5. summarise the data (by class)
amr_data_with_ncbi_cln_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))
# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)
# 6. join the data to the template (amr_summarised_meta)
amr_summarised_meta_with_data <- left_join(amr_summarised_meta, amr_data_with_ncbi_cln_summarised, by = join_by(sample == sample, Class == Class)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% mutate(Class = ifelse(is.na(Class), "None", Class))

# amr_data_with_ncbi_cln_meta <- left_join(amr_data_with_ncbi_cln, metadata, by = join_by(sample == Lane))
# View(amr_summarised_meta_with_data %>% filter(Class == 'AMINOGLYCOSIDE') %>% select(sample, Class, n, Group, Country) %>% arrange(Country, Group))

```
Graphs of overall comparison. Do not include carriers here, because of different sample handling in Bangladesh.

# todo - do the overall graphs for healthy vs acute vs carrier (?)

```{r overall_graphs_per_country}
number_of_amr_genes_per_sample <- amr_summarised_meta_with_data %>% filter(!(Group == 'Carrier')) %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
number_of_amr_genes_per_sample  %>% group_by(Country) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
# number of amr genes per sample, by country
p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  ylab('Number of AMR genes identified per sample') +
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_boxplot.png"), plot = p, width = 5, height = 4)

p
# rpk amr genes per sample by country
p <- ggplot(number_of_amr_genes_per_sample, aes(x = Country, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpkm_boxplot.png"), plot = p, width = 5, height = 4)

p
```


```{r overall_graphs_per_group}
number_of_amr_genes_per_sample_group <- amr_summarised_meta_with_data %>% filter(!(Group == 'Carrier' & Country == 'Bangladesh')) %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
number_of_amr_genes_per_sample_group$Group <- factor(number_of_amr_genes_per_sample_group$Group, levels = c("Control", "Acute typhoid", "Carrier"))

number_of_amr_genes_per_sample_group  %>% group_by(Group) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
# number of amr genes per sample, by country
p <- ggplot(number_of_amr_genes_per_sample_group, aes(x = Group, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  ylab('Number of AMR genes identified per sample') +
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_boxplot_group.png"), plot = p, width = 5, height = 4)

p
# rpk amr genes per sample by country
p <- ggplot(number_of_amr_genes_per_sample_group, aes(x = Group, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpkm_boxplot_group.png"), plot = p, width = 5, height = 4)

p
```


```{r amr_with_antibiotic_treatment}

number_of_amr_genes_per_sample_abu <- amr_summarised_meta_with_data %>% filter(Group == 'Acute typhoid') %>% filter(n > 0) %>% group_by(sample) %>% summarise(n = sum(n), sum_reads_per_kb = sum(sum_reads_per_kb))
number_of_amr_genes_per_sample_abu_meta <- left_join(number_of_amr_genes_per_sample_abu, metadata, by = join_by(sample == Lane))

number_of_amr_genes_per_sample_abu_meta  %>% group_by(Antibiotics_taken_before_sampling_yes_no_assumptions) %>% summarise(median_n = median(n), median_sum_rpk = median(sum_reads_per_kb))  %>% kbl() %>% kable_styling()
number_of_amr_genes_per_sample_abu_meta$Antibiotics_taken_before_sampling_yes_no_assumptions <- factor(number_of_amr_genes_per_sample_abu_meta$Antibiotics_taken_before_sampling_yes_no_assumptions, levels = c("Yes", "No", "Unknown"))

p <- ggplot(number_of_amr_genes_per_sample_abu_meta, aes(x = Antibiotics_taken_before_sampling_yes_no_assumptions, y = n)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45)
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_number_amu.png"), plot = p, width = 5, height = 4)

p

p <- ggplot(number_of_amr_genes_per_sample_abu_meta, aes(x = Antibiotics_taken_before_sampling_yes_no_assumptions, y = sum_reads_per_kb)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(alpha = 0.1, height = 0) +
  stat_compare_means() + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  scale_y_log10()
ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs', "overall_rpk_amu.png"), plot = p, width = 5, height = 4)

p

```

```{r ncbi_amr_graphs}

# amr_data_with_ncbi_cln_meta_summarised <- amr_data_with_ncbi_cln_meta %>% group_by(sample, Class) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% left_join(metadata, by = join_by(sample == Lane))
amr_summarised_meta_with_data$Group <- factor(amr_summarised_meta_with_data$Group, levels = c("Control", "Acute typhoid", "Carrier"))

# classes <- unique(amr_summarised_meta_with_data$Class)
# cut down classes with some manual removal of low numbers
classes <- c("AMINOGLYCOSIDE", "BETA-LACTAM", "LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN", "TETRACYCLINE", "TRIMETHOPRIM", "EFFLUX", "SULFONAMIDE")

# by country by group, number of genes
for (drug_class in classes) {
  # print(drug_class)
  drug_class_data <- filter(amr_summarised_meta_with_data, Class == drug_class)
  # View(drug_class_data)
  p <- ggviolin(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group", draw_quantiles = 0.5) + 
    stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
    stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
    rremove("x.text") + 
    rremove("xlab") + 
    rotate_x_text(angle = 45) +
    ggtitle(drug_class) +
    geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
  ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/number/by_country_by_group',  paste(gsub("/", "_", drug_class), "number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, number of genes
# remove carrier because not valid in bangladesh
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class, Group != 'Carrier')
    p <- ggboxplot(drug_class_data, y = "n", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/number/by_country',  paste(gsub("/", "_", drug_class), "number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}
# by country, by group, rpkm
for (drug_class in classes) {
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class)
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    # print(drug_class)
    # View(drug_class_data)
    
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) + scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/rpkb/by country by group', paste(gsub("/", "_", drug_class), "rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, rpkm
for (drug_class in classes) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data %>% filter(Class == drug_class, Group != 'Carrier')
    # add 1 to the reads so that the log plot works ok
    drug_class_data <- mutate(drug_class_data, sum_reads_per_kb = ifelse(sum_reads_per_kb == 0, 1, sum_reads_per_kb))
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = log10(max(drug_class_data$sum_reads_per_kb) * 0.9)) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0) +
      scale_y_log10()
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/rpkb/by country', paste(gsub("/", "_", drug_class), "rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```

```{r esbl_data}
esbls <- amr_data_with_ncbi_cln %>% filter(grepl('extended-spectrum', `Product name`))
esbls_template <- expand.grid(sample = samples, Subclass = 'ESBL')

esbls_meta <- left_join(esbls, metadata, by = join_by(sample == Lane))
esbls_summarised <- esbls_meta %>% group_by(sample) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))
esbls_all_samples_summarised <- left_join(esbls_template, esbls_summarised, by = join_by(sample == sample)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb))
esbls_summarised_meta <- left_join(esbls_all_samples_summarised, metadata, by = join_by(sample == Lane))
esbls_summarised_meta$Group <- factor(esbls_summarised_meta$Group, levels = c("Control", "Acute typhoid", "Carrier"))

```

```{r esbl_graphs}

p <- ggboxplot(esbls_summarised_meta, facet.by = "Country", y = "n", x = "Group", color = "Group") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
p
p <- ggboxplot(esbls_summarised_meta, y = "sum_reads_per_kb", x = "Country", color = "Country") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  scale_y_log10()
p

# ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)

write_csv(esbls_summarised_meta, file.path('~/Desktop', 'esbls_summarised_meta.csv'))

p <- ggboxplot(esbls_summarised_meta, y = "n", x = "Country", color = "Country") + 
  stat_compare_means(label.y = max(esbls_summarised_meta$n) * 0.9) + 
  stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +
  rremove("x.text") + 
  rremove("xlab") + 
  rotate_x_text(angle = 45) +
  ggtitle('ESBLs') +
  geom_jitter(aes(color = Country), width = 0.2, height = 0, alpha = 0.2)
p
```


```{r betalactam_graphs}
# 1. set the samples and classes to be analysed (samples set above)
# 2. make an empty data frame, with just the samples and classes
amr_summarised_betalac <- expand.grid(sample = samples, Subclass = subclasses)
# 3. add the metadata
amr_summarised_betalac_meta <- left_join(amr_summarised_betalac, metadata, by = join_by(sample == Lane))
# 4. take the found genes (can just use amr_data_with_ncbi_cln from above)
# 5. summarise the data (by Subclass)
amr_data_with_ncbi_cln_betalac_summarised <- amr_data_with_ncbi_cln %>% group_by(sample, Subclass) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb))

# 6. join the data to the template (amr_summarised_betalac_meta)
amr_summarised_meta_with_data_betalac <- left_join(amr_summarised_betalac_meta, amr_data_with_ncbi_cln_betalac_summarised, by = join_by(sample == sample, Subclass == Subclass)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb))
View(amr_summarised_meta_with_data_betalac %>% filter(Subclass == 'CARBAPENEM', n > 0))

# only four carbapenem samples, so taking out.
amr_summarised_meta_with_data_betalac <- amr_summarised_meta_with_data_betalac %>% filter(!(Subclass == 'CARBAPENEM'))
# after joining, everything that doesnt have a value in n is a zero (i.e. no genes of that class present in that sample)

amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control", "Acute typhoid", "Carrier"))

# amr_summarised_meta_with_data_betalac$Group <- factor(amr_summarised_meta_with_data_betalac$Group, levels = c("Control", "Acute typhoid", "Carrier"))
subclasses <- c('BETA-LACTAM', 'CEPHALOSPORIN')

# by country by group, boxplot of number of genes
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "n", x = "Group", color = "Group") + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0, alpha = 0.2)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of number of genes
for (drug_class in subclasses) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class, Group != 'Carrier')
    p <- ggboxplot(drug_class_data, y = "n", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$n) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) +       rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number',  paste(gsub("/", "_", drug_class), "bl_subclass_number_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country by group, boxplot of rpkm
for (drug_class in subclasses) {
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class)
    p <- ggboxplot(drug_class_data, facet.by = "Country", y = "sum_reads_per_kb", x = "Group", color = "Group", add = 'jitter') + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Control', 'Acute typhoid'), c('Control', 'Carrier'), c('Acute typhoid', 'Carrier'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/number', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot.png", sep = '_')), plot = p, width = 10, height = 6)
}

# by country, boxplot of rpkm
for (drug_class in subclasses) {
    # remove carrier because not valid in bangladesh
    drug_class_data <- amr_summarised_meta_with_data_betalac %>% filter(Subclass == drug_class, Group != 'Carrier')
    # View(drug_class_data)
    p <- ggboxplot(drug_class_data, y = "sum_reads_per_kb", x = "Country", outlier.shape = NA) + 
      stat_compare_means(label.y = max(drug_class_data$sum_reads_per_kb) * 0.9) + 
      stat_compare_means(comparisons = list(c('Nepal', 'Malawi'), c('Nepal', 'Bangladesh'), c('Bangladesh', 'Malawi'))) + 
      rremove("x.text") + 
      rremove("xlab") + 
      rotate_x_text(angle = 45) +
      ggtitle(drug_class) +
      geom_jitter(aes(color = Group), width = 0.2, height = 0)
    # print(drug_class)
    ggsave(filename =  file.path('/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/rgi/2023.08.30/ncbi_graphs/beta_lactam_subclasses/rpk', paste(gsub("/", "_", drug_class), "bl_subclass_rpkb_boxplot_country.png", sep = '_')), plot = p, width = 10, height = 6)
}

```

Do an analysis of only genes found in Typhi. allele_names and gene_family_names come from the pathogen watch paper supp table 2.

```{r typhi_only}
allele_names <- c('blaCMY-2', 'blaOXA-1', 'blaOXA-11', 'blaOXA-15', 'blaOXA-7', 'blaTEM-1', 'blaTEM-10', 'blaTEM-124', 'cmlA1')
gene_family_names <- c('blaAIM', 'blaBIC', 'blaCTX-M', 'blaGES', 'blaIMI', 'blaIMP', 'blaKPC', 'blaLMB', 'blaNDM', 'blaSHV', 'blaSIM', 'blaSPM', 'blaVIM', 'dfrA1', 'dfrA15', 'dfrA17', 'dfrA18', 'dfrA5', 'dfrA7', 'drfA14', 'erm(A)', 'erm(B)', 'erm(C)', 'mcr-1', 'mcr-2', 'mcr-3', 'mcr-4', 'mcr-5', 'mcr-6', 'mcr-7', 'mcr-8', 'mph(A)', 'mph(B)', 'qnrA', 'qnrB', 'qnrD', 'qnrS', 'sul1', 'sul2', 'tet(A)', 'tet(B)', 'tet(C)', 'tet(D)')

amr_data_typhi <- amr_data_with_ncbi_cln %>% filter(`#Allele` %in% allele_names | `Gene family` %in% gene_family_names)
typhi_classes <- unique(amr_data_typhi_meta$Class)
samples <- metadata %>% pull(Lane)

# 1. make a template with all the samples and classes
amr_summarised_typhi <- expand.grid(sample = samples, `Class` = typhi_classes)

# 2. count the number of genes per sample, per class.
tmp <- amr_data_typhi_meta %>% group_by(sample, Class) %>% summarise(n = n()) #%>% kbl() %>% kable_styling()

# 3. join the data to the template (amr_summarised_typhi)
amr_summarised_typhi_d <- left_join(amr_summarised_typhi, tmp, by = join_by(sample == sample, Class == Class)) %>% mutate(n = ifelse(is.na(n), 0, n)) %>% mutate(at_least_one = ifelse(n > 0, 1, 0)) 

# 4. join to metadata
amr_summarised_typhi_d_meta <- left_join(amr_summarised_typhi_d, select(metadata, c(Lane, Country, Group)), by = join_by(sample == Lane))


# 5. Group by class, country, and summarise
typhi_amr_class_country <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Country) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 
  # kbl() %>% kable_styling()

# function to run chi sq test on a given antibiotic class, assuming country and class.
run_chisq_on_amr <- function(df, antibiotic_class){
  test <- df %>% 
    ungroup() %>% 
    filter(Class == antibiotic_class) %>% 
    mutate(n_samples_without_gene = n_samples - n_samples_with_gene) %>% 
    select(n_samples_with_gene, n_samples_without_gene)
  test <- as.data.frame(test)

  # row.names(test) <- test$Country
  # test <- test[, -which(names(test) == "Country")]
  chisq.test(test)
}

# 6. run chi-sq tests on each antibiotic class
run_chisq_on_amr(typhi_amr_class_country, 'SULFONAMIDE')
run_chisq_on_amr(typhi_amr_class_country, 'TETRACYCLINE')
run_chisq_on_amr(typhi_amr_class_country, 'MACROLIDE')
run_chisq_on_amr(typhi_amr_class_country, 'QUINOLONE')
run_chisq_on_amr(typhi_amr_class_country, 'TRIMETHOPRIM')
run_chisq_on_amr(typhi_amr_class_country, 'BETA-LACTAM')
run_chisq_on_amr(typhi_amr_class_country, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN')

# 7. Group by class and group and summarise

typhi_amr_class_group <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Group) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 

typhi_amr_class_group %>% kbl() %>% kable_styling()

# 8. run chi sq 

run_chisq_on_amr(typhi_amr_class_group, 'SULFONAMIDE')
run_chisq_on_amr(typhi_amr_class_group, 'TETRACYCLINE')
run_chisq_on_amr(typhi_amr_class_group, 'MACROLIDE')
run_chisq_on_amr(typhi_amr_class_group, 'QUINOLONE')
run_chisq_on_amr(typhi_amr_class_group, 'TRIMETHOPRIM')
run_chisq_on_amr(typhi_amr_class_group, 'BETA-LACTAM')
run_chisq_on_amr(typhi_amr_class_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN')

# 7. Group by class, country, group, and summarise
typhi_amr_class_country_group <- amr_summarised_typhi_d_meta %>% 
  group_by(Class, Country, Group) %>% 
  summarise(n_samples = n(), n_samples_with_gene = sum(at_least_one), proportion_samples_with_gene = n_samples_with_gene / n_samples, n_genes = sum(n)) #%>% 
  # kbl() %>% kable_styling()

# function to run chi sq test on a given antibiotic class, assuming country, class and group.
run_chisq_on_amr_w_group <- function(df, antibiotic_class, country){
  test <- df %>% 
    ungroup() %>% 
    filter(Class == antibiotic_class, Country == country) %>% 
    select(Country, Group, n_samples_with_gene, n_samples) %>% 
    mutate(n_samples_without_gene = n_samples - n_samples_with_gene) %>% 
    select(-n_samples, Country)
  test <- as.data.frame(test)
  row.names(test) <- test$Group
  test <- test[, -which(names(test) == "Group")]
  test <- test[, -which(names(test) == "Country")]
  View(test)
  fisher.test(test)
}

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'BETA-LACTAM', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'LINCOSAMIDE/MACROLIDE/STREPTOGRAMIN', 'Nepal')

run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Bangladesh')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Malawi')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'MACROLIDE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'QUINOLONE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'SULFONAMIDE', 'Nepal')

# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Bangladesh')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Malawi')
# run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TETRACYCLINE', 'Nepal')

run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Bangladesh')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Malawi')
run_chisq_on_amr_w_group(typhi_amr_class_country_group, 'TRIMETHOPRIM', 'Nepal')

typhi_amr_class_country_group %>% kbl() %>% kable_styling()

amr_data_typhi_meta <- left_join(amr_data_typhi, metadata, by = join_by(sample == Lane))

# amr_data_typhi_meta %>% group_by(`Gene family`, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% arrange(`Gene family`, Country, Group) %>% kbl() %>% kable_styling()
# get all the unique class names
typhi_classes <- unique(amr_data_typhi_meta$Class)
# make an empty data frame, with just the classes, countries, and groups.
typhi_amr_classes <- expand.grid(Class = typhi_classes, Country = c('Bangladesh', 'Malawi', 'Nepal'), Group = c('Control', 'Acute typhoid', 'Carrier'))
# group by class, country, group, and summarise
tmp <- amr_data_typhi_meta %>% group_by(Class, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) %>% kbl() %>% kable_styling()
# join the data to the template (typhi_amr_classes), set any NAs as 0.
typhi_amr_classes <- left_join(typhi_amr_classes, tmp, by = join_by(Class == Class, Country == Country, Group == Group)) %>% 
  mutate(n = ifelse(is.na(n), 0, n)) %>% 
  mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% 
  arrange(Class, Country, Group)
# print the table
typhi_amr_classes %>% kbl() %>% kable_styling()

# same as above, but for gene families
typhi_gene_family_names <- unique(amr_data_typhi_meta$`Gene family`)
typhi_amr_gene_families <- expand.grid(`Gene family` = typhi_gene_family_names, Country = c('Bangladesh', 'Malawi', 'Nepal'), Group = c('Control', 'Acute typhoid', 'Carrier'))

tmp <- amr_data_typhi_meta %>% group_by(`Gene family`, Country, Group) %>% summarise(n = n(), sum_reads_per_kb = sum(reads_per_kb)) # %>% kbl() %>% kable_styling()
typhi_amr_families <- left_join(typhi_amr_gene_families, tmp, by = join_by(`Gene family` == `Gene family`, Country == Country, Group == Group)) %>% 
  mutate(n = ifelse(is.na(n), 0, n)) %>% 
  mutate(sum_reads_per_kb = ifelse(is.na(sum_reads_per_kb), 0, sum_reads_per_kb)) %>% 
  arrange(`Gene family`, Country, Group)

typhi_amr_families %>% kbl() %>% kable_styling()
```

## random forest analysis

Pre-processing.

```{r}
strataa_metaphlan_data_t <- as.data.frame(t(strataa_metaphlan_data))
names(strataa_metaphlan_data_t) <- gsub("\\|", "_", names(strataa_metaphlan_data_t))
strataa_metaphlan_data_t <- strataa_metaphlan_data_t %>% select(contains('_s__')) %>% select(!contains('_t__'))
strataa_metaphlan_data_t$sample_name <- rownames(strataa_metaphlan_data_t)
strataa_metaphlan_metadata_copy <- strataa_metaphlan_metadata %>% select(Group, Country) %>% mutate(sample_name = rownames(.))

# join the Group column from metadata to the metaphlan data
strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_t %>% left_join(strataa_metaphlan_metadata_copy, by = "sample_name") %>% filter(Group != "Acute typhoid", Country != 'Bangladesh') %>% mutate(Group = as.factor(Group))
row.names(strataa_metaphlan_data_heal_carr_rf) <- strataa_metaphlan_data_heal_carr_rf$sample_name
strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_heal_carr_rf %>% select(-sample_name) %>% filter(!is.na(Group))
# remove rows where the group is acute typhi and remove any column that sums to 0
# strataa_metaphlan_data_heal_carr_rf <- strataa_metaphlan_data_rf %>% filter(G #%>% select(where(~ !is.numeric(.) || sum(.) != 0))

```

RF - simple version.

```{r}
library(randomForest)


set.seed(12343)  # For reproducibility
train_index <- sample(1:nrow(strataa_metaphlan_data_heal_carr_rf), 0.8*nrow(strataa_metaphlan_data_heal_carr_rf))
train_data <- strataa_metaphlan_data_heal_carr_rf[train_index, ]
test_data <- strataa_metaphlan_data_heal_carr_rf[-train_index, ]

rf_model <- randomForest(Group ~ ., data=train_data, ntree=1500)

predictions <- predict(rf_model, test_data)
confusionMatrix <- table(test_data$Group, predictions)
print(confusionMatrix)
accuracy <- sum(diag(confusionMatrix)) / sum(confusionMatrix)
print(accuracy)

importance <- varImpPlot(rf_model)
View(importance)

```

CLR transform, makes no difference.

```{r rf_clr}

# Adding a small constant to avoid log(0)
strataa_metaphlan_data_adj <- strataa_metaphlan_data_t %>% select(-sample_name)
strataa_metaphlan_data_adj <- strataa_metaphlan_data_adj + 1e-6

# Calculating the geometric mean for each row
geom_mean <- apply(strataa_metaphlan_data_adj, 1, function(x) exp(mean(log(x))))

# Applying the CLR transformation
clr_transformed <- sweep(strataa_metaphlan_data_adj, 1, geom_mean, FUN = "/")
clr_transformed <- log(clr_transformed)

clr_transformed$sample_name <- rownames(clr_transformed)


clr_transformed_heal_carr_rf <- clr_transformed %>% left_join(strataa_metaphlan_metadata_copy, by = "sample_name") %>% filter(Group != "Acute typhoid", Country != 'Bangladesh') %>% mutate(Group = as.factor(Group))
row.names(clr_transformed_heal_carr_rf) <- clr_transformed_heal_carr_rf$sample_name
clr_transformed_heal_carr_rf <- clr_transformed_heal_carr_rf %>% select(-sample_name) %>% filter(!is.na(Group))

```


```{r clr_rf}
set.seed(12345)  # For reproducibility
train_index <- sample(1:nrow(clr_transformed_heal_carr_rf), 0.8*nrow(clr_transformed_heal_carr_rf))
train_data <- clr_transformed_heal_carr_rf[train_index, ]
test_data <- clr_transformed_heal_carr_rf[-train_index, ]

rf_model <- randomForest(Group ~ ., data=train_data, ntree=500)

predictions <- predict(rf_model, test_data)
confusionMatrix <- table(test_data$Group, predictions)
print(confusionMatrix)
accuracy <- sum(diag(confusionMatrix)) / sum(confusionMatrix)
print(accuracy)

importance <- varImpPlot(rf_model)
View(importance)
```

RF - more complex one

Using a strategy from one of Kevin Bonham's papers https://www.biorxiv.org/content/10.1101/2020.02.13.944181v5.full


# this creates three folds, each of which has 2/3rds of the data in it.
  # this is because for a 3 fold CV, we use 2/3 as training, and 1/3 as test.

```{r rf_complex}

library(randomForest)
library(caret)

# Assuming strataa_metaphlan_data_heal_carr_rf is your dataset
data <- strataa_metaphlan_data_heal_carr_rf
n_repeats <- 10  # Number of CV repetitions
n_folds <- 3      # Number of folds in each CV
n_rng_states <- 10  # Number of RNG states per fold

# Initialize list to store results
all_feature_importances <- vector("list", length = n_repeats)
all_accuracies <- vector("list", length = n_repeats)

# Set a seed for reproducibility
set.seed(123)

# Perform 100 repetitions of 3-fold CV
for (rep in 1:n_repeats) {
  
  # Create indices for the 3-fold CV
  fold_indices <- createFolds(data$Group, k = n_folds, list = TRUE, returnTrain = TRUE)
  
  # Initialize list to store fold results for this repetition
  rep_feature_importances <- vector("list", length = n_folds)
  rep_accuracies <- matrix(NA, nrow = n_folds, ncol = n_rng_states)
  
  # Loop over each fold
  for (fold in 1:n_folds) {
    
    # Get training and testing sets
    train_indices <- fold_indices[[fold]]
    test_indices <- setdiff(1:nrow(data), train_indices)
    train_data <- data[train_indices, ]
    test_data <- data[test_indices, ]
    
    # Initialize matrix to store feature importances for this fold
    fold_feature_importances <- matrix(NA, nrow = n_rng_states, ncol = ncol(data) - 1)  # Assuming the first column is the response
    
    # Loop over each RNG state
    for (rng_state in 1:n_rng_states) {
      
      # Set RNG state
      set.seed((rep * 1000) + (fold * 100) + rng_state)  # Unique seed for each model
      
      # Fit the random forest model
      rf <- randomForest(Group ~ ., data = train_data, ntree = 500)
      
      # Store the feature importance
      fold_feature_importances[rng_state, ] <- importance(rf)
      
      # Predict on the test set and calculate accuracy
      pred <- predict(rf, test_data)
      rep_accuracies[fold, rng_state] <- sum(pred == test_data$Group) / length(test_data$Group)
    }
    
    # Store the feature importances for this fold
    rep_feature_importances[[fold]] <- fold_feature_importances
  }
  
  # Store results of this repetition
  all_feature_importances[[rep]] <- rep_feature_importances
  all_accuracies[[rep]] <- rep_accuracies
}

# Now, `all_feature_importances` is a list of lists of matrices, where each matrix contains the feature importances for a fold and RNG state
# `all_accuracies` is a list of matrices, where each matrix contains the accuracies for each fold and RNG state

# Example of how to access feature importance for repetition 1, fold 1, RNG state 1
# print(all_feature_importances[[1]][[1]][1, ])

```



```{r}

library(randomForest)
library(caret)
# library(dplyr)
# library(tidyr)

# Assuming strataa_metaphlan_data_heal_carr_rf is your dataset
data <- strataa_metaphlan_data_heal_carr_rf %>% select(-Country)

# Parameters for the CV process
n_repeats <- 100
n_folds <- 3
n_rng_states <- 10

# Data frames to store the results
accuracy_df <- data.frame(rep = integer(), 
                          fold = integer(), 
                          rng_state = integer(), 
                          data_type = character(),  # Indicates train or test data
                          accuracy = numeric())

importance_df <- data.frame(feature = character(), 
                            rep = integer(), 
                            fold = integer(), 
                            rng_state = integer(), 
                            importance = numeric())

# Start the cross-validation process
for (rep in 1:n_repeats) {
  fold_indices <- createFolds(data$Group, k = n_folds, list = TRUE)
  
  for (fold in 1:n_folds) {
    train_indices <- fold_indices[[fold]]
    test_indices <- setdiff(1:nrow(data), train_indices)
    train_data <- data[train_indices, ]
    test_data <- data[test_indices, ]
    
    for (rng_state in 1:n_rng_states) {
      set.seed(1000 * rep + 100 * fold + rng_state)
      rf <- randomForest(Group ~ ., data = train_data, ntree = 500)
      
      # Calculate accuracy on training data
      train_pred <- predict(rf, train_data)
      train_accuracy <- sum(train_pred == train_data$Group) / nrow(train_data)
      
      # Calculate accuracy on test data
      test_pred <- predict(rf, test_data)
      test_accuracy <- sum(test_pred == test_data$Group) / nrow(test_data)
      
      # Add to accuracy dataframe
      accuracy_df <- rbind(accuracy_df, 
                           data.frame(rep = rep, 
                                      fold = fold, 
                                      rng_state = rng_state, 
                                      data_type = "Train",
                                      accuracy = train_accuracy),
                           data.frame(rep = rep, 
                                      fold = fold, 
                                      rng_state = rng_state, 
                                      data_type = "Test",
                                      accuracy = test_accuracy))
      
      # Capture feature importance
      imp <- as.data.frame(importance(rf))
      imp$feature <- row.names(imp)
      imp$rep <- rep
      imp$fold <- fold
      imp$rng_state <- rng_state
      
      # Reshape for tidy data and bind to the main importance dataframe
      imp_long <- pivot_longer(imp, 
                               cols = -c(feature, rep, fold, rng_state), 
                               names_to = "metric", 
                               values_to = "importance")
      
      # We only need the MeanDecreaseGini (or MeanDecreaseAccuracy) column here
      imp_filtered <- imp_long %>% 
                      filter(metric == "MeanDecreaseGini") %>% 
                      select(-metric)
      
      importance_df <- rbind(importance_df, imp_filtered)
    }
  }
}

# Clean the importance_df to have columns as specified
importance_df <- importance_df %>% 
                 select(feature, rep, fold, rng_state, importance)

# Print the data frames to check
print(head(accuracy_df))
print(head(importance_df))

# save(accuracy_df, importance_df, file = "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.RData")

# test_accuracy_df <- accuracy_df
# test_importance_df <- importance_df
```

```{r}

accuracy_df_w <- accuracy_df %>%
             pivot_wider(names_from = data_type, values_from = accuracy, 
                         names_prefix = "accuracy_") %>%
             mutate(sqrt_product_accuracy = sqrt(accuracy_Train * accuracy_Test)) %>%
             select(rep, fold, rng_state, sqrt_product_accuracy)


joined_df <- left_join(importance_df, accuracy_df_w, by = c("rep", "fold", "rng_state"))

# Calculate sqrt_product_accuracy weighted importance
joined_df <- joined_df %>%
             mutate(weighted_importance = sqrt_product_accuracy * importance)

# Group by feature and calculate the average of the weighted importances
average_importance_df <- joined_df %>%
                         group_by(feature) %>%
                         summarise(average_weighted_importance = mean(weighted_importance, na.rm = TRUE)) %>%
                         arrange(desc(average_weighted_importance))

# report the median and second and third quartiles

average_accuracy <- accuracy_df %>% group_by(data_type) %>% summarise(average_accuracy = median(accuracy, na.rm = TRUE), second_quantile = quantile(accuracy, 0.25, na.rm = TRUE), third_quantile = quantile(accuracy, 0.75, na.rm = TRUE))
View(average_accuracy)


# accuracy_df %>% filter(data_type == 'Test') %>% ggplot(aes(x = accuracy)) + geom_histogram(bins = 20)

# View the result
print(head(average_importance_df, 10))
save(average_importance_df, file = "/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.final.RData")
write_csv(average_importance_df, '/Users/flashton/Dropbox/GordonGroup/STRATAA_Microbiome/from_leo/Leonardos_analysis/random_forest/2024.01.09/rf_results.feature_importance.csv')
```

