---
title: "STRATAA taxonomic analysis"
author: "Leonardos Mageiros"
date: "08/06/2020"
output:
  html_document: default
  pdf_document: default
  word_document: default
---

# Taxonomical analysis of the STRATAA data at the species level

After executing kracken2 and bracken, here we analyse our taxonomic results fpr the STRATAA metagenomics project

## Load packages and set up environmental variables

We are going to need the following R packages and a source code file with all the fanctions we are going to use: 

```{r Setup, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Here we change the library path to local.
#Knitr has a problem producing files when operating in a network drive.
#Ignore that if you already have an R isntalation that opperates localy. 
#.libPaths("C:/CustomR")

library(broom)
library(tidyverse)
library(stringr)
library(readxl)
library(reshape2)
library(RColorBrewer)
library(viridis)
library(heatmaply)
library(ggplot2)
library(readr)
library(ggfortify)
library(cluster)
library(autoplotly)
library(htmlwidgets)
library(compositions)
library(zCompositions)
library(vegan)
library(dendextend)
library(dplyr)
library(ggpubr)
library(edgeR)
library(rmarkdown)
library(gridExtra)
library(microbiome)
library(phyloseq)
library(apeglm)
library(vsn)
library(DESeq2)
library(edgeR)
library(metacoder)
library(taxa)
library(purrr)
library(ggrepel)
library(devtools)
library(hrbrthemes)
library(MASS)

#this is a file with some functions we are going to use to anlyse our results
source("/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/bin/parse_bracken_functions.r")
source("/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/bin/pairwise_beta.R")
source("/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/bin/diff_expr.r")

```

## Define globals

We need to set specific variables for our analysis first. These contain the working dierectory, the metadata csv file, the out put of braken etc.

```{r Define working dierectory, load data and define variables, message = FALSE}


#STRATAA
# - set this as the location of this script.
knitr::opts_knit$set(root.dir = '/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/phil_running/3_analysis/reports')

#give metadata file
# either get this from email form leo, or convert the excel one.
meta <- read.csv("/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/0_metadata/full_meta.txt", header=T, sep = "\t")
names(meta)[names(meta) == "sample_ID"] <- "isolate"
names(meta)[names(meta) == "ID"] <- "isolate"

# 
#define the braken output folder, the level we are operating and a folder for all the output
# this folder should contain the the uncompressed contents of the rar file
# might want to do genus because have more power to detect effects with a smaller number of genera rather than species.
# do both, and then compare between the two to check consistent
# the species and genus level are separate bracken analyses
braken_folder = "/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/1_taxonomic_profiling/bracken_output/species/"
level = "species"
out_folder = "/Users/flashton/Dropbox/STRATAA_Microbiome_from_Leo/Leonardos_analysis/phil_running/1_species/"

#give filename patern as a reg exp
samples_regexp = "\\d+_\\d+_\\d+" #for patch and strata

```

## Parse, filter, load

The first step will be to read the braken output, parse the files, filter low abundance species (those present in <10% of samples) and load the results in a data frame. We are also storing everything in our hard drive. Step by step what this functon does is: 

* Check if the filtered file with the parsed bracken data exists. If yes load it and return
* Otherwise it reads all the files in the braken_folder 
* Puts them all in one file + create a histogram of the values
* Save all the files in the hard drive

```{r Parse, filter and load data, message = FALSE, results = FALSE}

data_table <- filter_data(braken_folder, level, samples_regexp, out_folder)

```
## Normalisation

There are three reasons that make normalisation a necessary step in metagenomic data. 

* Different sequencing depth between different samples
* Sparcity (data contain lots of 0 values)
* Compositionality (there is a fixed number reads that each sample can obtain. Thinking in terms of abundances the sum will always be 1)

There are a lot of ways to normalize metagenomic datasets. But most of them belong to one out of 3 gategories: 

* Rarefying (which actually subsamples the data to account for deferent depths - not recomended)
* Scaling (Which they calculate a factor the every value in our data will be multiplied with)
* Transforming the datafrom the linear to the logarithmic scale 

# he's using TMM rather than CLR, not much difference between them though.
We have implemented TMM (scaling) and CLR (logarithmic transformation)

The normalised tables are stored in the hard drive.  


```{r Normalise, message = FALSE, results = FALSE, warning=FALSE}

data_table.normalised.clr <- t(clr_transform(data_table, "kraken_assigned_reads", "clr_normalised_data.txt", out_folder))
colnames(data_table.normalised.clr) = gsub(pattern = "X", replacement = "", x = colnames(data_table.normalised.clr)) 

data_table.normalised.tmm <- t(tmm_transformation(data_table, "kraken_assigned_reads", "tmm_normalised_data.txt", out_folder))
colnames(data_table.normalised.tmm) = gsub(pattern = "X", replacement = "", x = colnames(data_table.normalised.tmm))

# beta and alpha diversity analyses require unnormalised data, as they do their own normalisation steps.
data_table.unnormalised <- data.matrix(read.csv("../1_species/summarised_filtered_species_otu.txt", header=T, sep = "\t"))
colnames(data_table.unnormalised) = gsub(pattern = "X", replacement = "", x = colnames(data_table.unnormalised))
```


We can observe the distribution between our unnormalised data and the two normalisation methods we used by creating scaterplots comparing the corresponding values:

```{r normalisation scatter plots, message = FALSE, results = FALSE}

x<-as.vector(data_table.unnormalised)

y<-as.vector(data_table.normalised.tmm)
g1 <- ggplot() + geom_point(aes(x=x, y=y)) + labs(x="Unnormalised values", y="TMM normalised values")

y2<-as.vector(data_table.normalised.clr)
g2 <- ggplot() + geom_point(aes(x=x, y=y2)) + labs(x="Unnormalised values", y="CLR normalised values")

g <- arrangeGrob(g1, g2, ncol=2) #generates g
plot(g)
# might be better to do png plot
ggsave(file= paste(out_folder, "1_normalized_data/normalization_distributions.pdf", sep=""), g)
```

It appears that our methods are performing equally good. 



## Beta diversity and PCoA

The PCA step performed above is actually identical with PCoA of Euclidian distances. 
Euclidian distances although simple and familiar can be missleading in metagenomic data. 

A more robust approach to visualise the same thing is to calculate pairwise Beta diversity among all samples and do PCoA using Bray-Curtis dissimilarities matrix. 

Lets see how the PCA plots differ from PCoA plots. 

```{r Beta diversity - PCoA , message = FALSE, results = FALSE}

# beta and alpha diversity analyses require unnormalised data, as they do their own normalisation steps.
calculate_beta(data_table.unnormalised, meta, quote("Country"), out_folder, "Country", level)

calculate_beta(data_table.unnormalised, meta, quote("Group"), out_folder, "Phenotype", level)





```

Let us also try and see how the phenotypes are looking if we seperate our data per city. 

```{r Beta diversity per city - PCoA , message = FALSE, results = FALSE}


sampledata <- sample_data(meta %>% remove_rownames %>% column_to_rownames(var="isolate"))
OTU = otu_table(t(data_table.unnormalised), taxa_are_rows = TRUE)
pseq <- phyloseq(OTU, sampledata)

malawi <- subset_samples(pseq, Country == "Malawi")
malawi.meta <- tibble::rownames_to_column(meta(malawi), "isolate")
calculate_beta(as.data.frame(t(abundances(malawi))), malawi.meta, quote("Group"), out_folder, "Malawi Phenotype", level)

bangladesh <- subset_samples(pseq, Country == "Bangladesh")
bangladesh.meta <- tibble::rownames_to_column(meta(bangladesh), "isolate")
calculate_beta(as.data.frame(t(abundances(bangladesh))), bangladesh.meta, quote("Group"), out_folder, "Bangladesh Phenotype", level)

nepal <- subset_samples(pseq, Country == "Nepal")
nepal.meta <- tibble::rownames_to_column(meta(nepal), "isolate")
calculate_beta(as.data.frame(t(abundances(nepal))), nepal.meta, quote("Group"), out_folder, "Nepal Phenotype", level)
```


## Alpha diversity 

Another very common step is to calculate the alpha diversity within each sample
This step is not suggested to be performed on normalized data 

```{r PCA alpha, message = FALSE,results = FALSE, warning=FALSE}
#you can change the 4th and 6th argument to perform the analysis using anither metadata column 
calculate_alpha(data_table.unnormalised, "kraken_assigned_reads", meta, quote("Country"), out_folder, "cities", level)


calculate_alpha(data_table, "kraken_assigned_reads", meta, quote("Group"), out_folder, "Phenotype", level, TRUE)


```


## Bar plots

Another common way to visualise the taxonomic profiles of metagenomes is bar plots. Here we create a bar plot per country on the class level. 

We notice that actinobacteria seem to be more prevalent in Bangladesh while Bacteroidia are more common  in Malawi and Nepal. 

```{r bar plots, message = FALSE, warning=FALSE}
#implement graph save to file
create_bar_plot("../phyloseq/input.xlsx", "Country")



```


## Create interactive heatmaply
```{r PCA heatmap, message = FALSE}
heatmaply(as.data.frame(data_table.normalised.clr), showticklabels = c(FALSE, TRUE), colors=colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100), fontsize_row = 5)
```


## Differential abundance
EdgeR uses Trimmed Mean of M-values (TMM). This normalization method is implemented in the edgeR Bioconductor package (version 2.4.0). It is also based on the hypothesis that most genes are not DE. The TMM factor is computed for each lane, with one lane being considered as a reference sample and the others as test samples. For each test sample, TMM is computed as the weighted mean of log ratios between this test and the reference, after exclusion of the most expressed genes and the genes with the largest log ratios. According to the hypothesis of low DE, this TMM should be close to 1. If it is not, its value provides an estimate of the correction factor that must be applied to the library sizes (and not the raw counts) in order to fulfill the hypothesis. The calcNormFactors() function in the edgeR Bioconductor package provides these scaling factors. To obtain normalized read counts, these normalization factors are re-scaled by the mean of the normalized library sizes. Normalized read counts are obtained by dividing raw read counts by these re-scaled normalization factors.

```{r diff edgeR_setup, message = FALSE, warning=FALSE, results = FALSE}

#give metadata file
full_meta <- read.csv("../full_meta.txt", header=T, row.names=1, sep = "\t")
sampledata <- sample_data(full_meta)

#OTU
otu <- read.csv("../1_species/summarised_filtered_species_otu.txt", header=T, row.names=1, sep = "\t")
colnames(otu) = gsub(pattern = "X", replacement = "", x = colnames(otu)) 
OTU = otu_table(otu, taxa_are_rows = TRUE)



#put them in a phyloseq object
pseq <- phyloseq(OTU, sampledata)
full_meta <- meta(pseq)
otu <- abundances(pseq)
```


```{r diff edgeR_all, message = FALSE, warning=FALSE, results = FALSE}
#keep control vs acute 
pseq_control_vs_acute <- subset_samples(pseq, Group =="Control_HealthySerosurvey" | Group =="Acute_Typhi")
subset_meta <- meta(pseq_control_vs_acute)
subset_otu <- t(abundances(pseq_control_vs_acute))
#EdgeR diff expression - HERE AGE SHOULD BE ADDED 
result <- glm.edgeR(x=subset_meta$Group, Y=subset_otu, covariates = subset_meta[ , c('Country', 'Sex')])
topTags(result, n=10)
write.table(topTags(result, n=Inf)$table, file='../1_species/5_glm/edgeR/results_all.txt',sep='\t',quote=FALSE, col.names=NA)

plot_glm(result, "../1_species/5_glm/edgeR/all.pdf", 40, 0.05, "Top 20 - All samples", "Accute Disease", "Control Healthy", "blue", "red")

sig_results <- topTags(result, n=Inf)$table
sig_results <- sig_results[sig_results$FDR < 0.05 , ]
sig_results <- sig_results[sig_results$logFC > 1 | sig_results$logFC < -1, ]
plot_one_gene(sig_results, subset_otu, subset_meta, "Group", "../1_species/5_glm/edgeR/test.pdf")

```


```{r diff per_country, message = FALSE, warning=FALSE, results = FALSE}
#MAlawi only
pseq_malawi <- subset_samples(pseq, Country =="Malawi")
pseq_malawi_control_vs_acute <- subset_samples(pseq_malawi, Group =="Control_HealthySerosurvey" | Group =="Acute_Typhi")
meta_malawi <- meta(pseq_malawi_control_vs_acute)
otu_malawi <- t(abundances(pseq_malawi_control_vs_acute))
#EdgeR diff expression
result_malawi <- glm.edgeR(x=meta_malawi$Group, Y=otu_malawi, covariates = meta_malawi[ , c('Sex', 'Age')])
topTags(result_malawi, n=10)
write.table(topTags(result_malawi, n=Inf)$table, file='../1_species/5_glm/edgeR/malwi_results.txt',sep='\t',quote=FALSE, col.names=NA)

#Nepal only
pseq_nepal <- subset_samples(pseq, Country =="Nepal")
pseq_nepal_control_vs_acute <- subset_samples(pseq_nepal, Group =="Control_HealthySerosurvey" | Group =="Acute_Typhi")
meta_nepal <- meta(pseq_nepal_control_vs_acute)
otu_nepal <- t(abundances(pseq_nepal_control_vs_acute))
#EdgeR diff expression
result_nepal <- glm.edgeR(x=meta_nepal$Group, Y=otu_nepal, covariates = meta_nepal[ , c('Sex', 'Age')])
topTags(result_malawi, n=10)
write.table(topTags(result_nepal, n=Inf)$table, file='../1_species/5_glm/edgeR/nepal_results.txt',sep='\t',quote=FALSE, col.names=NA)

#Bangladesh only




```


